<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quick start for the sommer package ‚Ä¢ sommer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quick start for the sommer package">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quick start for the sommer package</h1>
                        <h4 data-toc-skip class="author">Giovanny
Covarrubias-Pazaran</h4>
            
            <h4 data-toc-skip class="date">2024-09-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/covaruber/sommer/blob/HEAD/vignettes/v1.sommer.quick.start.Rmd"><code>vignettes/v1.sommer.quick.start.Rmd</code></a></small>
      <div class="d-none name"><code>v1.sommer.quick.start.Rmd</code></div>
    </div>

    
    
<p>The sommer package was developed to provide R users with a powerful
and reliable multivariate mixed model solver. ThThe package is focused
on two approaches: 1) p &gt; n (more effects to estimate than
observations) using the mmer() function, and 2) n &gt; p (more
observations than effects to estimate) using the mmec() function. The
core algorithms are coded in C++ using the Armadillo library. This
package allows the user to fit mixed models with the advantage of
specifying the variance-covariance structure for the random effects,
specifying heterogeneous variances, and obtaining other parameters such
as BLUPs, BLUEs, residuals, fitted values, variances for fixed and
random effects, etc.</p>
<p>The purpose of this quick start guide is to show the flexibility of
the package under certain common scenarios:</p>
<p><strong>SECTION 1: Introduction</strong></p>
<ol style="list-style-type: decimal">
<li>Background on mixed models</li>
<li>Background on covariance structures</li>
</ol>
<p><strong>SECTION 2: Different models enabled in sommer</strong></p>
<ol style="list-style-type: decimal">
<li>Univariate homogeneous variance models</li>
<li>Univariate heterogeneous variance models</li>
<li>Univariate unstructured variance models</li>
<li>Multivariate homogeneous variance models</li>
<li>Details on special functions for variance models
<ul>
<li>the major vsr() and vsc() functions for special variance models and
its auxiliars:</li>
<li>atr() and atc() specific levels heterogeneous variance
structure</li>
<li>dsr() and dsc() diagonal covariance structure</li>
<li>usr() and usc() unstructured covariance</li>
<li>csr() and csc() customized covariance structure</li>
<li>isc() function for main effects</li>
</ul>
</li>
<li>The specification of constraints in the variance-covariance
structures
<ul>
<li>unsm() unstructured constraint</li>
<li>fixm() fixed constraint</li>
<li>fcm() constraints on fixed effects</li>
</ul>
</li>
<li>Special functions for special models
<ul>
<li>Random regression models</li>
<li>Overlayed models</li>
<li>Spatial models</li>
<li>GWAS models</li>
<li>Customized random effects</li>
</ul>
</li>
<li>Genomic selection (predicting mendelian sampling)
<ul>
<li>GBLUP</li>
<li>rrBLUP</li>
</ul>
</li>
<li>Likelihood ratio tests</li>
<li>Final remarks</li>
</ol>
<p><strong>SECTION 3: The predict function</strong></p>
<ol style="list-style-type: decimal">
<li>Background on prediction</li>
<li>Predicting means</li>
</ol>
<div class="section level2">
<h2 id="section-1-introduction">SECTION 1: Introduction<a class="anchor" aria-label="anchor" href="#section-1-introduction"></a>
</h2>
<div class="section level3">
<h3 id="background-on-mixed-models">1) Background on mixed models<a class="anchor" aria-label="anchor" href="#background-on-mixed-models"></a>
</h3>
<p>The core of the package is the <code>mmer</code> function which
solves the mixed model equations. The functions are an interface to call
the <code>NR</code> Direct-Inversion Newton-Raphson or Average
Information algorithms (Tunnicliffe 1989; Gilmour et al.¬†1995; Lee et
al.¬†2016). From version 2.0, sommer can handle multivariate models.
Following Maier et al.¬†(2015), the multivariate (and by extension the
univariate) mixed model implemented has the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mn>1</mn></msub><mo>=</mo><msub><mi>X</mi><mn>1</mn></msub><msub><mi>Œ≤</mi><mn>1</mn></msub><mo>+</mo><msub><mi>Z</mi><mn>1</mn></msub><msub><mi>u</mi><mn>1</mn></msub><mo>+</mo><msub><mi>œµ</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">y_1 = X_1\beta_1 + Z_1u_1 + \epsilon_1</annotation></semantics></math></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mn>2</mn></msub><mo>=</mo><msub><mi>X</mi><mn>2</mn></msub><msub><mi>Œ≤</mi><mn>2</mn></msub><mo>+</mo><msub><mi>Z</mi><mn>2</mn></msub><msub><mi>u</mi><mn>2</mn></msub><mo>+</mo><msub><mi>œµ</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">y_2 = X_2\beta_2 + Z_2u_2 + \epsilon_2</annotation></semantics></math></p>
<p>‚Ä¶</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mi>i</mi></msub><mo>=</mo><msub><mi>X</mi><mi>i</mi></msub><msub><mi>Œ≤</mi><mi>i</mi></msub><mo>+</mo><msub><mi>Z</mi><mi>i</mi></msub><msub><mi>u</mi><mi>i</mi></msub><mo>+</mo><msub><mi>œµ</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">y_i = X_i\beta_i + Z_iu_i + \epsilon_i</annotation></semantics></math></p>
<p><br></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>y</mi><mi>i</mi></msub><annotation encoding="application/x-tex">y_i</annotation></semantics></math>
is a vector of trait phenotypes,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ≤</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\beta_i</annotation></semantics></math>
is a vector of fixed effects,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mi>i</mi></msub><annotation encoding="application/x-tex">u_i</annotation></semantics></math>
is a vector of random effects for individuals and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>e</mi><mi>i</mi></msub><annotation encoding="application/x-tex">e_i</annotation></semantics></math>
are residuals for trait ‚Äòi‚Äô (i = 1, ‚Ä¶, t). The random effects
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mn>1</mn></msub><annotation encoding="application/x-tex">u_1</annotation></semantics></math>
‚Ä¶
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mi>i</mi></msub><annotation encoding="application/x-tex">u_i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>e</mi><mi>i</mi></msub><annotation encoding="application/x-tex">e_i</annotation></semantics></math>)
are assumed to be normally distributed with mean zero. X and Z are
incidence matrices for fixed and random effects respectively. The
distribution of the multivariate response and the phenotypic variance
covariance (V) are:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><mi>U</mi><mo>+</mo><msub><mi>œµ</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">Y = X\beta + ZU + \epsilon_i</annotation></semantics></math></p>
<p><br></p>
<p>Y ~
MVN(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>Œ≤</mi></mrow><annotation encoding="application/x-tex">X\beta</annotation></semantics></math>,
V)</p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêò</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mn>1</mn></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mn>2</mn></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mi>t</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{Y} = \left[\begin{array}
{r}
y_1 \\
y_2 \\
... \\
y_t \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêó</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>X</mi><mn>1</mn></msub></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>X</mi><mi>t</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{X} = \left[\begin{array}
{rrr}
X_1 &amp; 0 &amp; 0 \\
\vdots &amp; \ddots &amp; \vdots\\
0 &amp; 0 &amp; X_t \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêï</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msubsup><mi>œÉ</mi><msub><mi>g</mi><mn>1</mn></msub><mn>2</mn></msubsup><msub><mi>Z</mi><mn>1</mn></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mn>1</mn></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msub><mi>œÉ</mi><msub><mi>g</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msub><mi>œÉ</mi><msub><mi>œµ</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msub><mi>œÉ</mi><msub><mi>g</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msub><mi>œÉ</mi><msub><mi>œµ</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mi>t</mi></msub><mi>K</mi><msubsup><mi>œÉ</mi><msub><mi>g</mi><mi>t</mi></msub><mn>2</mn></msubsup><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mi>t</mi></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{V} = \left[\begin{array}
{rrr}
Z_1 K{\sigma^2_{g_{1}}} Z_1' + H{\sigma^2_{\epsilon_{1}}} &amp; ... &amp; Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}} \\
\vdots &amp; \ddots &amp; \vdots\\
Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}} &amp; ... &amp; Z_t K{\sigma^2_{g_{t}}} Z_t' + H{\sigma^2_{\epsilon_{t}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p>where K is the relationship or covariance matrix for the kth random
effect (u=1,‚Ä¶,k), and H=I is an identity matrix or a partial identity
matrix for the residual term. The terms
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{g_{i}}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mi>i</mi></msub><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{\epsilon_{i}}</annotation></semantics></math>
denote the genetic (or any of the kth random terms) and residual
variance of trait ‚Äòi‚Äô, respectively and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></msub></msub><annotation encoding="application/x-tex">\sigma_{g_{_{ij}}}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÉ</mi><msub><mi>œµ</mi><msub><mi></mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></msub></msub><annotation encoding="application/x-tex">\sigma_{\epsilon_{_{ij}}}</annotation></semantics></math>
the genetic (or any of the kth random terms) and residual covariance
between traits ‚Äòi‚Äô and ‚Äòj‚Äô (i=1,‚Ä¶,t, and j=1,‚Ä¶,t). The algorithm
implemented optimizes the log likelihood:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><mi>g</mi><mi>L</mi><mo>=</mo><mn>1</mn><mi>/</mi><mn>2</mn><mo>*</mo><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">|</mo><mi>V</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>X</mi><mi>‚Ä≤</mi><mrow><mo stretchy="true" form="prefix">|</mo><mi>V</mi><mo stretchy="true" form="postfix">|</mo></mrow><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>Y</mi><mi>‚Ä≤</mi><mi>P</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">logL = 1/2 * ln(|V|) + ln(X'|V|X) + Y'PY</annotation></semantics></math></p>
<p><br></p>
<p>where || is the determinant of a matrix. And the REML estimates are
updated using a Newton optimization algorithm of the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>Œ∏</mi><mrow><mi>k</mi><mo>+</mo><mn>1</mn></mrow></msup><mo>=</mo><msup><mi>Œ∏</mi><mi>k</mi></msup><mo>+</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>H</mi><mi>k</mi></msup><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mo>*</mo><mfrac><mrow><mi>d</mi><mi>L</mi></mrow><mrow><mi>d</mi><msubsup><mi>œÉ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></mfrac><mo stretchy="false" form="prefix">|</mo><msup><mi>Œ∏</mi><mi>k</mi></msup></mrow><annotation encoding="application/x-tex">\theta^{k+1} = \theta^{k} + (H^{k})^{-1}*\frac{dL}{d\sigma^2_i}|\theta^k</annotation></semantics></math></p>
<p><br></p>
<p>Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∏</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>
is the vector of variance components for random effects and covariance
components among traits,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>H</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><annotation encoding="application/x-tex">H^{-1}</annotation></semantics></math>
is the inverse of the Hessian matrix of second derivatives for the kth
cycle,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><mi>d</mi><mi>L</mi></mrow><mrow><mi>d</mi><msubsup><mi>œÉ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{dL}{d\sigma^2_i}</annotation></semantics></math>
is the vector of first derivatives of the likelihood with respect to the
variance-covariance components. The Eigen decomposition of the
relationship matrix proposed by Lee and Van Der Werf (2016) was included
in the Newton-Raphson algorithm to improve time efficiency.
Additionally, the popular <code><a href="../reference/vpredict.html">vpredict()</a></code> function to estimate
standard errors for linear combinations of variance components
(i.e.¬†heritabilities and genetic correlations) was added to the package
as well.</p>
<p>Please refer to the canonical papers listed in the Literature section
to check how the algorithms work. We have tested widely the methods to
make sure they provide the same solution when the likelihood behaves
well but for complex problems they might lead to slightly different
answers. If you have any concern please contact me at <a href="mailto:cova_ruber@live.com.mx" class="email">cova_ruber@live.com.mx</a>.</p>
<p>In the following section we will go in detail over several examples
on how to use mixed models in univariate and multivariate case and their
use in quantitative genetics.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="background-on-covariance-structures">2) Background on covariance structures<a class="anchor" aria-label="anchor" href="#background-on-covariance-structures"></a>
</h3>
<p>One of the major strenghts of linear mixed models is the flexibility
to specify variance-covariance structures at all levels. In general,
variance structures of mixed models can be seen as tensor (kronecker)
products of multiple variance-covariance stuctures. For example, a
multi-response model (i.e.¬†2 traits) where ‚Äúg‚Äù individuals (i.e.¬†100
individuals) are tested in ‚Äúe‚Äù treatments (i.e.¬†3 environments), the
variance-covariance for the random effect ‚Äúindividuals‚Äù can be seen as
the following multiplicative model:</p>
<p>T
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>‚äó</mi><annotation encoding="application/x-tex">\otimes</annotation></semantics></math>
G
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>‚äó</mi><annotation encoding="application/x-tex">\otimes</annotation></semantics></math>
A</p>
<p>where:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêì</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>t</mi><mn>1</mn><mo>,</mo><mi>t</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>t</mi><mn>1</mn><mo>,</mo><mi>t</mi><mn>2</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>t</mi><mn>2</mn><mo>,</mo><mi>t</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>t</mi><mn>2</mn><mo>,</mo><mi>t</mi><mn>2</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{T} = \left[\begin{array}
{rr}
{\sigma^2_{g_{_{t1,t1}}}} &amp; {\sigma_{g_{_{t1,t2}}}} \\
{\sigma_{g_{_{t2,t1}}}} &amp; {\sigma^2_{g_{_{t2,t2}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>is the covariance structure for individuals among traits.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêÜ</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{G} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; {\sigma_{g_{_{e1,e2}}}} &amp; {\sigma_{g_{_{e1,e3}}}} \\
{\sigma_{g_{_{e2,e1}}}} &amp; {\sigma^2_{g_{_{e2,e2}}}} &amp; {\sigma_{g_{_{e2,e3}}}} \\
{\sigma_{g_{_{e3,e1}}}} &amp; {\sigma_{g_{_{e3,e2}}}} &amp; {\sigma^2_{g_{_{e3,e3}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>is the covariance structure for individuals among environments.</p>
<p>and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>A</mi><annotation encoding="application/x-tex">A</annotation></semantics></math>
is a square matrix representing the covariance among the levels of the
individuals (any known relationship matrix).</p>
<p>The T and G covariance structures shown above are unknown matrices to
be estimated whereas A is known. The T and G matrices shown above are
called as unstructured (US) covariance matrices, although this type is
just one example from several covariance structures that the linear
mixed models enable. For example, other popular covariance structures
are:</p>
<p>Diagonal (DIAG) covariance structures</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mi>i</mi><mo>,</mo><mi>e</mi><mi>i</mi></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; 0 &amp; 0 \\
\vdots &amp; \ddots &amp; \vdots \\
0 &amp; 0 &amp; {\sigma^2_{g_{_{ei,ei}}}} \\
\end{array}\right]
</annotation></semantics></math> Compound simmetry (CS) covariance
structures</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>œÉ</mi><mrow><mi>g</mi><mi>e</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>œÉ</mi><mrow><mi>g</mi><mi>e</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><mi>g</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>œÉ</mi><mrow><mi>g</mi><mi>e</mi></mrow><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrrr}
{\sigma^2_{g}} + {\sigma^2_{ge}} &amp; {\sigma^2_{g}} &amp; {\sigma^2_{g}} &amp; {\sigma^2_{g}} \\
{\sigma^2_{g}} &amp; {\sigma^2_{g}} + {\sigma^2_{ge}} &amp; {\sigma^2_{g}} &amp; {\sigma^2_{g}}\\
\vdots &amp; \vdots &amp; \ddots &amp; \vdots\\
{\sigma^2_{g}} &amp; {\sigma^2_{g}} &amp; {\sigma^2_{g}} &amp; {\sigma^2_{g}} + {\sigma^2_{ge}}\\
\end{array}\right]
</annotation></semantics></math></p>
<p>First order autoregressive (AR1) covariance structures</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><msup><mi>œÉ</mi><mn>2</mn></msup><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mn>1</mn></mtd><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>2</mn></msup></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>3</mn></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd><mtd columnalign="right" style="text-align: right"><mn>1</mn></mtd><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>2</mn></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>2</mn></msup></mtd><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd><mtd columnalign="right" style="text-align: right"><mn>1</mn></mtd><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>3</mn></msup></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>œÅ</mi><mn>2</mn></msup></mtd><mtd columnalign="right" style="text-align: right"><mi>œÅ</mi></mtd><mtd columnalign="right" style="text-align: right"><mn>1</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \sigma^2 \left[\begin{array}
{rrrr}
1 &amp; {\rho} &amp; {\rho^2} &amp; {\rho^3} \\
{\rho} &amp; 1 &amp; {\rho} &amp; {\rho^2}\\
{\rho^2} &amp; {\rho} &amp; 1 &amp; {\rho} \\
{\rho^3} &amp; {\rho^2} &amp; {\rho}  &amp; 1  \\
\end{array}\right]
</annotation></semantics></math></p>
<p>or the already mentioned Unstructured (US) covariance structures</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; {\sigma_{g_{_{e1,e2}}}} &amp; {\sigma_{g_{_{e1,e3}}}} \\
\vdots &amp; \ddots &amp; \vdots \\
{\sigma_{g_{_{e3,e1}}}} &amp; {\sigma_{g_{_{e3,e2}}}} &amp; {\sigma^2_{g_{_{e3,e3}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>among others. Sommer has the capabilities to fit some of these
covariance structures in the mixed model machinery.</p>
</div>
</div>
<div class="section level2">
<h2 id="section-2-different-models-enabled-in-sommer">SECTION 2: Different models enabled in sommer<a class="anchor" aria-label="anchor" href="#section-2-different-models-enabled-in-sommer"></a>
</h2>
<div class="section level3">
<h3 id="univariate-homogeneous-variance-models">1) Univariate homogeneous variance models<a class="anchor" aria-label="anchor" href="#univariate-homogeneous-variance-models"></a>
</h3>
<p>This type of model refers to single response models where a variable
of interest (i.e.¬†genotypes) needs to be analyzed as interacting with a
2nd random effect (i.e.¬†environments), but you assume that across
environments the genotypes have the same variance component. This is the
so-called compound symmetry (CS) model.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/covaruber/sommer">sommer</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Matrix</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MASS</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: crayon</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span><span class="co">## solving for r records</span></span>
<span><span class="va">ans1r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span> <span class="va">Name</span> <span class="op">+</span> <span class="va">Env</span><span class="op">:</span><span class="va">Name</span>,</span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="va">units</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans1r</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                       VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## Name.Yield-Yield     3.681877 1.6909561 2.177394   Positive</span></span>
<span><span class="co">## Env:Name.Yield-Yield 5.173062 1.4952313 3.459707   Positive</span></span>
<span><span class="co">## units.Yield-Yield    4.366285 0.6470458 6.748031   Positive</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## solving for c coefficients</span></span>
<span><span class="va">ans1c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span> <span class="va">Name</span> <span class="op">+</span> <span class="va">Env</span><span class="op">:</span><span class="va">Name</span>,</span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="va">units</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans1c</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                   VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## Name:isc:isc     3.683153  1.880974 1.958109   Positive</span></span>
<span><span class="co">## Env:Name:isc:isc 5.174116  2.420486 2.137635   Positive</span></span>
<span><span class="co">## units:isc:isc    4.359651  2.270467 1.920156   Positive</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="univariate-heterogeneous-variance-models">2) Univariate heterogeneous variance models<a class="anchor" aria-label="anchor" href="#univariate-heterogeneous-variance-models"></a>
</h3>
<p>Very often in multi-environment trials, the assumption that the
genetic variance or the residual variance is the same across locations
may be too naive. Because of that, specifying a general genetic
component and a location specific genetic variance is the way to go.
This requires a CS+DIAG model (also called heterogeneous CS model).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span> </span>
<span><span class="va">ans2r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span><span class="va">Name</span> <span class="op">+</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">Name</span><span class="op">)</span>,</span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans2r</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                             VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## Name.Yield-Yield           2.962851 1.4962000 1.980251   Positive</span></span>
<span><span class="co">## CA.2011:Name.Yield-Yield  10.146369 4.5073271 2.251083   Positive</span></span>
<span><span class="co">## CA.2012:Name.Yield-Yield   1.877530 1.8697568 1.004158   Positive</span></span>
<span><span class="co">## CA.2013:Name.Yield-Yield   6.629152 2.5028114 2.648682   Positive</span></span>
<span><span class="co">## CA.2011:units.Yield-Yield  4.942450 1.5245057 3.242001   Positive</span></span>
<span><span class="co">## CA.2012:units.Yield-Yield  5.724963 1.3123015 4.362536   Positive</span></span>
<span><span class="co">## CA.2013:units.Yield-Yield  2.559880 0.6399685 4.000010   Positive</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span><span class="op">=</span><span class="va">DT</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">DT</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">ans2c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span><span class="va">Name</span> <span class="op">+</span> <span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsc.html">dsc</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">Name</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsc.html">dsc</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans2c</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                             VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## Name:isc:isc               2.961884 1.4528416 2.038683   Positive</span></span>
<span><span class="co">## Env:Name:CA.2011:CA.2011  10.148234 4.5621620 2.224435   Positive</span></span>
<span><span class="co">## Env:Name:CA.2012:CA.2012   1.878549 1.8883269 0.994822   Positive</span></span>
<span><span class="co">## Env:Name:CA.2013:CA.2013   6.629514 2.4890963 2.663422   Positive</span></span>
<span><span class="co">## Env:units:CA.2011:CA.2011  4.942515 1.5201275 3.251382   Positive</span></span>
<span><span class="co">## Env:units:CA.2012:CA.2012  5.724688 1.3325966 4.295890   Positive</span></span>
<span><span class="co">## Env:units:CA.2013:CA.2013  2.559825 0.6395343 4.002639   Positive</span></span></code></pre>
<p>As you can see the special function <code>atr</code> or
<code>diag</code> can be used to indicate that there‚Äôs a different
variance for the genotypes in each environment. The same was done for
the residual. The difference between <code>atr</code> and
<code>diag</code> is that the <code>atr</code> function can be used to
specify the levels or specific environments where the variance is
different.</p>
</div>
<div class="section level3">
<h3 id="unstructured-variance-models">3) Unstructured variance models<a class="anchor" aria-label="anchor" href="#unstructured-variance-models"></a>
</h3>
<p>A more relaxed asumption than the CS+DIAG model is the unstructured
model (US) which assumes that among the levels of certain factor
(i.e.¬†Environments) there‚Äôs a covariance struture of a second random
effect (i.e.¬†Genotypes). This can be done in sommer using the
<code>usr(.)</code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span></span>
<span><span class="va">ans3r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/usr.html">usr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">Name</span><span class="op">)</span>,</span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>, </span>
<span>             data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans3r</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                                     VarComp VarCompSE    Zratio Constraint</span></span>
<span><span class="co">## CA.2011:Name.Yield-Yield         15.6650010 5.4206906 2.8898534   Positive</span></span>
<span><span class="co">## CA.2012:CA.2011:Name.Yield-Yield  6.1101600 2.4850649 2.4587527   Unconstr</span></span>
<span><span class="co">## CA.2012:Name.Yield-Yield          4.5296090 1.8208107 2.4876881   Positive</span></span>
<span><span class="co">## CA.2013:CA.2011:Name.Yield-Yield  6.3844808 3.0658977 2.0824181   Unconstr</span></span>
<span><span class="co">## CA.2013:CA.2012:Name.Yield-Yield  0.3929997 1.5233985 0.2579757   Unconstr</span></span>
<span><span class="co">## CA.2013:Name.Yield-Yield          8.5972750 2.4837814 3.4613654   Positive</span></span>
<span><span class="co">## CA.2011:units.Yield-Yield         4.9698460 1.5322540 3.2434870   Positive</span></span>
<span><span class="co">## CA.2012:units.Yield-Yield         5.6729333 1.3007862 4.3611574   Positive</span></span>
<span><span class="co">## CA.2013:units.Yield-Yield         2.5570940 0.6392821 3.9999462   Positive</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span><span class="op">=</span><span class="va">DT</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">DT</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">ans3c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span> <span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/usc.html">usc</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">Name</span><span class="op">)</span><span class="op">)</span>,</span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsc.html">dsc</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans3c</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                             VarComp VarCompSE    Zratio Constraint</span></span>
<span><span class="co">## Env:Name:CA.2011:CA.2011  14.492642 3.3536970 4.3213927   Positive</span></span>
<span><span class="co">## Env:Name:CA.2011:CA.2012   5.804743 1.9402333 2.9917760   Unconstr</span></span>
<span><span class="co">## Env:Name:CA.2012:CA.2012   4.456845 2.4466754 1.8215925   Positive</span></span>
<span><span class="co">## Env:Name:CA.2011:CA.2013   5.628452 1.5551412 3.6192546   Unconstr</span></span>
<span><span class="co">## Env:Name:CA.2012:CA.2013   0.487732 1.4943128 0.3263922   Unconstr</span></span>
<span><span class="co">## Env:Name:CA.2013:CA.2013   8.152593 2.0718946 3.9348494   Positive</span></span>
<span><span class="co">## Env:units:CA.2011:CA.2011  4.960032 1.5272879 3.2476074   Positive</span></span>
<span><span class="co">## Env:units:CA.2012:CA.2012  5.634496 1.3028875 4.3246219   Positive</span></span>
<span><span class="co">## Env:units:CA.2013:CA.2013  2.557155 0.6385661 4.0045273   Positive</span></span></code></pre>
<p>As can be seen the <code>usr(Env)</code> indicates that the genotypes
(Name) can have a covariance structure among environments (Env).</p>
</div>
<div class="section level3">
<h3 id="multivariate-homogeneous-variance-models">4) Multivariate homogeneous variance models<a class="anchor" aria-label="anchor" href="#multivariate-homogeneous-variance-models"></a>
</h3>
<p>Currently there‚Äôs a great push for multi-response models. This is
motivated by the correlation that certain variables hide and that could
benefit in the prediction perspective. In sommer to specify multivariate
models the response requires the use of the <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code>
function in the response, and the <code>usr(trait)</code>,
<code>diag(trait)</code>, or <code>atr(trait)</code> functions in the
random part of the model.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">EnvName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">Env</span>,<span class="va">DT</span><span class="op">$</span><span class="va">Name</span><span class="op">)</span></span>
<span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">Yield</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">Yield</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">Weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">Weight</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ans4r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">Yield</span>, <span class="va">Weight</span><span class="op">)</span> <span class="op">~</span> <span class="va">Env</span>,</span>
<span>             random<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Name</span>, Gtc<span class="op">=</span><span class="fu"><a href="../reference/unsm.html">unsm</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>             rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span>, Gtc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans4r</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                         VarComp  VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## u:Name.Yield-Yield    0.2448029 0.07278135 3.363540   Positive</span></span>
<span><span class="co">## u:Name.Yield-Weight   0.3177637 0.07457957 4.260733   Unconstr</span></span>
<span><span class="co">## u:Name.Weight-Weight  0.2813217 0.08054357 3.492789   Positive</span></span>
<span><span class="co">## u:units.Yield-Yield   0.3725767 0.04331247 8.602066   Positive</span></span>
<span><span class="co">## u:units.Weight-Weight 0.3586951 0.04189443 8.561882   Positive</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/reshape.html" class="external-link">reshape</a></span><span class="op">(</span><span class="va">DT</span>, idvar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Name"</span>,<span class="st">"Env"</span>,<span class="st">"Block"</span><span class="op">)</span>, varying <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">6</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>,</span>
<span>        v.names <span class="op">=</span> <span class="st">"y"</span>, direction <span class="op">=</span> <span class="st">"long"</span>, timevar <span class="op">=</span> <span class="st">"trait"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">DT2</span><span class="op">$</span><span class="va">trait</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT2</span><span class="op">$</span><span class="va">trait</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ans4c &lt;- mmec(y ~ Env:trait,</span></span>
<span><span class="co">#              random= ~ vsc(usc(trait),isc(Name)),</span></span>
<span><span class="co">#              rcov= ~ vsc(dsc(trait),isc(units)), returnParam = T, </span></span>
<span><span class="co">#              data=DT2, verbose = T)</span></span>
<span><span class="co"># summary(ans4c)$varcomp</span></span></code></pre></div>
<p>You may notice that we have added the <code>usr(trait)</code> behind
the random effects. This is to indicate the structure that should be
assumed in the multivariate model. The <code>diag(trait)</code> used
behind a random effect (i.e.¬†Name) indicates that for the traits modeled
(Yield and Weight) there‚Äôs no covariance component and should not be
estimated, whereas <code>usr(trait)</code> assumes that for such a
random effect, there‚Äôs a covariance component to be estimated
(i.e.¬†covariance between Yield and Weight for the random effect Name).
The same applies for the residual part (rcov).</p>
<p>Any number of random effects can be specified with different
structures.</p>
</div>
<div class="section level3">
<h3 id="details-on-special-functions-for-variance-models">5) Details on special functions for variance models<a class="anchor" aria-label="anchor" href="#details-on-special-functions-for-variance-models"></a>
</h3>
<div class="section level4">
<h4 id="the-vsr-function-for-special-variance-models-and-its-auxiliaries">the vsr() function for special variance models and its
auxiliaries<a class="anchor" aria-label="anchor" href="#the-vsr-function-for-special-variance-models-and-its-auxiliaries"></a>
</h4>
<p>The sommer function <code><a href="../reference/vsr.html">vsr()</a></code> allows constructing complex
variance models that are passed to the <code><a href="../reference/mmer.html">mmer()</a></code> function
which constitutes one of the most important features of the sommer
package. The specification of the <code><a href="../reference/vsr.html">vsr()</a></code> function has the
form:</p>
<p><code>random=~vsr(..., Gu, Gti, Gtc)</code></p>
<p>The idea is that the <code><a href="../reference/vsr.html">vsr()</a></code> function reflects the special
variance structure that each random effect could have in the matrix
notation:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>u</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>T</mi><mo>‚®Ç</mo><mi>E</mi><mo>‚®Ç</mo><mi>.</mi><mi>.</mi><mi>.</mi><mo>‚®Ç</mo><mi>A</mi></mrow><annotation encoding="application/x-tex">var(u) = T \bigotimes E \bigotimes ... \bigotimes A</annotation></semantics></math></p>
<p>where the ‚Ä¶ argument in the <code><a href="../reference/vsr.html">vsr()</a></code> function is used to
specify the kronecker products from all matrices that form the variance
for the random effect , where the auxiliary functions
<code><a href="../reference/dsr.html">dsr()</a></code>, <code><a href="../reference/usr.html">usr()</a></code>, <code><a href="../reference/csr.html">csr()</a></code>,
<code><a href="../reference/atr.html">atr()</a></code>, can be used to define such structured variance
structures. The idea is that a variance model for a random effect x
(i.e.¬†individuals) might require a more flexible model than just:</p>
<p><code>random=~x</code></p>
<p>For example, if individuals are tested in different time-points and
environments, we can assume a different variance and covariance
components among the individuals in the different environment-timepoint
combinations. An example of variance structure of the type:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>u</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>T</mi><mo>‚®Ç</mo><mi>E</mi><mo>‚®Ç</mo><mi>S</mi><mo>‚®Ç</mo><mi>A</mi></mrow><annotation encoding="application/x-tex">var(u) = T \bigotimes E \bigotimes S \bigotimes A</annotation></semantics></math></p>
<p>would be specified in the <code><a href="../reference/vsr.html">vsr()</a></code> function as:</p>
<p><code>random=~vsr(usr(e),usr(s),x, Gu=A, Gtc=T)</code></p>
<p>where the <code>e</code> would be a column vector in a data frame for
the environments, <code>s</code> a column vector in the dataframe for
the time points, <code>x</code> is the vector in the datrame for the
identifier of individuals, <code>A</code> is a known square variance
covariance matrix among individuals (usually an identity matrix; default
if not specified), and T is a square matrices with as many rows and
columns as the number of traits that specifies the trait covariance
structure.</p>
<p>The auxiliary functions to build the variance models for the random
effect are: + dsr() diagonal covariance structure + usr() unstructured
covariance + atr() specific levels heterogeneous variance structure +
csr() customized covariance structure</p>
</div>
<div class="section level4">
<h4 id="the-vsc-function-for-special-variance-models-and-its-auxiliaries">the vsc() function for special variance models and its
auxiliaries<a class="anchor" aria-label="anchor" href="#the-vsc-function-for-special-variance-models-and-its-auxiliaries"></a>
</h4>
<p>The sommer function <code><a href="../reference/vsc.html">vsc()</a></code> allows constructing complex
variance models that are passed to the <code><a href="../reference/mmec.html">mmec()</a></code> function
which constitutes one of the most important features of the sommer
package. The specification of the <code><a href="../reference/vsc.html">vsc()</a></code> function has the
form:</p>
<p><code>random=~vsc(..., Gu)</code></p>
<p>where the different variance structures are input as before but the
constraints are specified within the auxiliar function with the argument
theta and thetaC. Please notice that all functions for c coefficients
end in the letter c instead or r.</p>
</div>
<div class="section level4">
<h4 id="dsr-and-dsc-to-specify-a-diagonal-diag-covariance-structures">dsr() and dsc() to specify a diagonal (DIAG) covariance
structures<a class="anchor" aria-label="anchor" href="#dsr-and-dsc-to-specify-a-diagonal-diag-covariance-structures"></a>
</h4>
<p>A diagonal covariance structure looks like this:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mi>i</mi><mo>,</mo><mi>e</mi><mi>i</mi></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; 0 &amp; 0 \\
\vdots &amp; \ddots &amp; \vdots \\
0 &amp; 0 &amp; {\sigma^2_{g_{_{ei,ei}}}} \\
\end{array}\right]
</annotation></semantics></math> Considering an example for one random
effect (g; indicating i.e.¬†individuals) evaluated in different treatment
levels (e; indicating i.e.¬†the different treatments) the model would
look like:</p>
<p><code>random=~vsr(dsr(e),g)</code> for r records
<code>random=~vsc(dsc(e),isc(g))</code> for c coefficients</p>
</div>
<div class="section level4">
<h4 id="usr-and-usc-to-specify-an-unstructured-us-covariance">usr() and usc() to specify an unstructured (US) covariance<a class="anchor" aria-label="anchor" href="#usr-and-usc-to-specify-an-unstructured-us-covariance"></a>
</h4>
<p>An unstructured covariance looks like this:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêÜ</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>2</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>3</mn><mo>,</mo><mi>e</mi><mn>3</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{G} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; {\sigma_{g_{_{e1,e2}}}} &amp; {\sigma_{g_{_{e1,e3}}}} \\
{\sigma_{g_{_{e2,e1}}}} &amp; {\sigma^2_{g_{_{e2,e2}}}} &amp; {\sigma_{g_{_{e2,e3}}}} \\
{\sigma_{g_{_{e3,e1}}}} &amp; {\sigma_{g_{_{e3,e2}}}} &amp; {\sigma^2_{g_{_{e3,e3}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>Considering same example for one random effect (g; indicating
i.e.¬†individuals) evaluated in different treatment levels (e; indicating
i.e.¬†the different treatments) the model would look like:</p>
<p><code>random=~vsr(usr(e),g)</code> for r records
<code>random=~vsc(usc(e),isc(g))</code> for c coefficients</p>
</div>
<div class="section level4">
<h4 id="atr-and-atc-to-specify-a-level-specific-heterogeneous-variance">atr() and atc() to specify a level-specific heterogeneous
variance<a class="anchor" aria-label="anchor" href="#atr-and-atc-to-specify-a-level-specific-heterogeneous-variance"></a>
</h4>
<p>A diagonal covariance structure for specific levels of the second
random effect looks like this:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mi>i</mi><mo>,</mo><mi>e</mi><mi>i</mi></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; 0 &amp; 0 \\
\vdots &amp; \ddots &amp; \vdots \\
0 &amp; 0 &amp; {\sigma^2_{g_{_{ei,ei}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>Considering the same example for one random effect (g; indicating
i.e.¬†individuals) evaluated in different treatment levels (e; indicating
i.e.¬†the different treatments A,B,C) the model would look like:</p>
<p><code>random=~vsr(atr(e,c("A","B")),g)</code>
<code>random=~vsc(atc(e,c("A","B")),isc(g))</code></p>
<p>where the variance component for g is only fitted at levels A and
B.</p>
</div>
<div class="section level4">
<h4 id="csr-and-csc-to-specify-a-level-specific-variance-covariance-structure">csr() and csc() to specify a level-specific variance-covariance
structure<a class="anchor" aria-label="anchor" href="#csr-and-csc-to-specify-a-level-specific-variance-covariance-structure"></a>
</h4>
<p>A customized covariance structure for specific levels of the second
random effect (variance and covariances) looks i.e.¬†like this:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùö∫</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>1</mn></mrow></msub></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mn>1</mn><mo>,</mo><mi>e</mi><mn>2</mn></mrow></msub></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><msubsup><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>e</mi><mi>i</mi><mo>,</mo><mi>e</mi><mi>i</mi></mrow></msub></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{\Sigma} = \left[\begin{array}
{rrr}
{\sigma^2_{g_{_{e1,e1}}}} &amp; {\sigma_{g_{_{e1,e2}}}} &amp; 0 \\
\vdots &amp; \ddots &amp; \vdots \\
0 &amp; 0 &amp; {\sigma^2_{g_{_{ei,ei}}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p>Considering same example for one random effect (g; indicating
i.e.¬†individuals) evaluated in different treatment levels (e; indicating
i.e.¬†the different treatments A,B,C) the model would look like:</p>
<p><code>random=~vsr(csr(e,c("A","B")),g)</code>
<code>random=~vsc(csc(e,c("A","B")),isc(g))</code></p>
<p>where <code>mm</code> indicates which variance and covariance
components are estimated for <code>g</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-specification-of-multi-trait-constraints-in-the-variance-components-for-the-vsr-function-gtc-argument">6) The specification of multi-trait constraints in the variance
components for the vsr function (Gtc argument)<a class="anchor" aria-label="anchor" href="#the-specification-of-multi-trait-constraints-in-the-variance-components-for-the-vsr-function-gtc-argument"></a>
</h3>
<p>One of the major strengths of sommer is its extreme flexibility to
specify variance-covariance structures in the multi-trait framework.
Since sommer 3.7 this is easily achieved by the use of the
<code><a href="../reference/vsr.html">vsr()</a></code> function and it‚Äôs argument <code>Gtc</code>. The
<code>Gtc</code> argument expects a matrix of constraints for the
variance-covariance components for the random effect filled with numbers
according to the following rules:</p>
<ul>
<li>0: parameter not to be estimated</li>
<li>1: estimated and constrained to be positive</li>
<li>2: estimated and unconstrained</li>
<li>3: not to be estimated but fixed value provided in
<code>Gti</code>
</li>
</ul>
<p>Some useful functions to specify quickly the contraint matrices are
<code><a href="../reference/unsm.html">unsm()</a></code> for unstructured, <code><a href="../reference/fixm.html">fixm()</a></code> for fixed
constraint, and <code><a href="../reference/fcm.html">fcm()</a></code> for fixed effect constraints.</p>
<p>Consider a multi-trait model with 4 traits (y1,‚Ä¶,y4) and 1 random
effect (u) and 1 fixed effect (x)</p>
<p><code>fixed=cbind(y1,y2,y3,y4)~x</code></p>
<p><code>random= ~vsr(u, Gtc=?)</code></p>
<p>The constraint for the 4 x 4 matrix of variance covariance components
to be estimated can be an:</p>
<ol style="list-style-type: lower-alpha">
<li>unstructured (variance components have to be positive and
covariances either positive or negative)
<code>random= ~vsr(u, Gtc=unsm(4))</code> and
<code>random= ~vsr(dsc(x,thetaC=unsm(4)),isc(u))</code>
</li>
</ol>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/unsm.html">unsm</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">## [1,]    1    2    2    2</span></span>
<span><span class="co">## [2,]    2    1    2    2</span></span>
<span><span class="co">## [3,]    2    2    1    2</span></span>
<span><span class="co">## [4,]    2    2    2    1</span></span></code></pre>
<ol start="3" style="list-style-type: lower-alpha">
<li>fixed (variance or covariance components indicated with a 3 are
considered fixed and values are provided in the Gti argument)
<code>random= ~vsr(u, Gtc=fixm(4), Gti=mm)</code> and
<code>random= ~vsr(dsc(x,thetaC=fixm(4)),isc(u))</code>
</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fixm.html">fixm</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">## [1,]    3    3    3    3</span></span>
<span><span class="co">## [2,]    3    3    3    3</span></span>
<span><span class="co">## [3,]    3    3    3    3</span></span>
<span><span class="co">## [4,]    3    3    3    3</span></span></code></pre>
<p>where mm is a 4 x 4 matrix with initial values for the variance
components.</p>
<ol start="4" style="list-style-type: lower-alpha">
<li>constraints for fixed effects
<code>fixed= cbind(y1,y2,y3,y4)~vsr(x, Gtc=fcm(c(1,0,1,0)))</code>
</li>
</ol>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fcm.html">fcm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2]</span></span>
<span><span class="co">## [1,]    1    0</span></span>
<span><span class="co">## [2,]    0    0</span></span>
<span><span class="co">## [3,]    0    1</span></span>
<span><span class="co">## [4,]    0    0</span></span></code></pre>
<p>where 1‚Äôs and 0‚Äôs indicate the traits where the fixed effect will be
estimated (1‚Äôs) and where it won‚Äôt (0‚Äôs).</p>
</div>
<div class="section level3">
<h3 id="special-functions-for-special-models">7) Special functions for special models<a class="anchor" aria-label="anchor" href="#special-functions-for-special-models"></a>
</h3>
<div class="section level4">
<h4 id="random-regression-models">Random regression models<a class="anchor" aria-label="anchor" href="#random-regression-models"></a>
</h4>
<p>In order to fit random regression models the user can use the
<code><a href="../reference/leg.html">leg()</a></code> function to fit Legendre polynomials. This can be
combined with other special covariance structures such as
<code>dsr() / dsc()</code>, <code>usr() / usc()</code>, etc.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">orthopolynom</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_legendre</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_legendre</span></span>
<span></span>
<span><span class="va">mRR2r</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Y</span><span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">Xf</span></span>
<span>           , random<span class="op">=</span><span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/usr.html">usr</a></span><span class="op">(</span><span class="fu"><a href="../reference/leg.html">leg</a></span><span class="op">(</span><span class="va">X</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>,<span class="va">SUBJECT</span><span class="op">)</span></span>
<span>           , rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span></span>
<span>           , data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mRR2r</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                         VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## leg0:SUBJECT.Y-Y      2.5782969 0.6717242 3.838326   Positive</span></span>
<span><span class="co">## leg1:leg0:SUBJECT.Y-Y 0.4765431 0.2394975 1.989763   Unconstr</span></span>
<span><span class="co">## leg1:SUBJECT.Y-Y      0.3497299 0.2183229 1.601893   Positive</span></span>
<span><span class="co">## u:units.Y-Y           2.6912226 0.3825197 7.035513   Positive</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mRR2c</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Y</span><span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">Xf</span></span>
<span>           , random<span class="op">=</span><span class="op">~</span> <span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/usc.html">usc</a></span><span class="op">(</span><span class="fu"><a href="../reference/leg.html">leg</a></span><span class="op">(</span><span class="va">X</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>,<span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">SUBJECT</span><span class="op">)</span><span class="op">)</span></span>
<span>           , rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span><span class="op">)</span></span>
<span>           , data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mRR2c</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                       VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## X:SUBJECT:leg0:leg0 2.5631091 0.4926224 5.202989   Positive</span></span>
<span><span class="co">## X:SUBJECT:leg0:leg1 0.4681312 0.1908480 2.452900   Unconstr</span></span>
<span><span class="co">## X:SUBJECT:leg1:leg1 0.3477737 0.2828824 1.229393   Positive</span></span>
<span><span class="co">## units:isc:isc       2.6914702 0.4614278 5.832918   Positive</span></span></code></pre>
<p>Here, a numeric covariate X is used to explain the trajectory of the
SUBJECTs and combined with an unstructured covariance matrix. The
details can be found in the theory.</p>
</div>
<div class="section level4">
<h4 id="gwas-models">GWAS models<a class="anchor" aria-label="anchor" href="#gwas-models"></a>
</h4>
<p>Although genome-wide association studies can be conducted through a
variety of approaches, the use of mixed models to find associations
between markers and phenotypes still one of the most popular approaches.
One of the most classical and popular approaches is to test marker by
marker trough mixed modeling (1 model by marker) to obtain the marker
effect and a statistic reflecting the level of association usually
provided as the -log10 p-value. The second most popular approach is to
assume that the genetic variance component is similar for all markers
and therefore the variance components are only estimated once (1 model
for all markers) and use the inverse of the phenotypic variance matrix
(V.inverse) to test all markers in the generalized linear model
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>b</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>X</mi><mi>V</mi><mo>‚àí</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>‚àí</mo><mi>X</mi><mi>V</mi><mo>‚àí</mo><mi>y</mi></mrow><annotation encoding="application/x-tex">b=(XV-X)-XV-y</annotation></semantics></math>.
This makes the GWAS much faster and efficient without major loses. Given
the straightforward extension, sommer provides the <code>GWAS</code>
function which can fit both type of approaches (be aware that these are
two methods among many existant in the literature) in univariate and
multivariate models, that way genetically correlated traits can be
tested together to increase the power of detection. In summary the GWAS
model implemented in sommer to obtain marker effect is a generalized
linear model of the form:</p>
<p>b = (X‚ÄôV-X)X‚ÄôV-y</p>
<p>with X = ZMi</p>
<p>where: - b is the marker effect (dimensions 1 x mt) - y is the
response variable (univariate or multivariate) (dimensions 1 x nt) - V-
is the inverse of the phenotypic variance matrix (dimensions nt x nt) -
Z is the incidence matrix for the random effect selected (gTerm
argument) to perform the GWAS (dimensions nt x ut) Mi is the ith column
of the marker matrix (M argument) (dimensions u x m)</p>
<p>for t traits, n observations, m markers and u levels of the random
effect. Depending if P3D is TRUE or FALSE the V- matrix will be
calculated once and used for all marker tests (P3D=TRUE) or estimated
through REML for each marker (P3D=FALSE).</p>
<p>Here we show a simple GWAS model for an univariate example.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">#### create the variance-covariance matrix</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="co">#### look at the data and fit the model</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DT</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        id Row Col Year      color  Yield FruitAver Firmness Rowf Colf</span></span>
<span><span class="co">## P003 P003   3   1 2014 0.10075269 154.67     41.93  588.917    3    1</span></span>
<span><span class="co">## P004 P004   4   1 2014 0.13891940 186.77     58.79  640.031    4    1</span></span>
<span><span class="co">## P005 P005   5   1 2014 0.08681502  80.21     48.16  671.523    5    1</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">MP</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 Locus Position Chrom</span></span>
<span><span class="co">## 1  scaffold_77830_839        0     1</span></span>
<span><span class="co">## 2  scaffold_39187_895        0     1</span></span>
<span><span class="co">## 3 scaffold_50439_2379        0     1</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GT</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      scaffold_50439_2381 scaffold_39344_153 uneak_3436043 uneak_2632033</span></span>
<span><span class="co">## P003                   0                  0             0             1</span></span>
<span><span class="co">## P004                   0                  0             0             1</span></span>
<span><span class="co">## P005                   0                 -1             0             1</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>,Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span></span>
<span>             <span class="op">+</span> <span class="va">Rowf</span> <span class="op">+</span> <span class="va">Colf</span>,</span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             M<span class="op">=</span><span class="va">GT</span>, gTerm <span class="op">=</span> <span class="st">"u:id"</span>,</span>
<span>             verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BB0000;">Performing GWAS evaluation</span></span></span>
<span><span class="co"><span style="color: #BB0000;">## </span></span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">mix1</span><span class="op">$</span><span class="va">scores</span><span class="op">)</span></span>
<span><span class="va">ms</span><span class="op">$</span><span class="va">Locus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span></span>
<span><span class="va">MP2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">MP</span>,<span class="va">ms</span>,by<span class="op">=</span><span class="st">"Locus"</span>,all.x <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>;</span>
<span><span class="fu"><a href="../reference/manhattan.html">manhattan</a></span><span class="op">(</span><span class="va">MP2</span>, pch<span class="op">=</span><span class="fl">20</span>,cex<span class="op">=</span><span class="fl">.5</span>, PVCN <span class="op">=</span> <span class="st">"color"</span><span class="op">)</span></span></code></pre></div>
<p><img src="v1.sommer.quick.start_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Be aware that the marker matrix <code>M</code> has to be imputed (no
missing data allowed) and make sure that the number of rows in the M
matrix is equivalent to the levels of the gTerm specified (i.e.¬†if the
gTerm is ‚Äúid‚Äù and has 300 levels or in other words 300 individuals, then
M has dimensions 300 x m, where m is the number of markers).</p>
</div>
<div class="section level4">
<h4 id="overlayed-models-the-overlay-function">Overlayed models [the overlay() function]<a class="anchor" aria-label="anchor" href="#overlayed-models-the-overlay-function"></a>
</h4>
<p>Another very useful function is the <code>overlay</code> function,
which allows for the overlay of matrices of different random effects and
estimate a single variance component for the overlayed terms.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"DT_halfdiallel"</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_halfdiallel</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   rep geno male female     sugar</span></span>
<span><span class="co">## 1   1   12    1      2 13.950509</span></span>
<span><span class="co">## 2   2   12    1      2  9.756918</span></span>
<span><span class="co">## 3   1   13    1      3 13.906355</span></span>
<span><span class="co">## 4   2   13    1      3  9.119455</span></span>
<span><span class="co">## 5   1   14    1      4  5.174483</span></span>
<span><span class="co">## 6   2   14    1      4  8.452221</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span><span class="op">$</span><span class="va">femalef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">female</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">malef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">male</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">genof</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="co">#### model using overlay</span></span>
<span><span class="va">modhr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">sugar</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/overlay.html">overlay</a></span><span class="op">(</span><span class="va">femalef</span>,<span class="va">malef</span><span class="op">)</span><span class="op">)</span> </span>
<span>             <span class="op">+</span> <span class="va">genof</span>, data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">modhc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">sugar</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="fu"><a href="../reference/overlay.html">overlay</a></span><span class="op">(</span><span class="va">femalef</span>,<span class="va">malef</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span>             <span class="op">+</span> <span class="va">genof</span>,data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>here the <code>femalef</code> and <code>malef</code> random effects
are overlayed becoming a single random effect that has the same variance
component.</p>
</div>
<div class="section level4">
<h4 id="spatial-models-using-the-2-dimensional-spline">Spatial models (using the 2-dimensional spline)<a class="anchor" aria-label="anchor" href="#spatial-models-using-the-2-dimensional-spline"></a>
</h4>
<p>We will use the CPdata to show the use of 2-dimensional splines for
accomodating spatial effects in field experiments. In early generation
variety trials the availability of seed is low, which makes the use of
an unreplicated design a neccesity more than anything else. Experimental
designs such as augmented designs and partially-replicated (p-rep)
designs have become more common these days.</p>
<p>In order to do a good job modeling the spatial trends happening in
the field, special covariance structures have been proposed to
accomodate such spatial trends (i.e.¬†autoregressive residuals; ar1).
Unfortunately, some of these covariance structures make the modeling
rather unstable. More recently other research groups have proposed the
use of 2-dimensional splines to overcome such issues and have a more
robust modeling of the spatial terms (Lee et al.¬†2013; Rodr√≠guez-√Ålvarez
et al.¬†2018).</p>
<p>In this example we assume an unreplicated population where row and
range information is available which allows usr to fit a 2 dimensional
spline model.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">### mimic two fields</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span><span class="va">mix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Rowf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Colf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/spl2Da.html">spl2Da</a></span><span class="op">(</span><span class="va">Row</span>,<span class="va">Col</span><span class="op">)</span>,</span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ============================================================</span></span>
<span><span class="co">##          Multivariate Linear Mixed Model fit by REML         </span></span>
<span><span class="co">## **********************  sommer 4.3  ********************** </span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">##          logLik      AIC     BIC Method Converge</span></span>
<span><span class="co">## Value -151.2647 304.5293 308.421     NR    FALSE</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Variance-Covariance components:</span></span>
<span><span class="co">##                     VarComp VarCompSE Zratio Constraint</span></span>
<span><span class="co">## u:id.Yield-Yield      792.6     317.6 2.4959   Positive</span></span>
<span><span class="co">## u:Rowf.Yield-Yield    807.6     371.3 2.1750   Positive</span></span>
<span><span class="co">## u:Colf.Yield-Yield    183.2     139.7 1.3121   Positive</span></span>
<span><span class="co">## A:all.Yield-Yield     515.8     701.3 0.7354   Positive</span></span>
<span><span class="co">## u:units.Yield-Yield  2918.4     292.8 9.9667   Positive</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Fixed effects:</span></span>
<span><span class="co">##   Trait      Effect Estimate Std.Error t.value</span></span>
<span><span class="co">## 1 Yield (Intercept)    132.1     8.761   15.08</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Groups and observations:</span></span>
<span><span class="co">##        Yield</span></span>
<span><span class="co">## u:id     363</span></span>
<span><span class="co">## u:Rowf    13</span></span>
<span><span class="co">## u:Colf    36</span></span>
<span><span class="co">## A:all    168</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Use the '$' sign to access results and parameters</span></span></code></pre>
<p>Notice that the job is done by the <code><a href="../reference/spl2Da.html">spl2Da()</a></code> function
that takes the <code>Row</code> and <code>Col</code> information to fit
a spatial kernel.</p>
</div>
<div class="section level4">
<h4 id="customized-random-effects">Customized random effects<a class="anchor" aria-label="anchor" href="#customized-random-effects"></a>
</h4>
<p>One of the most powerful features of sommer is the ability to provide
any customized matrix and estimate any random effect. For example:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span></span>
<span><span class="co">#### look at the data and fit the model</span></span>
<span><span class="va">mix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>              random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span><span class="op">)</span>,</span>
<span>              rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>the matrix GT is provided as a random effect by encapsulating the
matrix in a list and provided in the <code><a href="../reference/vsr.html">vsr()</a></code> function.</p>
</div>
</div>
<div class="section level3">
<h3 id="genomic-selection">10) Genomic selection<a class="anchor" aria-label="anchor" href="#genomic-selection"></a>
</h3>
<p>In this section I decided to show the way you can fit an rrBLUP and
GBLUP model in sommer using some wheat example data from CIMMYT in the
genomic selection framework. This is the case of prediction of specific
individuals within a population. It basically uses a similar model of
the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><mi>u</mi><mo>+</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">y = X\beta + Zu + \epsilon</annotation></semantics></math></p>
<p><br></p>
<p>and takes advantage of the variance covariance matrix for the
genotype effect known as the additive relationship matrix (A) and
calculated using the <code>A.mat</code> function to establish
connections among all individuals and predict the BLUPs for individuals
that were not measured. In case the interest is to get BLUPs for markers
the random effect is the actual marker matrix and the relationship among
markers can be specified as well but in this example is assumed to be
diagonal.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_wheat</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_wheat</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_wheat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"X"</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span>;<span class="va">DT</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># select environment 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="co"># GBLUP pedigree-based approach</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">y.trn</span> <span class="op">&lt;-</span> <span class="va">DT</span></span>
<span><span class="va">vv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y.trn</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co">## GBLUP</span></span>
<span><span class="va">ans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">X1</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>,Gu<span class="op">=</span><span class="va">K</span><span class="op">)</span>, </span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">y.trn</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># kinship based</span></span>
<span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"id"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">[</span><span class="va">vv</span>,<span class="op">]</span>,<span class="va">DT</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span>, use<span class="op">=</span><span class="st">"complete"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5733721</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rrBLUP</span></span>
<span><span class="va">ans2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">X1</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span>, buildGu <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, </span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, getPEV <span class="op">=</span> <span class="cn">FALSE</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             data<span class="op">=</span><span class="va">y.trn</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># kinship based</span></span>
<span></span>
<span><span class="va">u</span> <span class="op">&lt;-</span> <span class="va">GT</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">ans2</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:GT`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span> <span class="co"># BLUPs for individuals</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">u</span><span class="op">[</span><span class="va">vv</span>,<span class="op">]</span>,<span class="va">DT</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span><span class="op">)</span> <span class="co"># same correlation</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5771181</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the same can be applied in multi-response models in GBLUP or rrBLUP</span></span>
<span><span class="co"># same can be achieved with the mmec function (see ?DT_wheat)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="likelihood-ratio-tests">11) Likelihood ratio tests<a class="anchor" aria-label="anchor" href="#likelihood-ratio-tests"></a>
</h3>
<p>The Likelihood ratio test (LRT) is a good way to investigate the
significance of random effects or specific variance-covariance
components.</p>
<div class="section level4">
<h4 id="testing-the-significance-of-a-variance-component">11.1) Testing the significance of a variance component<a class="anchor" aria-label="anchor" href="#testing-the-significance-of-a-variance-component"></a>
</h4>
<p>For example, imagine that a researcher would like to know if his
model improves when adding the effect of a spatial kernel to capture the
spatial trend in the field, his base model may look like this:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">### mimic two fields</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Rowf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Colf</span><span class="op">)</span>,</span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>And the model with the spatial kernel is the following:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Rowf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Colf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/spl2Da.html">spl2Da</a></span><span class="op">(</span><span class="va">Row</span>,<span class="va">Col</span><span class="op">)</span>,</span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Then to test if the second model brings value let usr fit the
likelihood ratio test as follows:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">mix1</span>, <span class="va">mix2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Likelihood ratio test for mixed models</span></span>
<span><span class="co">## ==============================================================</span></span>
<span><span class="co">##      Df      AIC      BIC     loLik   Chisq ChiDf  PrChisq</span></span>
<span><span class="co">## mod2  8 304.5293 308.4210 -151.2647                       </span></span>
<span><span class="co">## mod1  7 305.1825 309.0741 -151.5912 0.65315     1 0.41899 </span></span>
<span><span class="co">## ==============================================================</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
<p>As can be seen the test turns out to not be very significant despite
the increase in the likelihood.</p>
</div>
<div class="section level4">
<h4 id="testing-the-significance-of-a-covariance-component">11.2) Testing the significance of a covariance component<a class="anchor" aria-label="anchor" href="#testing-the-significance-of-a-covariance-component"></a>
</h4>
<p>Sometimes the researcher is more interested in knowing if a
covariance structure is relevant or not. Assume we have two multi-trait
models, 1) fitting no-covariance (independent) among traits, and 2) one
fitting the genetic covariance among yield and color in the following
population:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">EnvName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">Env</span>,<span class="va">DT</span><span class="op">$</span><span class="va">Name</span><span class="op">)</span></span>
<span><span class="va">modelBase</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">Yield</span>, <span class="va">Weight</span><span class="op">)</span> <span class="op">~</span> <span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Name</span>, Gtc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, <span class="co"># here is diag()</span></span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span>, Gtc<span class="op">=</span><span class="fu"><a href="../reference/unsm.html">unsm</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">modelCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">Yield</span>, <span class="va">Weight</span><span class="op">)</span> <span class="op">~</span> <span class="va">Env</span>,</span>
<span>              random<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/usr.html">usr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">Name</span>, Gtc<span class="op">=</span><span class="fu"><a href="../reference/unsm.html">unsm</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, <span class="co"># here is unsm()</span></span>
<span>              rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="../reference/dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">units</span>, Gtc<span class="op">=</span><span class="fu"><a href="../reference/unsm.html">unsm</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">modelBase</span>, <span class="va">modelCov</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Likelihood ratio test for mixed models</span></span>
<span><span class="co">## ==============================================================</span></span>
<span><span class="co">##      Df       AIC       BIC    loLik     Chisq ChiDf                  PrChisq</span></span>
<span><span class="co">## mod2 45 -340.4760 -316.9949 176.2380                                         </span></span>
<span><span class="co">## mod1 23 -205.0695 -181.5885 108.5347 135.40646    22 2.57838297804313e-18 ***</span></span>
<span><span class="co">## ==============================================================</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
<p>As can be seen, in this case fitting the covariance among the
genotypes improves the model fit considerably and the probablity of the
Chi-square distribution is &lt; 0.05. Then you the model with the
covariance is the preferred model.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="section-3-the-predict-function">SECTION 3: The predict function<a class="anchor" aria-label="anchor" href="#section-3-the-predict-function"></a>
</h2>
<div class="section level3">
<h3 id="background-on-prediction">1) Background on prediction<a class="anchor" aria-label="anchor" href="#background-on-prediction"></a>
</h3>
<p>In a linear mixed model where y is an nx1 vector of observations, the
linear mixed model can be written as:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>œÑ</mi><mo>+</mo><mi>Z</mi><mi>u</mi><mo>+</mo><mi>e</mi><mo>=</mo><mi>W</mi><mi>Œ≤</mi><mo>+</mo><mi>e</mi></mrow><annotation encoding="application/x-tex">y = X\tau + Zu + e =  W \beta + e</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œÑ</mi><annotation encoding="application/x-tex">\tau</annotation></semantics></math>
is a vector tx1 of fixed effects, X is an nxt design matrix which
associates observations with the appropriate combinations of fixed
effects, u is the qx1 vector of random effects, Z is the nxq matrix
which associates observations with the appropriate random effects, and e
is the nx1 vector of residual errors. For shorthand W and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≤</mi><annotation encoding="application/x-tex">\beta</annotation></semantics></math>
represent the combined design matrix and vector of efects, respectively.
It is assumed:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>u</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>e</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mo>,</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>G</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ≥</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mn>0</mn></mtd><mtd columnalign="right" style="text-align: right"><mi>R</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>œï</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex"> \left[\begin{array}
{r}
u \\
e \\
\end{array}\right] = N(\left[\begin{array}
{r}
0 \\
0 \\
\end{array}\right] , \left[\begin{array}
{rr}
G{(\gamma)} &amp; 0 \\
0 &amp; R{(\phi)} \\
\end{array}\right]) 
</annotation></semantics></math></p>
<p>where the covariance matrices G and R for random effects and
residuals are functions of parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≥</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œï</mi><annotation encoding="application/x-tex">\phi</annotation></semantics></math>
respectively. The covariance matrix of data can be written as:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>y</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mi>œÉ</mi><mn>2</mn></msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><mi>G</mi><mi>Z</mi><mi>‚Ä≤</mi><mo>+</mo><mi>R</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">var(y) = \sigma^2 (ZGZ'+R)</annotation></semantics></math></p>
<p>The variance parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≥</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œï</mi><annotation encoding="application/x-tex">\phi</annotation></semantics></math>
are usually estimated by maximum likelihood or REML (Patterson and
Thompson, 1971).</p>
<p>The BLUP of a linear combination for known D,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≥</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œï</mi><annotation encoding="application/x-tex">\phi</annotation></semantics></math>
is then:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mi>ùêÉ</mi><mover><mi>ùõÉ</mi><mo accent="true">¬Ø</mo></mover></mrow><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>D</mi><mi>œÑ</mi></msub></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>D</mi><mi>u</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mover><mi>œÑ</mi><mo accent="true">¬Ø</mo></mover></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mover><mi>u</mi><mo accent="true">¬Ø</mo></mover></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mrow><mi>D</mi><mover><mi>œÑ</mi><mo accent="true">¬Ø</mo></mover></mrow><mo>+</mo><mrow><msub><mi>D</mi><mi>u</mi></msub><mover><mi>u</mi><mo accent="true">¬Ø</mo></mover></mrow></mrow><annotation encoding="application/x-tex">\mathbf{D\overline{\beta}} = \left[\begin{array}
{rr}
D_{\tau} &amp; D_u \\
\end{array}\right]  \left[\begin{array}
{r}
\overline{\tau} \\
\overline{u} \\
\end{array}\right] = {D \overline{\tau}} + {D_{u}\overline{u}}
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>œÑ</mi><mo>,</mo><mi>u</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\overline{\beta} = (\tau, u)</annotation></semantics></math>
is the solution of the mixed moel equations:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>X</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>X</mi></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>X</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>Z</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>Z</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>X</mi></mtd><mtd columnalign="right" style="text-align: right"><msup><mi>Z</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>Z</mi><mo>+</mo><msup><mi>G</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>œÑ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>u</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>X</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>y</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msup><mi>Z</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>X</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex"> \left[\begin{array}
{rr}
X^TR^{-1}X &amp; X^TR^{-1}Z  \\
Z^TR^{-1}X &amp; Z^TR^{-1}Z + G^{-1} \\
\end{array}\right]  \left[\begin{array}
{r}
\tau \\
u \\
\end{array}\right] = \left[\begin{array}
{r}
X^TR^{-1}y \\
Z^TR^{-1}X \\
\end{array}\right] 
</annotation></semantics></math></p>
<p>These can also be written as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover><mo>=</mo><msup><mi>W</mi><mi>T</mi></msup><msup><mi>R</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mi>y</mi></mrow><annotation encoding="application/x-tex">C \overline{\beta} = W^TR^{-1}y</annotation></semantics></math>.
The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>œÑ</mi><mo accent="true">¬Ø</mo></mover><annotation encoding="application/x-tex">\overline{\tau}</annotation></semantics></math>
is the best linear unbiased estimator (BLUE) of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œÑ</mi><annotation encoding="application/x-tex">\tau</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>u</mi><mo accent="true">¬Ø</mo></mover><annotation encoding="application/x-tex">\overline{u}</annotation></semantics></math>
is the best linear unbiased predictor (BLUP) of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>u</mi><annotation encoding="application/x-tex">u</annotation></semantics></math>,
this with variance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover><mo>‚àí</mo><mi>Œ≤</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mi>C</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup></mrow><annotation encoding="application/x-tex">var(\overline{\beta}-\beta)=C^{-1}</annotation></semantics></math>.
Consideration of the values required to form the confidence intervals
make it clear that is the prediction error variance (PEV or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover><mo>‚àí</mo><mi>Œ≤</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">var(\overline{\beta}-\beta)</annotation></semantics></math>),
rather than the variance of the estimator
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">var(\overline{\beta})</annotation></semantics></math>
that is usually of interest. The prediction error variance for a linear
combination
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover></mrow><annotation encoding="application/x-tex">D\overline{\beta}</annotation></semantics></math>
is then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><msup><mi>C</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><msup><mi>D</mi><mi>T</mi></msup></mrow><annotation encoding="application/x-tex">DC^{-1}D^T</annotation></semantics></math>.
Since the variance parameters are unknown, we replace the unknown
variance parameters by their REML estimates and use the empirical
values.</p>
<p>The <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> function of the sommer package then builts
the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>C</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><annotation encoding="application/x-tex">C^{-1}</annotation></semantics></math>
matrix from the mixed model equations and the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>
matrix of linear combinations to obtain the PEV and SEs for the
predictions. For the means it uses the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>D</mi><annotation encoding="application/x-tex">D</annotation></semantics></math>
matrix of linear combinations times the vector of required fixed and
random effects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mover><mi>œÑ</mi><mo accent="true">¬Ø</mo></mover><mo>+</mo><mi>Z</mi><mover><mi>u</mi><mo accent="true">¬Ø</mo></mover></mrow><annotation encoding="application/x-tex">X\overline{\tau} + Z \overline{u}</annotation></semantics></math>
or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mover><mi>Œ≤</mi><mo accent="true">¬Ø</mo></mover></mrow><annotation encoding="application/x-tex">D\overline{\beta}</annotation></semantics></math>,
where the matrix D is a linear combination of specific fixed and/or
random effects from matrices
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and/or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>.</p>
</div>
<div class="section level3">
<h3 id="predicting-means">2) Predicting means<a class="anchor" aria-label="anchor" href="#predicting-means"></a>
</h3>
<p>The sommer package is equiped with a <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> function
that can be used to calculate means and standard errors for fixed and
random effects specified in the models fitted with the
<code><a href="../reference/mmer.html">mmer()</a></code> and <code>mmec</code> functions. Using the yatesoats
dataset we will will fit some fixed and random effects.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/covaruber/sommer">sommer</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_yatesoats</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_yatesoats</span></span>
<span><span class="va">m3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span>fixed<span class="op">=</span><span class="va">Y</span> <span class="op">~</span> <span class="va">V</span> <span class="op">+</span> <span class="va">N</span> <span class="op">+</span> <span class="va">V</span><span class="op">:</span><span class="va">N</span>,</span>
<span>           random <span class="op">=</span> <span class="op">~</span> <span class="va">B</span> <span class="op">+</span> <span class="va">B</span><span class="op">:</span><span class="va">MP</span>,</span>
<span>           rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>,</span>
<span>           data <span class="op">=</span> <span class="va">DT</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">m3</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##            VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## B.Y-Y     214.4477 168.62790 1.271722   Positive</span></span>
<span><span class="co">## B:MP.Y-Y  106.0508  67.83280 1.563415   Positive</span></span>
<span><span class="co">## units.Y-Y 177.0883  37.34293 4.742217   Positive</span></span></code></pre>
<p>Now, the model can be used together with the <code>classify</code>
argument to obtain means for the <code>classify</code> argument. For
example, the model includes in the fixed formula the terms ‚ÄúV‚Äù for
variety, ‚ÄúN‚Äù for nitrogen treatments, and ‚ÄúV:N‚Äù for the interaction
between variety and nitrogen. The <code>classify</code> argument can be
used to specify the term for which the means are desired. In the
following example the means for the nitrogen treatments are obtained as
follows:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Dt</span> <span class="op">&lt;-</span> <span class="va">m3</span><span class="op">$</span><span class="va">Dtable</span>; <span class="va">Dt</span></span></code></pre></div>
<pre><code><span><span class="co">##     type term include average</span></span>
<span><span class="co">## 1  fixed    1   FALSE   FALSE</span></span>
<span><span class="co">## 2  fixed    V   FALSE   FALSE</span></span>
<span><span class="co">## 3  fixed    N   FALSE   FALSE</span></span>
<span><span class="co">## 4  fixed  V:N   FALSE   FALSE</span></span>
<span><span class="co">## 5 random    B   FALSE   FALSE</span></span>
<span><span class="co">## 6 random B:MP   FALSE   FALSE</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first fixed effect just average</span></span>
<span><span class="va">Dt</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"average"</span><span class="op">]</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="co"># second fixed effect include</span></span>
<span><span class="va">Dt</span><span class="op">[</span><span class="fl">2</span>,<span class="st">"include"</span><span class="op">]</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="co"># third fixed effect include and average</span></span>
<span><span class="va">Dt</span><span class="op">[</span><span class="fl">3</span>,<span class="st">"include"</span><span class="op">]</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="va">Dt</span><span class="op">[</span><span class="fl">3</span>,<span class="st">"average"</span><span class="op">]</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="va">Dt</span></span></code></pre></div>
<pre><code><span><span class="co">##     type term include average</span></span>
<span><span class="co">## 1  fixed    1   FALSE    TRUE</span></span>
<span><span class="co">## 2  fixed    V    TRUE   FALSE</span></span>
<span><span class="co">## 3  fixed    N    TRUE    TRUE</span></span>
<span><span class="co">## 4  fixed  V:N   FALSE   FALSE</span></span>
<span><span class="co">## 5 random    B   FALSE   FALSE</span></span>
<span><span class="co">## 6 random B:MP   FALSE   FALSE</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pp</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">m3</span>, Dtable<span class="op">=</span><span class="va">Dt</span>, D<span class="op">=</span><span class="st">"N"</span><span class="op">)</span></span>
<span><span class="va">pp</span><span class="op">$</span><span class="va">pvals</span></span></code></pre></div>
<pre><code><span><span class="co">##       N predicted.value std.error</span></span>
<span><span class="co">## 0     0        78.16667  13.31581</span></span>
<span><span class="co">## 0.2 0.2        87.41667  14.88566</span></span>
<span><span class="co">## 0.4 0.4        95.50000  14.88566</span></span>
<span><span class="co">## 0.6 0.6       100.58333  14.88566</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="literature">Literature<a class="anchor" aria-label="anchor" href="#literature"></a>
</h2>
<p>Covarrubias-Pazaran G. 2016. Genome assisted prediction of
quantitative traits using the R package sommer. PLoS ONE 11(6):1-15.</p>
<p>Covarrubias-Pazaran G. 2018. Software update: Moving the R package
sommer to multivariate mixed models for genome-assisted prediction. doi:
<a href="https://doi.org/10.1101/354639" class="external-link uri">https://doi.org/10.1101/354639</a></p>
<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants.
Second edition. Stemma Press. 390 pp.</p>
<p>Gilmour et al.¬†1995. Average Information REML: An efficient algorithm
for variance parameter estimation in linear mixed models. Biometrics
51(4):1440-1450.</p>
<p>Henderson C.R. 1975. Best Linear Unbiased Estimation and Prediction
under a Selection Model. Biometrics vol.¬†31(2):423-447.</p>
<p>Kang et al.¬†2008. Efficient control of population structure in model
organism association mapping. Genetics 178:1709-1723.</p>
<p>Lee, D.-J., Durban, M., and Eilers, P.H.C. (2013). Efficient
two-dimensional smoothing with P-spline ANOVA mixed models and nested
bases. Computational Statistics and Data Analysis, 61, 22 - 37.</p>
<p>Lee et al.¬†2015. MTG2: An efficient algorithm for multivariate linear
mixed model analysis based on genomic information. Cold Spring Harbor.
doi: <a href="http://dx.doi.org/10.1101/027201" class="external-link uri">http://dx.doi.org/10.1101/027201</a>.</p>
<p>Maier et al.¬†2015. Joint analysis of psychiatric disorders increases
accuracy of risk prediction for schizophrenia, bipolar disorder, and
major depressive disorder. Am J Hum Genet; 96(2):283-294.</p>
<p>Rodriguez-Alvarez, Maria Xose, et al.¬†Correcting for spatial
heterogeneity in plant breeding experiments with P-splines. Spatial
Statistics 23 (2018): 52-71.</p>
<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML
estimates of variance components. Paper invited for the 1993 American
Statistical Association Meeting, San Francisco.</p>
<p>Yu et al.¬†2006. A unified mixed-model method for association mapping
that accounts for multiple levels of relatedness. Genetics
38:203-208.</p>
<p>Tunnicliffe W. 1989. On the use of marginal likelihood in time series
model estimation. JRSS 51(1):15-27.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
