<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quantitative genetics using the sommer package ‚Ä¢ sommer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantitative genetics using the sommer package">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quantitative genetics using the sommer package</h1>
                        <h4 data-toc-skip class="author">Giovanny
Covarrubias-Pazaran</h4>
            
            <h4 data-toc-skip class="date">2024-09-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/covaruber/sommer/blob/HEAD/vignettes/v3.sommer.qg.Rmd"><code>vignettes/v3.sommer.qg.Rmd</code></a></small>
      <div class="d-none name"><code>v3.sommer.qg.Rmd</code></div>
    </div>

    
    
<p>The sommer package was developed to provide R users with a powerful
and reliable multivariate mixed model solver for different genetic and
non-genetic analyses in diploid and polyploid organisms. This package
allows the user to estimate variance components for a mixed model with
the advantages of specifying the variance-covariance structure of the
random effects, specifying heterogeneous variances, and obtaining other
parameters such as BLUPs, BLUEs, residuals, fitted values, variances for
fixed and random effects, etc. The core algorithms of the package are
coded in C++ using the Armadillo library to optimize dense matrix
operations common in the derect-inversion algorithms. Although the
vignette shows examples using the mmer function, the mmec function can
be faster when working with more records than coefficients to be
estimated and we highly recommend to shift to the use of the
<code>mmec</code> function.</p>
<p>The package is focused on problems of the type p &gt; n related to
genomic prediction (hybrid prediction &amp; genomic selection) and GWAS
analysis, although any general mixed model can be fitted as well. The
package provides kernels to estimate additive (<code>A.mat</code>),
dominance (<code>D.mat</code>), and epistatic (<code>E.mat</code>)
relationship matrices that have been shown to increase prediction
accuracy under certain scenarios or simply to estimate the variance
components of such. The package provides flexibility to fit other
genetic models such as full and half diallel models as well.</p>
<p>The vignettes aim to provide several examples in how to use the
sommer package under different scenarios. We will spend the rest of the
space providing examples for:</p>
<p><strong>SECTION 1: Introduction </strong></p>
<ol style="list-style-type: decimal">
<li>Background in linear algebra</li>
</ol>
<p><strong>SECTION 2: Topics in quantitative genetics </strong></p>
<ol style="list-style-type: decimal">
<li>Heritability
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>)
calculation</li>
<li>Specifying heterogeneous variances in mixed models</li>
<li>Using the <code><a href="../reference/vpredict.html">vpredict()</a></code> calculator</li>
<li>Half and full diallel designs (using the overlay)</li>
<li>Genomic selection (predicting mendelian sampling)
<ul>
<li>GBLUP</li>
<li>rrBLUP</li>
</ul>
</li>
<li>Indirect genetic effects</li>
<li>Single cross prediction (hybrid prediction)</li>
<li>Spatial modeling (using the 2-dimensional splines)</li>
<li>Multivariate genetic models and genetic correlations</li>
</ol>
<p><strong>SECTION 3: Special topics in quantitative
genetics</strong></p>
<ol style="list-style-type: decimal">
<li>Partitioned model</li>
<li>UDU‚Äô decomposition</li>
<li>Mating designs</li>
<li>GWAS by GBLUP</li>
<li>Reduced models</li>
</ol>
<div class="section level2">
<h2 id="section-1-introduction">SECTION 1: Introduction<a class="anchor" aria-label="anchor" href="#section-1-introduction"></a>
</h2>
<div class="section level3">
<h3 id="backgrounds-in-linear-algebra">Backgrounds in linear algebra<a class="anchor" aria-label="anchor" href="#backgrounds-in-linear-algebra"></a>
</h3>
<p>The core of the package is the <code><a href="../reference/mmer.html">mmer()</a></code> and
<code>mmec</code> functions which solve the mixed model equations. The
functions are an interface to call the <code>NR</code> Direct-Inversion
Newton-Raphson or Average Information or mme-based Average Information
(Tunnicliffe 1989; Gilmour et al.¬†1995; Lee et al.¬†2016). Since version
2.0, sommer can handle multivariate models. Following Maier et
al.¬†(2015), the multivariate (and by extension the univariate) mixed
model implemented has the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mn>1</mn></msub><mo>=</mo><msub><mi>X</mi><mn>1</mn></msub><msub><mi>Œ≤</mi><mn>1</mn></msub><mo>+</mo><msub><mi>Z</mi><mn>1</mn></msub><msub><mi>u</mi><mn>1</mn></msub><mo>+</mo><msub><mi>œµ</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">y_1 = X_1\beta_1 + Z_1u_1 + \epsilon_1</annotation></semantics></math></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mn>2</mn></msub><mo>=</mo><msub><mi>X</mi><mn>2</mn></msub><msub><mi>Œ≤</mi><mn>2</mn></msub><mo>+</mo><msub><mi>Z</mi><mn>2</mn></msub><msub><mi>u</mi><mn>2</mn></msub><mo>+</mo><msub><mi>œµ</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">y_2 = X_2\beta_2 + Z_2u_2 + \epsilon_2</annotation></semantics></math></p>
<p>‚Ä¶</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>y</mi><mi>i</mi></msub><mo>=</mo><msub><mi>X</mi><mi>i</mi></msub><msub><mi>Œ≤</mi><mi>i</mi></msub><mo>+</mo><msub><mi>Z</mi><mi>i</mi></msub><msub><mi>u</mi><mi>i</mi></msub><mo>+</mo><msub><mi>œµ</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">y_i = X_i\beta_i + Z_iu_i + \epsilon_i</annotation></semantics></math></p>
<p><br></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>y</mi><mi>i</mi></msub><annotation encoding="application/x-tex">y_i</annotation></semantics></math>
is a vector of trait phenotypes,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ≤</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\beta_i</annotation></semantics></math>
is a vector of fixed effects,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mi>i</mi></msub><annotation encoding="application/x-tex">u_i</annotation></semantics></math>
is a vector of random effects for individuals and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>e</mi><mi>i</mi></msub><annotation encoding="application/x-tex">e_i</annotation></semantics></math>
are residuals for trait <code>i</code> (i = 1, ‚Ä¶, t). The random effects
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mn>1</mn></msub><annotation encoding="application/x-tex">u_1</annotation></semantics></math>
‚Ä¶
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mi>i</mi></msub><annotation encoding="application/x-tex">u_i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>e</mi><mi>i</mi></msub><annotation encoding="application/x-tex">e_i</annotation></semantics></math>)
are assumed to be normally distributed with mean zero. X and Z are
incidence matrices for fixed and random effects respectively. The
distributions of the multivariate response and the phenotypic variance
covariance (V) are:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><mi>U</mi><mo>+</mo><msub><mi>œµ</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">Y = X\beta + ZU + \epsilon_i</annotation></semantics></math></p>
<p><br></p>
<p>Y ~
MVN(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mi>Œ≤</mi></mrow><annotation encoding="application/x-tex">X\beta</annotation></semantics></math>,
V)</p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêò</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mn>1</mn></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mn>2</mn></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mi>t</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{Y} = \left[\begin{array}
{r}
y_1 \\
y_2 \\
... \\
y_t \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêó</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>X</mi><mn>1</mn></msub></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>X</mi><mi>t</mi></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{X} = \left[\begin{array}
{rrr}
X_1 &amp; ... &amp; ... \\
\vdots &amp; \ddots &amp; \vdots\\
... &amp; ... &amp; X_t \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùêï</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">[</mo><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msubsup><mi>œÉ</mi><msub><mi>g</mi><mn>1</mn></msub><mn>2</mn></msubsup><msub><mi>Z</mi><mn>1</mn></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mn>1</mn></msub><mn>2</mn></msubsup></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msub><mi>œÉ</mi><msub><mi>g</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msub><mi>œÉ</mi><msub><mi>œµ</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd><mtd columnalign="right" style="text-align: right"><mo>‚ã±</mo></mtd><mtd columnalign="right" style="text-align: right"><mi>‚ãÆ</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mn>1</mn></msub><mi>K</mi><msub><mi>œÉ</mi><msub><mi>g</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msub><mi>œÉ</mi><msub><mi>œµ</mi><mrow><mn>1</mn><mo>,</mo><mi>t</mi></mrow></msub></msub></mtd><mtd columnalign="right" style="text-align: right"><mi>.</mi><mi>.</mi><mi>.</mi></mtd><mtd columnalign="right" style="text-align: right"><msub><mi>Z</mi><mi>t</mi></msub><mi>K</mi><msubsup><mi>œÉ</mi><msub><mi>g</mi><mi>t</mi></msub><mn>2</mn></msubsup><msub><mi>Z</mi><mi>t</mi></msub><mi>‚Ä≤</mi><mo>+</mo><mi>H</mi><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mi>t</mi></msub><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd></mtr></mtable><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbf{V} = \left[\begin{array}
{rrr}
Z_1 K{\sigma^2_{g_{1}}} Z_1' + H{\sigma^2_{\epsilon_{1}}} &amp; ... &amp; Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}}\\
 \vdots &amp; \ddots &amp; \vdots\\
Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}} &amp; ... &amp; Z_t K{\sigma^2_{g_{t}}} Z_t' + H{\sigma^2_{\epsilon_{t}}} \\
\end{array}\right]
</annotation></semantics></math></p>
<p><br></p>
<p>where K is the relationship or covariance matrix for the kth random
effect (u=1,‚Ä¶,k), and R=I is an identity matrix for the residual term.
The terms
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><msub><mi>g</mi><mi>i</mi></msub><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{g_{i}}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><msub><mi>œµ</mi><mi>i</mi></msub><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{\epsilon_{i}}</annotation></semantics></math>
denote the genetic (or any of the kth random terms) and residual
variance of trait <code>i</code>, respectively and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÉ</mi><msub><mi>g</mi><msub><mi></mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></msub></msub><annotation encoding="application/x-tex">\sigma_{g_{_{ij}}}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÉ</mi><msub><mi>œµ</mi><msub><mi></mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></msub></msub><annotation encoding="application/x-tex">\sigma_{\epsilon_{_{ij}}}</annotation></semantics></math>
the genetic (or any of the kth random terms) and residual covariance
between traits <code>i</code> and <code>j</code> (i=1,‚Ä¶,t, and j=1,‚Ä¶,t).
The algorithm implemented optimizes the log likelihood:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><mi>g</mi><mi>L</mi><mo>=</mo><mn>1</mn><mi>/</mi><mn>2</mn><mo>*</mo><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">|</mo><mi>V</mi><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>l</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>X</mi><mi>‚Ä≤</mi><mrow><mo stretchy="true" form="prefix">|</mo><mi>V</mi><mo stretchy="true" form="postfix">|</mo></mrow><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>Y</mi><mi>‚Ä≤</mi><mi>P</mi><mi>Y</mi></mrow><annotation encoding="application/x-tex">logL = 1/2 * ln(|V|) + ln(X'|V|X) + Y'PY</annotation></semantics></math></p>
<p><br></p>
<p>where || is the determinant of a matrix. The REML estimates are
updated using a Newton optimization algorithm of the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>Œ∏</mi><mrow><mi>k</mi><mo>+</mo><mn>1</mn></mrow></msup><mo>=</mo><msup><mi>Œ∏</mi><mi>k</mi></msup><mo>+</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>H</mi><mi>k</mi></msup><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><mo>*</mo><mfrac><mrow><mi>d</mi><mi>L</mi></mrow><mrow><mi>d</mi><msubsup><mi>œÉ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></mfrac><mo stretchy="false" form="prefix">|</mo><msup><mi>Œ∏</mi><mi>k</mi></msup></mrow><annotation encoding="application/x-tex">\theta^{k+1} = \theta^{k} + (H^{k})^{-1}*\frac{dL}{d\sigma^2_i}|\theta^k</annotation></semantics></math></p>
<p><br></p>
<p>Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∏</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>
is the vector of variance components for random effects and covariance
components among traits,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>H</mi><mrow><mi>‚àí</mi><mn>1</mn></mrow></msup><annotation encoding="application/x-tex">H^{-1}</annotation></semantics></math>
is the inverse of the Hessian matrix of second derivatives for the kth
cycle,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><mi>d</mi><mi>L</mi></mrow><mrow><mi>d</mi><msubsup><mi>œÉ</mi><mi>i</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{dL}{d\sigma^2_i}</annotation></semantics></math>
is the vector of first derivatives of the likelihood with respect to the
variance-covariance components. The Eigen decomposition of the
relationship matrix proposed by Lee and Van Der Werf (2016) was included
in the Newton-Raphson algorithm to improve time efficiency.
Additionally, the popular <code><a href="../reference/vpredict.html">vpredict()</a></code> function to estimate
standard errors for linear combinations of variance components
(i.e.¬†heritabilities and genetic correlations) was added to the package
as well.</p>
<p>Please refer to the canonical papers listed in the Literature section
to check how the algorithms work. We have tested widely the methods to
make sure they provide the same solution when the likelihood behaves
well, but for complex problems they might lead to slightly different
answers. If you have any concern please contact me at <a href="mailto:cova_ruber@live.com.mx" class="email">cova_ruber@live.com.mx</a>.</p>
<p>In the following section we will go in detail over several examples
on how to use mixed models in univariate and multivariate case and their
use in quantitative genetics.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="section-2-topics-in-quantitative-genetics">SECTION 2: Topics in quantitative genetics<a class="anchor" aria-label="anchor" href="#section-2-topics-in-quantitative-genetics"></a>
</h2>
<div class="section level3">
<h3 id="marker-and-non-marker-based-heritability-calculation">1) Marker and non-marker based heritability calculation<a class="anchor" aria-label="anchor" href="#marker-and-non-marker-based-heritability-calculation"></a>
</h3>
<p>Heritability is one of the most popular parameters among the breeding
and genetics communities because of the insight it provides in the
inheritance of the trait and potential selection response. Heritability
is usually estimated as narrow sense
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>;
only additive variance in the numerator
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>),
and broad sense
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>H</mi><mn>2</mn></msup><annotation encoding="application/x-tex">H^2</annotation></semantics></math>;
all genetic variance in the numerator
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>G</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_G</annotation></semantics></math>).</p>
<p>In a classical breeding experiment with no molecular markers, special
designs are performed to estimate and dissect the additive
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>)
and non-additive (e.g., dominance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>D</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_D</annotation></semantics></math>,
and epistatic
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>E</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_E</annotation></semantics></math>)
variance along with environmental variability. Designs such as
generation analysis, North Carolina designs are used to dissect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>D</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_D</annotation></semantics></math>
to estimate the narrow sense heritability
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>)
using only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>
in the numerator. When no special design is available we can still
disect the genetic variance
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>G</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_G</annotation></semantics></math>)
and estimate the broad sense heritability. In this first example we will
show the broad sense estimation which doesn‚Äôt use covariance matrices
for the genotypic effect (e.g., genomic-additive relationship matrices).
For big models with no relationship matrices, sommer‚Äôs direct inversion
is a bad idea to use but we will still show how to do it, but keep in
mind that for very sparse models with no relationship matrices or other
special covariance structures we recommend using the <code>lmer()</code>
function from the lme4 package or any other package using MME-based
algorithms (e.g., asreml-R).</p>
<p>The following dataset has 41 potato lines evaluated in 5 locations
across 3 years in an RCBD design. We show how to fit the model and
extract the variance components to calculate the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/covaruber/sommer">sommer</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Matrix</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MASS</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: crayon</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="va">A_example</span></span>
<span></span>
<span><span class="va">ans1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span> <span class="op">~</span> <span class="va">Name</span> <span class="op">+</span> <span class="va">Env</span> <span class="op">+</span> <span class="va">Env</span><span class="op">:</span><span class="va">Name</span> <span class="op">+</span> <span class="va">Env</span><span class="op">:</span><span class="va">Block</span>,</span>
<span>             rcov<span class="op">=</span> <span class="op">~</span> <span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans1</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                        VarComp    VarCompSE       Zratio Constraint</span></span>
<span><span class="co">## Name:isc:isc      5.922491e+00 4.033870e-16 1.468191e+16   Positive</span></span>
<span><span class="co">## Env:isc:isc       1.115162e+02 1.001923e-14 1.113021e+16   Positive</span></span>
<span><span class="co">## Env:Name:isc:isc  1.973759e+01 1.155087e-16 1.708752e+17   Positive</span></span>
<span><span class="co">## Env:Block:isc:isc 1.000000e-10 1.692545e-04 5.908264e-07      Fixed</span></span>
<span><span class="co">## units:isc:isc     1.000000e-10 1.098857e-05 9.100366e-06   Positive</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">n.env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">Env</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vpredict(ans1, h2 ~ V1 / ( V1 + (V3/n.env) + (V5/(2*n.env)) ) )</span></span></code></pre></div>
<p>That is an estimate of broad-sense heritability.</p>
<p>Recently with markers becoming cheaper, thousand of markers can be
run in the breeding materials. When markers are available, a special
design is not neccesary to dissect the additive genetic variance. The
availability of the additive, dominance and epistatic relationship
matrices allow us to estimate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>D</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_D</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>I</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_I</annotation></semantics></math>,
although given that A, D and E are not orthogonal the interpretation of
models that fit more than the A matrix at the same time becomes
cumbersome.</p>
<p>Assume you have a population (even unreplicated) in the field but in
addition we have genetic markers. Now we can fit the model and estimate
the genomic heritability that explains a portion of the additive genetic
variance (with high marker density
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>
=
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mrow><mi>m</mi><mi>a</mi><mi>r</mi><mi>k</mi><mi>e</mi><mi>r</mi><mi>s</mi></mrow><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{markers}</annotation></semantics></math>)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">idd</span> <span class="op">&lt;-</span><span class="va">DT</span><span class="op">$</span><span class="va">id</span>; <span class="va">DT</span><span class="op">$</span><span class="va">ide</span> <span class="op">&lt;-</span><span class="va">DT</span><span class="op">$</span><span class="va">id</span></span>
<span><span class="co">### look at the data</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/D.mat.html">D.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># dominance relationship matrix</span></span>
<span><span class="va">E</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/E.mat.html">E.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># epistatic relationship matrix</span></span>
<span><span class="va">ans.ADE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>                 random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>,Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">idd</span>,Gu<span class="op">=</span><span class="va">D</span><span class="op">)</span>, </span>
<span>                 rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>                 data<span class="op">=</span><span class="va">DT</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans.ADE</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       VarComp    VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## u:id.color-color  0.003486639 0.0011076709 3.147721   Positive</span></span>
<span><span class="co">## u:idd.color-color 0.001286778 0.0005071774 2.537136   Positive</span></span>
<span><span class="co">## units.color-color 0.002152054 0.0002976821 7.229372   Positive</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vpredict.html">vpredict</a></span><span class="op">(</span><span class="va">ans.ADE</span>, <span class="va">h2</span> <span class="op">~</span> <span class="op">(</span><span class="va">V1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span> <span class="va">V1</span><span class="op">+</span><span class="va">V3</span><span class="op">)</span> <span class="op">)</span> <span class="co"># narrow sense</span></span></code></pre></div>
<pre><code><span><span class="co">##     Estimate         SE</span></span>
<span><span class="co">## h2 0.6183417 0.08605268</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vpredict.html">vpredict</a></span><span class="op">(</span><span class="va">ans.ADE</span>, <span class="va">h2</span> <span class="op">~</span> <span class="op">(</span><span class="va">V1</span><span class="op">+</span><span class="va">V2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span> <span class="va">V1</span><span class="op">+</span><span class="va">V2</span><span class="op">+</span><span class="va">V3</span><span class="op">)</span> <span class="op">)</span> <span class="co"># broad-sense</span></span></code></pre></div>
<pre><code><span><span class="co">##     Estimate         SE</span></span>
<span><span class="co">## h2 0.6892552 0.05896317</span></span></code></pre>
<p>In this example we showed how to estimate the additive
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>A</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_A</annotation></semantics></math>)
and dominance
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>œÉ</mi><mi>D</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_D</annotation></semantics></math>)
variance components based on markers and estimate broad
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>H</mi><mn>2</mn></msup><annotation encoding="application/x-tex">H^2</annotation></semantics></math>)
and narrow-sense heritability
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>).
Notice that we used the <code><a href="../reference/vsr.html">vsr()</a></code> function which indicates that
the random effect inside the parenthesis (i.e.¬†<code>id</code>,
<code>idd</code> or <code>ide</code>) has a covariance matrix (A, D, or
E), that will be specified in the <code>Gu</code> argument of the
<code><a href="../reference/vsr.html">vsr()</a></code> function. Please DO NOT provide the inverse, but
rather the original covariance matrix.</p>
</div>
<div class="section level3">
<h3 id="specifying-heterogeneous-variances-in-univariate-models">2) Specifying heterogeneous variances in univariate models<a class="anchor" aria-label="anchor" href="#specifying-heterogeneous-variances-in-univariate-models"></a>
</h3>
<p>Very often in multi-environment trials, the assumption that genetic
variance is the same across locations may be too naive. Because of that,
specifying a general genetic component and a location-specific genetic
variance is the way to go.</p>
<p>We estimate variance components for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>C</mi><msub><mi>A</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">GCA_2</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>C</mi><mi>A</mi></mrow><annotation encoding="application/x-tex">SCA</annotation></semantics></math>
specifying the variance structure.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data(DT_cornhybrids)</span></span>
<span><span class="co"># DT &lt;- DT_cornhybrids</span></span>
<span><span class="co"># DTi &lt;- DTi_cornhybrids</span></span>
<span><span class="co"># GT &lt;- GT_cornhybrids</span></span>
<span><span class="co"># ### fit the model</span></span>
<span><span class="co"># modFD &lt;- mmec(Yield~1, </span></span>
<span><span class="co">#               random=~ vsc(atc(Location,c("3","4")),isc(GCA2)), </span></span>
<span><span class="co">#               rcov= ~ vsc(dsc(Location),isc(units)), nIters=3,</span></span>
<span><span class="co">#               returnParam = F,</span></span>
<span><span class="co">#               data=DT, verbose = FALSE)</span></span>
<span><span class="co"># summary(modFD)</span></span></code></pre></div>
<p>In the previous example we showed how the <code><a href="../reference/atr.html">atr()</a></code> function
is used in the <code><a href="../reference/mmer.html">mmer()</a></code> solver. By using the
<code><a href="../reference/atr.html">atr()</a></code> function you can specify that i.e.¬†the GCA2 has a
different variance in different Locations, in this case locations 3 and
4, but also a main GCA variance. This is considered a CS + DIAG
(compound symmetry + diagonal) model.</p>
<p>In addition, other functions can be added on top to fit models with
covariance structures, i.e.¬†the <code>Gu</code> argument from the
<code><a href="../reference/vsr.html">vsr()</a></code> function to indicate a covariance matrix (A, pedigree
or genomic relationship matrix)</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data(DT_cornhybrids)</span></span>
<span><span class="co"># DT &lt;- DT_cornhybrids</span></span>
<span><span class="co"># DTi &lt;- DTi_cornhybrids</span></span>
<span><span class="co"># GT &lt;- as(GT_cornhybrids, Class = "dgCMatrix")</span></span>
<span><span class="co"># GT[1:4,1:4]</span></span>
<span><span class="co"># DT=DT[with(DT, order(Location)), ]</span></span>
<span><span class="co"># ### fit the model</span></span>
<span><span class="co"># modFD &lt;- mmec(Yield~1, </span></span>
<span><span class="co">#               random=~ vsc(atc(Location,c("3","4")),isc(GCA2),Gu=GT), </span></span>
<span><span class="co">#               rcov= ~ vsc(dsc(Location),isc(units)), nIters=3,</span></span>
<span><span class="co">#               data=DT, verbose = FALSE)</span></span>
<span><span class="co"># summary(modFD)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-the-vpredict-calculator">3) Using the vpredict calculator<a class="anchor" aria-label="anchor" href="#using-the-vpredict-calculator"></a>
</h3>
<p>Sometimes the user needs to calculate ratios or functions of specific
variance-covariance components and obtain the standard errors for such
parameters. Examples of these are the genetic correlations,
heritabilities, etc. Using the CPdata we will show how to estimate the
heritability and the standard error using the <code><a href="../reference/vpredict.html">vpredict()</a></code>
function that uses the delta method to come up with these parameters.
This can be extended for any linear combination of the variance
components.</p>
<div class="section level4">
<h4 id="standar-error-for-heritability">3.1) Standar error for heritability<a class="anchor" aria-label="anchor" href="#standar-error-for-heritability"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">### look at the data</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="va">ans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>                random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>,Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span>, </span>
<span>                rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>                data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans.ADE</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       VarComp    VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## u:id.color-color  0.003486639 0.0011076709 3.147721   Positive</span></span>
<span><span class="co">## u:idd.color-color 0.001286778 0.0005071774 2.537136   Positive</span></span>
<span><span class="co">## units.color-color 0.002152054 0.0002976821 7.229372   Positive</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vpredict.html">vpredict</a></span><span class="op">(</span><span class="va">ans</span>, <span class="va">h2</span> <span class="op">~</span> <span class="op">(</span><span class="va">V1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span> <span class="va">V1</span><span class="op">+</span><span class="va">V2</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Estimate         SE</span></span>
<span><span class="co">## h2 0.6405467 0.05805209</span></span></code></pre>
<p>The same can be used for multivariate models. Please check the
documentation of the <code>vpredict</code> function to see more
examples.</p>
</div>
<div class="section level4">
<h4 id="standar-error-for-genetic-correlation">3.2) Standar error for genetic correlation<a class="anchor" aria-label="anchor" href="#standar-error-for-genetic-correlation"></a>
</h4>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## just silenced to avoid too much time building the vignettes</span></span>
<span><span class="co"># data(DT_btdata)</span></span>
<span><span class="co"># DT &lt;- DT_btdata</span></span>
<span><span class="co"># mix3 &lt;- mmer(cbind(tarsus, back) ~ sex,</span></span>
<span><span class="co">#                random = ~ vsr(dam, Gtc=unsm(2)) + vsr(fosternest,Gtc=diag(2)),</span></span>
<span><span class="co">#                rcov=~vsr(units,Gtc=unsm(2)), nIters=3,</span></span>
<span><span class="co">#                data = DT, verbose = FALSE)</span></span>
<span><span class="co"># summary(mix3)</span></span>
<span><span class="co"># #### calculate the genetic correlation</span></span>
<span><span class="co"># vpredict(mix3, gen.cor ~ V2 / sqrt(V1*V3))</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="half-and-full-diallel-designs-use-of-the-overlay">4) Half and full diallel designs (use of the overlay)<a class="anchor" aria-label="anchor" href="#half-and-full-diallel-designs-use-of-the-overlay"></a>
</h3>
<p>When breeders are looking for the best single-cross combinations,
diallel designs have been by far the most used design in crops like
maize. There are 4 types of diallel designs depending on whether
reciprocal and self-crosses (omission of parents) are performed (full
diallel with parents n^2; full diallel without parents n(n-1); half
diallel with parents 1/2 * n(n+1); half diallel without parents 1/2 *
n(n-1) ). In this example we will show a full diallel design (reciprocal
crosses are performed) and half diallel designs (only one of the
directions is performed).</p>
<p>In the first data set we show a full diallel among 40 lines from 2
heterotic groups, 20 in each. Therefore 400 possible hybrids are
possible. We have pehnotypic data for 100 of them across 4 locations. We
use the data available to fit a model of the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mn>1</mn></msub><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mn>2</mn></msub><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mi>S</mi></msub><mo>+</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">y = X\beta + Zu_1 + Zu_2 + Zu_S + \epsilon</annotation></semantics></math></p>
<p><br></p>
<p>We estimate variance components for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>C</mi><msub><mi>A</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">GCA_1</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>C</mi><msub><mi>A</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">GCA_2</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>C</mi><mi>A</mi></mrow><annotation encoding="application/x-tex">SCA</annotation></semantics></math>
and use them to estimate heritability. Additionally BLUPs for GCA and
SCA effects can be used to predict crosses.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cornhybrids</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cornhybrids</span></span>
<span><span class="va">DTi</span> <span class="op">&lt;-</span> <span class="va">DTi_cornhybrids</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cornhybrids</span></span>
<span></span>
<span><span class="va">modFD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Location</span>, </span>
<span>              random<span class="op">=</span><span class="op">~</span><span class="va">GCA1</span><span class="op">+</span><span class="va">GCA2</span><span class="op">+</span><span class="va">SCA</span>, </span>
<span>              rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Updated VC is not positive definite, changing to EM step</span></span>
<span><span class="co">## Update using constraints</span></span>
<span><span class="co">## Updated VC is not positive definite, changing to EM step</span></span>
<span><span class="co">## Update using constraints</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">suma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">modFD</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    VarComp VarCompSE       Zratio Constraint</span></span>
<span><span class="co">## GCA1:isc:isc  1.000000e-10  36.89059 2.710718e-12      Fixed</span></span>
<span><span class="co">## GCA2:isc:isc  4.629216e+01  37.14297 1.246324e+00   Positive</span></span>
<span><span class="co">## SCA:isc:isc   8.937350e+01  21.56529 4.144322e+00   Positive</span></span>
<span><span class="co">## units:isc:isc 2.388020e+02  28.83389 8.281992e+00   Positive</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Vgca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">suma</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">Vsca</span> <span class="op">&lt;-</span> <span class="va">suma</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">Ve</span> <span class="op">&lt;-</span> <span class="va">suma</span><span class="op">[</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">Va</span> <span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="va">Vgca</span></span>
<span><span class="va">Vd</span> <span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="va">Vsca</span></span>
<span><span class="va">Vg</span> <span class="op">&lt;-</span> <span class="va">Va</span> <span class="op">+</span> <span class="va">Vd</span></span>
<span><span class="op">(</span><span class="va">H2</span> <span class="op">&lt;-</span> <span class="va">Vg</span> <span class="op">/</span> <span class="op">(</span><span class="va">Vg</span> <span class="op">+</span> <span class="op">(</span><span class="va">Ve</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.6944174</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">Va</span> <span class="op">/</span> <span class="op">(</span><span class="va">Vg</span> <span class="op">+</span> <span class="op">(</span><span class="va">Ve</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2369507</span></span></code></pre>
<p>Don‚Äôt worry too much about the <code>h2</code> value, the data was
simulated to be mainly dominance variance, therefore the <code>Va</code>
was simulated extremely small leading to such value of narrow sense
<code>h2</code>.</p>
<p>In the second data set we show a small half diallel with 7 parents
crossed in one direction. There are n(n-1)/2 possible crosses; 7(6)/2 =
21 unique crosses. Parents appear as males or females indistictly. Each
with two replications in a CRD. For a half diallel design a single GCA
variance component for both males and females can be estimated and an
SCA as well
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>œÉ</mi><mi>G</mi><mn>2</mn></msubsup><mi>C</mi><mi>A</mi></mrow><annotation encoding="application/x-tex">\sigma^2_GCA</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>œÉ</mi><mi>S</mi><mn>2</mn></msubsup><mi>C</mi><mi>A</mi></mrow><annotation encoding="application/x-tex">\sigma^2_SCA</annotation></semantics></math>
respectively), and BLUPs for GCA and SCA of the parents can be
extracted. We will show first how to do so with the <code><a href="../reference/mmer.html">mmer()</a></code>
function using the <code><a href="../reference/overlay.html">overlay()</a></code> function. The specific model
here is:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mi>g</mi></msub><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mi>s</mi></msub><mo>+</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">y = X\beta + Zu_g + Zu_s + \epsilon</annotation></semantics></math></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"DT_halfdiallel"</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_halfdiallel</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   rep geno male female     sugar</span></span>
<span><span class="co">## 1   1   12    1      2 13.950509</span></span>
<span><span class="co">## 2   2   12    1      2  9.756918</span></span>
<span><span class="co">## 3   1   13    1      3 13.906355</span></span>
<span><span class="co">## 4   2   13    1      3  9.119455</span></span>
<span><span class="co">## 5   1   14    1      4  5.174483</span></span>
<span><span class="co">## 6   2   14    1      4  8.452221</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span><span class="op">$</span><span class="va">femalef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">female</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">malef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">male</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">genof</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="co">#### model using overlay</span></span>
<span><span class="va">modh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">sugar</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="fu"><a href="../reference/overlay.html">overlay</a></span><span class="op">(</span><span class="va">femalef</span>,<span class="va">malef</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span>             <span class="op">+</span> <span class="va">genof</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">modh</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                            VarComp    VarCompSE       Zratio Constraint</span></span>
<span><span class="co">## femalef:malef:isc:isc 5.780585e+00 8.281722e-17 6.979932e+16   Positive</span></span>
<span><span class="co">## genof:isc:isc         2.133867e+01 7.516095e-17 2.839064e+17   Positive</span></span>
<span><span class="co">## units:isc:isc         1.000000e-10 1.697717e-05 5.890264e-06   Positive</span></span></code></pre>
<p>Notice how the <code><a href="../reference/overlay.html">overlay()</a></code> argument makes the overlap of
incidence matrices possible making sure that male and female are joint
into a single random effect.</p>
</div>
<div class="section level3">
<h3 id="genomic-selection-predicting-mendelian-sampling">5) Genomic selection: predicting mendelian sampling<a class="anchor" aria-label="anchor" href="#genomic-selection-predicting-mendelian-sampling"></a>
</h3>
<p>In this section we will use wheat data from CIMMYT to show how
genomic selection is performed. This is the case of prediction of
specific individuals within a population. It basically uses a similar
model of the form:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><mi>u</mi><mo>+</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">y = X\beta + Zu + \epsilon</annotation></semantics></math></p>
<p><br></p>
<p>and takes advantage of the variance covariance matrix for the
genotype effect known as the additive relationship matrix (A) and
calculated using the <code>A.mat</code> function to establish
connections among all individuals and predict the BLUPs for individuals
that were not measured. The prediction accuracy depends on several
factors such as the heritability
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>h</mi><mn>2</mn></msup><annotation encoding="application/x-tex">h^2</annotation></semantics></math>),
training population used (TP), size of TP, etc.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_wheat</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_wheat</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_wheat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">200</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"X"</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span>;<span class="va">DT</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># select environment 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="co"># GBLUP pedigree-based approach</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">y.trn</span> <span class="op">&lt;-</span> <span class="va">DT</span></span>
<span><span class="va">vv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y.trn</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">y.trn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              X1          X2          X3         X4   id</span></span>
<span><span class="co">## 775   1.6716295 -1.72746986 -1.89028479  0.0509159  775</span></span>
<span><span class="co">## 2166 -0.2527028  0.40952243  0.30938553 -1.7387588 2166</span></span>
<span><span class="co">## 2167         NA -0.64862633 -0.79955921 -1.0535691 2167</span></span>
<span><span class="co">## 2465  0.7854395  0.09394919  0.57046773  0.5517574 2465</span></span>
<span><span class="co">## 3881  0.9983176 -0.28248062  1.61868192 -0.1142848 3881</span></span>
<span><span class="co">## 3889  2.3360969  0.62647587  0.07353311  0.7195856 3889</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## GBLUP</span></span>
<span><span class="va">ans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">X1</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>,Gu<span class="op">=</span><span class="va">K</span><span class="op">)</span>, </span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>,nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">y.trn</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># kinship based</span></span>
<span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"id"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">ans</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">X1</span><span class="op">[</span><span class="va">vv</span>,<span class="op">]</span>,<span class="va">DT</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span>, use<span class="op">=</span><span class="st">"complete"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4310372</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rrBLUP</span></span>
<span><span class="va">ans2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">X1</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span>, buildGu <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, </span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, getPEV <span class="op">=</span> <span class="cn">FALSE</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             data<span class="op">=</span><span class="va">y.trn</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># kinship based</span></span>
<span></span>
<span><span class="va">u</span> <span class="op">&lt;-</span> <span class="va">GT</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">ans2</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:GT`</span><span class="op">$</span><span class="va">X1</span><span class="op">)</span> <span class="co"># BLUPs for individuals</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">u</span><span class="op">[</span><span class="va">vv</span>,<span class="op">]</span>,<span class="va">DT</span><span class="op">[</span><span class="va">vv</span>,<span class="st">"X1"</span><span class="op">]</span><span class="op">)</span> <span class="co"># same correlation</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4370181</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the same can be applied in multi-response models in GBLUP or rrBLUP</span></span></code></pre></div>
<p>Please notice that when specifying the marker matrix as a random
effect we used the argument ‚ÄòbuildGu=FALSE‚Äô to inform the ‚Äòmmer‚Äô
function that a covariance matrix for the levels of the random effect
shouldn‚Äôt be built. Imagine a model with 100,000 markers, that would
imply a relationship matrix of 100,000 x 100,000. If that matrix is a
diagonal it would only compromise the speed and memory of the function.
By setting ‚ÄòbuildGu=FALSE‚Äô the mmer solver will avoid the matrix
multiplications using that huge diagonal matrix. If you want to specify
a relationship matrix for the marker matrix then you cannot use that
‚ÄòbuildGu‚Äô argument.</p>
</div>
<div class="section level3">
<h3 id="indirect-genetic-effects">6) Indirect genetic effects<a class="anchor" aria-label="anchor" href="#indirect-genetic-effects"></a>
</h3>
<p>General variance structures can be used to fit indirect genetic
effects. Here, we use an example dataset to show how we can fit the
variance and covariance components between two or more different random
effects.</p>
<p>We first fit a direct genetic effects model:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data(DT_ige)</span></span>
<span><span class="co"># DT &lt;- DT_ige</span></span>
<span><span class="co"># Af &lt;- A_ige</span></span>
<span><span class="co"># An &lt;- A_ige</span></span>
<span></span>
<span><span class="co">## Direct genetic effects model</span></span>
<span><span class="co"># modDGE &lt;- mmec(trait ~ block,</span></span>
<span><span class="co">#                random = ~ focal,</span></span>
<span><span class="co">#                rcov = ~ units, nIters=3,</span></span>
<span><span class="co">#                data = DT, verbose=FALSE)</span></span>
<span><span class="co"># summary(modDGE)$varcomp</span></span></code></pre></div>
<p>We now fit the indirect genetic effects model without covariance
between DGE and IGE:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data(DT_ige)</span></span>
<span><span class="co"># DT &lt;- DT_ige</span></span>
<span><span class="co"># A &lt;- A_ige</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># ## Indirect genetic effects model</span></span>
<span><span class="co"># modIGE &lt;- mmec(trait ~ block, dateWarning = FALSE,</span></span>
<span><span class="co">#                random = ~ focal + neighbour, verbose = FALSE,</span></span>
<span><span class="co">#                rcov = ~ units, nIters=100,</span></span>
<span><span class="co">#               data = DT)</span></span>
<span><span class="co"># summary(modIGE)$varcomp</span></span></code></pre></div>
<p>We now fit the indirect genetic effects model with covariance between
DGE and IGE for which we will use the <code><a href="../reference/gvsr.html">gvsr()</a></code> function:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ### Indirect genetic effects model</span></span>
<span><span class="co"># modIGE &lt;- mmec(trait ~ block, dateWarning = FALSE,</span></span>
<span><span class="co">#                random = ~ covc( vsc(isc(focal)), vsc(isc(neighbour)) ),</span></span>
<span><span class="co">#                rcov = ~ units, nIters=100, verbose = FALSE,</span></span>
<span><span class="co">#               data = DT)</span></span>
<span><span class="co"># summary(modIGE)$varcomp</span></span></code></pre></div>
<p>On top of that we can include a relationship matrix for the two
random effects that are being forced to co-vary</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Indirect genetic effects model</span></span>
<span><span class="co"># Ai &lt;- as( solve(A_ige + diag(1e-5, nrow(A_ige),nrow(A_ige) )), Class="dgCMatrix")</span></span>
<span><span class="co"># # Indirect genetic effects model with covariance between DGE and IGE using relationship matrices</span></span>
<span><span class="co"># modIGE &lt;- mmec(trait ~ block, dateWarning = FALSE,</span></span>
<span><span class="co">#                random = ~ covc( vsc(isc(focal), Gu=Ai), vsc(isc(neighbour), Gu=Ai) ),</span></span>
<span><span class="co">#                rcov = ~ units, nIters=100, verbose = FALSE,</span></span>
<span><span class="co">#               data = DT)</span></span>
<span><span class="co"># summary(modIGE)$varcomp</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genomic-selection-single-cross-prediction">7) Genomic selection: single cross prediction<a class="anchor" aria-label="anchor" href="#genomic-selection-single-cross-prediction"></a>
</h3>
<p>When doing prediction of single cross performance the phenotype can
be dissected in three main components, the general combining abilities
(GCA) and specific combining abilities (SCA). This can be expressed with
the same model analyzed in the diallel experiment mentioned before:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>X</mi><mi>Œ≤</mi><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mn>1</mn></msub><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mn>2</mn></msub><mo>+</mo><mi>Z</mi><msub><mi>u</mi><mi>S</mi></msub><mo>+</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">y = X\beta + Zu_1 + Zu_2 + Zu_S + \epsilon</annotation></semantics></math></p>
<p><br></p>
<p>with:</p>
<p><br></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mn>1</mn></msub><annotation encoding="application/x-tex">u_1</annotation></semantics></math>
~ N(0,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mn>1</mn></msub><annotation encoding="application/x-tex">K_1</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>œÉ</mi><mi>u</mi><mn>2</mn></msubsup><mn>1</mn></mrow><annotation encoding="application/x-tex">\sigma^2_u1</annotation></semantics></math>)</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mn>2</mn></msub><annotation encoding="application/x-tex">u_2</annotation></semantics></math>
~ N(0,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mn>2</mn></msub><annotation encoding="application/x-tex">K_2</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>œÉ</mi><mi>u</mi><mn>2</mn></msubsup><mn>2</mn></mrow><annotation encoding="application/x-tex">\sigma^2_u2</annotation></semantics></math>)</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>u</mi><mi>s</mi></msub><annotation encoding="application/x-tex">u_s</annotation></semantics></math>
~ N(0,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mn>3</mn></msub><annotation encoding="application/x-tex">K_3</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>œÉ</mi><mi>u</mi><mn>2</mn></msubsup><mi>s</mi></mrow><annotation encoding="application/x-tex">\sigma^2_us</annotation></semantics></math>)</p>
<p><br></p>
<p>And we can specify the K matrices. The main difference between this
model and the full and half diallel designs is the fact that this model
will include variance covariance structures in each of the three random
effects (GCA1, GCA2 and SCA) to be able to predict the crosses that have
not ocurred yet. We will use the data published by Technow et al.¬†(2015)
to show how to do prediction of single crosses.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_technow</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_technow</span></span>
<span><span class="va">Md</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">Md_technow</span><span class="op">*</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">Mf</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">Mf_technow</span><span class="op">*</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">Ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">Md</span><span class="op">)</span></span>
<span><span class="va">Af</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">Mf</span><span class="op">)</span></span>
<span><span class="va">Adi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/solve-methods.html" class="external-link">solve</a></span><span class="op">(</span><span class="va">Ad</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">1e-4</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Ad</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Ad</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, Class<span class="op">=</span><span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span><span class="va">Afi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/solve-methods.html" class="external-link">solve</a></span><span class="op">(</span><span class="va">Af</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">1e-4</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Af</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Af</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, Class<span class="op">=</span><span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span><span class="co"># RUN THE PREDICTION MODEL</span></span>
<span><span class="va">y.trn</span> <span class="op">&lt;-</span> <span class="va">DT</span></span>
<span><span class="va">vv1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">GY</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vv2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">vv1</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">y.trn</span><span class="op">[</span><span class="va">vv2</span>,<span class="st">"GY"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">anss2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmec.html">mmec</a></span><span class="op">(</span><span class="va">GY</span><span class="op">~</span><span class="fl">1</span>, </span>
<span>              random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">dent</span><span class="op">)</span>,Gu<span class="op">=</span><span class="va">Adi</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/vsc.html">vsc</a></span><span class="op">(</span><span class="fu"><a href="../reference/isc.html">isc</a></span><span class="op">(</span><span class="va">flint</span><span class="op">)</span>,Gu<span class="op">=</span><span class="va">Afi</span><span class="op">)</span>, </span>
<span>              rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">15</span>,</span>
<span>              data<span class="op">=</span><span class="va">y.trn</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">anss2</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span></span></code></pre></div>
<pre><code><span><span class="co">##                    VarComp VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## dent:Adi:isc:isc  15.86817  2.992973 5.301809   Positive</span></span>
<span><span class="co">## flint:Afi:isc:isc 11.29153  3.280194 3.442337   Positive</span></span>
<span><span class="co">## units:isc:isc     15.97828  2.831564 5.642916   Positive</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># zu1 &lt;- model.matrix(~dent-1,y.trn) %*% anss2$uList$`vsc(isc(dent), Gu = Adi)`</span></span>
<span><span class="co"># zu2 &lt;- model.matrix(~flint-1,y.trn) %*% anss2$uList$`vsc(isc(flint), Gu = Afi)`</span></span>
<span><span class="co"># u &lt;- zu1+zu2+as.vector(anss2$b)</span></span>
<span><span class="co"># cor(u[vv2,], DT$GY[vv2])</span></span></code></pre></div>
<p>In the previous model we only used the GCA effects (GCA1 and GCA2)
for practicity, altough it‚Äôs been shown that the SCA effect doesn‚Äôt
actually help that much in increasing prediction accuracy, but does
increase a lot the computation intensity required since the variance
covariance matrix for SCA is the kronecker product of the variance
covariance matrices for the GCA effects, resulting in a 10578 x 10578
matrix that increases in a very intensive manner the computation
required.</p>
<p>A model without covariance structures would show that the SCA
variance component is insignificant compared to the GCA effects. This is
why including the third random effect doesn‚Äôt increase the prediction
accuracy.</p>
</div>
<div class="section level3">
<h3 id="spatial-modeling-using-the-2-dimensional-spline">8) Spatial modeling: using the 2-dimensional spline<a class="anchor" aria-label="anchor" href="#spatial-modeling-using-the-2-dimensional-spline"></a>
</h3>
<p>We will use the CPdata to show the use of 2-dimensional splines for
accomodating spatial effects in field experiments. In early generation
variety trials the availability of seed is low, which makes the use of
unreplicated designs a neccesity more than anything else. Experimental
designs such as augmented designs and partially-replicated (p-rep)
designs are becoming ever more common these days.</p>
<p>In order to do a good job modeling the spatial trends happening in
the field, special covariance structures have been proposed to
accomodate such spatial trends (i.e.¬†autoregressive residuals; ar1).
Unfortunately, some of these covariance structures make the modeling
rather unstable. More recently, other research groups have proposed the
use of 2-dimensional splines to overcome such issues and have a more
robust modeling of the spatial terms (Lee et al.¬†2013; Rodr√≠guez-√Ålvarez
et al.¬†2018).</p>
<p>In this example we assume an unreplicated population where row and
range information is available which allows us to fit a 2 dimensional
spline model.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">### mimic two fields</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span></span>
<span><span class="va">mix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Rowf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Colf</span><span class="op">)</span> <span class="op">+</span></span>
<span>              <span class="fu"><a href="../reference/spl2Da.html">spl2Da</a></span><span class="op">(</span><span class="va">Row</span>,<span class="va">Col</span><span class="op">)</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">units</span><span class="op">)</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ============================================================</span></span>
<span><span class="co">##          Multivariate Linear Mixed Model fit by REML         </span></span>
<span><span class="co">## **********************  sommer 4.3  ********************** </span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">##          logLik      AIC     BIC Method Converge</span></span>
<span><span class="co">## Value -151.2647 304.5293 308.421     NR    FALSE</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Variance-Covariance components:</span></span>
<span><span class="co">##                     VarComp VarCompSE Zratio Constraint</span></span>
<span><span class="co">## u:id.Yield-Yield      792.6     317.6 2.4959   Positive</span></span>
<span><span class="co">## u:Rowf.Yield-Yield    807.6     371.3 2.1750   Positive</span></span>
<span><span class="co">## u:Colf.Yield-Yield    183.2     139.7 1.3121   Positive</span></span>
<span><span class="co">## A:all.Yield-Yield     515.8     701.3 0.7354   Positive</span></span>
<span><span class="co">## u:units.Yield-Yield  2918.4     292.8 9.9667   Positive</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Fixed effects:</span></span>
<span><span class="co">##   Trait      Effect Estimate Std.Error t.value</span></span>
<span><span class="co">## 1 Yield (Intercept)    132.1     8.761   15.08</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Groups and observations:</span></span>
<span><span class="co">##        Yield</span></span>
<span><span class="co">## u:id     363</span></span>
<span><span class="co">## u:Rowf    13</span></span>
<span><span class="co">## u:Colf    36</span></span>
<span><span class="co">## A:all    168</span></span>
<span><span class="co">## ============================================================</span></span>
<span><span class="co">## Use the '$' sign to access results and parameters</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make a plot to observe the spatial effects found by the spl2D()</span></span>
<span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">DT</span>,<span class="fu"><a href="../reference/spl2Da.html">spl2Da</a></span><span class="op">(</span><span class="va">Row</span>,<span class="va">Col</span><span class="op">)</span><span class="op">)</span> <span class="co"># 2D spline incidence matrix</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="va">W</span><span class="op">$</span><span class="va">Z</span><span class="op">$</span><span class="va">`A:all`</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">mix</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`A:all`</span><span class="op">$</span><span class="va">Yield</span> <span class="co"># 2D spline BLUPs</span></span>
<span><span class="co"># lattice::levelplot(spatial~Row*Col, data=DT) # plot the spatial effect by row and column</span></span></code></pre></div>
<p>Notice that the job is done by the <code><a href="../reference/spl2Da.html">spl2Da()</a></code> function
that takes the <code>Row</code> and <code>Col</code> information to fit
a spatial kernel.</p>
</div>
<div class="section level3">
<h3 id="multivariate-genetic-models-and-genetic-correlations">9) Multivariate genetic models and genetic correlations<a class="anchor" aria-label="anchor" href="#multivariate-genetic-models-and-genetic-correlations"></a>
</h3>
<p>Sometimes is important to estimate genetic variance-covariance among
traits‚Äìmulti-reponse models are very useful for such a task. Let see an
example with 3 traits (<code>color</code>, <code>Yield</code>, and
<code>Firmness</code>) and a single random effect (genotype;
<code>id</code>) although multiple effects can be modeled as well. We
need to use a variance covariance structure for the random effect to be
able to obtain the genetic covariance among traits.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data(DT_cpdata)</span></span>
<span><span class="co"># DT &lt;- DT_cpdata</span></span>
<span><span class="co"># GT &lt;- GT_cpdata</span></span>
<span><span class="co"># MP &lt;- MP_cpdata</span></span>
<span><span class="co"># A &lt;- A.mat(GT)</span></span>
<span><span class="co"># ans.m &lt;- mmer(cbind(Yield,color)~1,</span></span>
<span><span class="co">#                random=~ vsr(id, Gu=A, Gtc=unsm(2))</span></span>
<span><span class="co">#                + vsr(Rowf,Gtc=diag(2))</span></span>
<span><span class="co">#                + vsr(Colf,Gtc=diag(2)),</span></span>
<span><span class="co">#                rcov=~ vsr(units, Gtc=unsm(2)), nIters=3,</span></span>
<span><span class="co">#                data=DT, verbose = FALSE)</span></span></code></pre></div>
<p>Now you can extract the BLUPs using <code>randef(ans.m)</code> or
simply <code>ans.m$U</code>. Also, genetic correlations and
heritabilities can be calculated easily.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cov2cor(ans.m$sigma$`u:id`)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="section-3-special-topics-in-quantitative-genetics">SECTION 3: Special topics in Quantitative genetics<a class="anchor" aria-label="anchor" href="#section-3-special-topics-in-quantitative-genetics"></a>
</h2>
<div class="section level3">
<h3 id="partitioned-model">1) Partitioned model<a class="anchor" aria-label="anchor" href="#partitioned-model"></a>
</h3>
<p>The partitioned model was popularized by () to show that marker
effects can be obtained by fitting a GBLUP model to reduce the
computational burden and then recover them by creating some special
matrices MM‚Äô for GBLUP and M‚Äô(M‚ÄôM)- to recover marker effects. Here we
show a very easy example using the DT_cpdata:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/covaruber/sommer">sommer</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"DT_cpdata"</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span></span>
<span></span>
<span><span class="co">################</span></span>
<span><span class="co"># MARKER MODEL</span></span>
<span><span class="co">################</span></span>
<span><span class="va">mix.marker</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>                   random<span class="op">=</span><span class="op">~</span><span class="va">Rowf</span><span class="op">+</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span>,</span>
<span>                   rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>,data<span class="op">=</span><span class="va">DT</span>, </span>
<span>                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">me.marker</span> <span class="op">&lt;-</span> <span class="va">mix.marker</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:M`</span><span class="op">$</span><span class="va">color</span></span>
<span></span>
<span><span class="co">################</span></span>
<span><span class="co"># PARTITIONED GBLUP MODEL</span></span>
<span><span class="co">################</span></span>
<span></span>
<span><span class="va">MMT</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/matmult-methods.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="co">## MM' = additive relationship matrix </span></span>
<span><span class="va">MMTinv</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/solve-methods.html" class="external-link">solve</a></span><span class="op">(</span><span class="va">MMT</span><span class="op">)</span> <span class="co">## inverse</span></span>
<span><span class="va">MTMMTinv</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">MMTinv</span> <span class="co"># M' %*% (M'M)-</span></span>
<span></span>
<span><span class="va">mix.part</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>                 random<span class="op">=</span><span class="op">~</span><span class="va">Rowf</span><span class="op">+</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">MMT</span><span class="op">)</span>,</span>
<span>                 rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>,data<span class="op">=</span><span class="va">DT</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#convert BLUPs to marker effects me=M'(M'M)- u</span></span>
<span><span class="va">me.part</span><span class="op">&lt;-</span><span class="va">MTMMTinv</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="va">mix.part</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">color</span>,ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compare marker effects between both models</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">me.marker</span>,<span class="va">me.part</span><span class="op">)</span></span></code></pre></div>
<p><img src="v3.sommer.qg_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>As can be seen, these two models are equivalent with the exception
that the partitioned model is more computationally efficient.</p>
</div>
<div class="section level3">
<h3 id="udu-decomposition">2) UDU‚Äô decomposition<a class="anchor" aria-label="anchor" href="#udu-decomposition"></a>
</h3>
<p>Lee and Van der Warf (2015) proposed a decomposition of the
relationship matrix A=UDU‚Äô together with a transformation of the
response and fixed effects Uy = Ux + UZ + e, to fit a model where the
phenotypic variance matrix V is a diagonal because the relationship
matrix is the diagonal matrix D from the decomposition that can be
inverted easily and make multitrait models more feasible.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data("DT_wheat")</span></span>
<span><span class="co"># rownames(GT_wheat) &lt;- rownames(DT_wheat)</span></span>
<span><span class="co"># G &lt;- A.mat(GT_wheat)</span></span>
<span><span class="co"># Y &lt;- data.frame(DT_wheat)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># # make the decomposition</span></span>
<span><span class="co"># UD&lt;-eigen(G) # get the decomposition: G = UDU'</span></span>
<span><span class="co"># U&lt;-UD$vectors</span></span>
<span><span class="co"># D&lt;-diag(UD$values)# This will be our new 'relationship-matrix'</span></span>
<span><span class="co"># rownames(D) &lt;- colnames(D) &lt;- rownames(G)</span></span>
<span><span class="co"># X&lt;-model.matrix(~1, data=Y) # here: only one fixed effect (intercept)</span></span>
<span><span class="co"># UX&lt;-t(U)%*%X # premultiply X and y by U' </span></span>
<span><span class="co"># UY &lt;- t(U) %*% as.matrix(Y) # multivariate</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># # dataset for decomposed model</span></span>
<span><span class="co"># DTd&lt;-data.frame(id = rownames(G) ,UY, UX =UX[,1])</span></span>
<span><span class="co"># DTd$id&lt;-as.character(DTd$id)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># modeld &lt;- mmer(cbind(X1,X2) ~ UX - 1, </span></span>
<span><span class="co">#               random = ~vsr(id,Gu=D), </span></span>
<span><span class="co">#               rcov = ~vsr(units),</span></span>
<span><span class="co">#               data=DTd, verbose = FALSE)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># # dataset for normal model</span></span>
<span><span class="co"># DTn&lt;-data.frame(id = rownames(G) , DT_wheat)</span></span>
<span><span class="co"># DTn$id&lt;-as.character(DTn$id)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># modeln &lt;- mmer(cbind(X1,X2) ~ 1, </span></span>
<span><span class="co">#               random = ~vsr(id,Gu=G), </span></span>
<span><span class="co">#               rcov = ~vsr(units),</span></span>
<span><span class="co">#               data=DTn, verbose = FALSE)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># ## compare regular and transformed blups</span></span>
<span><span class="co"># plot(x=(solve(t(U)))%*%modeld$U$`u:id`$X2[colnames(D)], </span></span>
<span><span class="co">#      y=modeln$U$`u:id`$X2[colnames(D)], xlab="UDU blup",</span></span>
<span><span class="co">#      ylab="blup")</span></span></code></pre></div>
<p>As can be seen, the two models are equivalent. Despite the fact that
sommer doesn‚Äôt take a great advantage of this trick because it was built
for dense matrices using the Armadillo library. Other software may be
better using this trick.</p>
</div>
<div class="section level3">
<h3 id="mating-designs">3) Mating designs<a class="anchor" aria-label="anchor" href="#mating-designs"></a>
</h3>
<p>Estimating variance components has been a topic of interest for the
breeding community for a long time. Here we show how to calculate
additive and dominance variance using the North Carolina Design I
(Nested design) and North Carolina Design II (Factorial design) using
the classical Expected Mean Squares method and the REML methods from
sommer and how these two are equivalent.</p>
<div class="section level4">
<h4 id="north-carolina-design-i-nested-design">North Carolina Design I (Nested design)<a class="anchor" aria-label="anchor" href="#north-carolina-design-i-nested-design"></a>
</h4>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_expdesigns</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_expdesigns</span><span class="op">$</span><span class="va">car1</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span><span class="va">set</span><span class="op">+</span><span class="va">male</span><span class="op">+</span><span class="va">female</span><span class="op">+</span><span class="va">rep</span>, data<span class="op">=</span><span class="va">DT</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">setf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">set</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">repf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">rep</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">malef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">male</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">femalef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">female</span><span class="op">)</span></span>
<span><span class="co">#levelplot(yield~male*female|set, data=DT, main="NC design I")</span></span>
<span><span class="co">##############################</span></span>
<span><span class="co">## Expected Mean Square method</span></span>
<span><span class="co">##############################</span></span>
<span><span class="va">mix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span> <span class="va">setf</span> <span class="op">+</span> <span class="va">setf</span><span class="op">:</span><span class="va">repf</span> <span class="op">+</span> <span class="va">femalef</span><span class="op">:</span><span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">malef</span><span class="op">:</span><span class="va">setf</span>, data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="va">MS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">mix1</span><span class="op">)</span>; <span class="va">MS</span></span></code></pre></div>
<pre><code><span><span class="co">## Analysis of Variance Table</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Response: yield</span></span>
<span><span class="co">##                    Df Sum Sq Mean Sq F value   Pr(&gt;F)    </span></span>
<span><span class="co">## setf                1 0.1780 0.17796  1.6646 0.226012    </span></span>
<span><span class="co">## setf:repf           2 0.9965 0.49824  4.6605 0.037141 *  </span></span>
<span><span class="co">## setf:malef          4 7.3904 1.84759 17.2822 0.000173 ***</span></span>
<span><span class="co">## setf:femalef:malef  6 1.6083 0.26806  2.5074 0.095575 .  </span></span>
<span><span class="co">## Residuals          10 1.0691 0.10691                     </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms1</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"setf:malef"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">ms2</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"setf:femalef:malef"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"Residuals"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">nrep</span><span class="op">=</span><span class="fl">2</span></span>
<span><span class="va">nfem</span><span class="op">=</span><span class="fl">2</span></span>
<span><span class="va">Vfm</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">ms2</span><span class="op">-</span><span class="va">mse</span><span class="op">)</span><span class="op">/</span><span class="va">nrep</span></span>
<span><span class="va">Vm</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">ms1</span><span class="op">-</span><span class="va">ms2</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">nrep</span><span class="op">*</span><span class="va">nfem</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate Va and Vd</span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vm</span> <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Vd</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="op">(</span><span class="va">Vfm</span><span class="op">-</span><span class="va">Vm</span><span class="op">)</span> <span class="co"># assuming no inbreeding(4/(1+F)^2)</span></span>
<span><span class="va">Vg</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Va</span>,<span class="va">Vd</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Vg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Va"</span>,<span class="st">"Vd"</span><span class="op">)</span>; <span class="va">Vg</span></span></code></pre></div>
<pre><code><span><span class="co">##        Va        Vd </span></span>
<span><span class="co">##  1.579537 -1.257241</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##############################</span></span>
<span><span class="co">## REML method</span></span>
<span><span class="co">##############################</span></span>
<span><span class="va">mix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span> <span class="va">setf</span> <span class="op">+</span> <span class="va">setf</span><span class="op">:</span><span class="va">repf</span>,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="va">femalef</span><span class="op">:</span><span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">malef</span><span class="op">:</span><span class="va">setf</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mix2</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span>; <span class="va">vc</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  VarComp  VarCompSE   Zratio Constraint</span></span>
<span><span class="co">## femalef:malef:setf.yield-yield 0.0795986 0.07959526 1.000042   Positive</span></span>
<span><span class="co">## malef:setf.yield-yield         0.3875096 0.27561291 1.405992   Positive</span></span>
<span><span class="co">## units.yield-yield              0.1080557 0.05294578 2.040875   Positive</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Vfm</span> <span class="op">&lt;-</span> <span class="va">vc</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"VarComp"</span><span class="op">]</span></span>
<span><span class="va">Vm</span> <span class="op">&lt;-</span> <span class="va">vc</span><span class="op">[</span><span class="fl">2</span>,<span class="st">"VarComp"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Calculate Va and Vd</span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vm</span> <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Vd</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="op">(</span><span class="va">Vfm</span><span class="op">-</span><span class="va">Vm</span><span class="op">)</span> <span class="co"># assuming no inbreeding(4/(1+F)^2)</span></span>
<span><span class="va">Vg</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Va</span>,<span class="va">Vd</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Vg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Va"</span>,<span class="st">"Vd"</span><span class="op">)</span>; <span class="va">Vg</span></span></code></pre></div>
<pre><code><span><span class="co">##        Va        Vd </span></span>
<span><span class="co">##  1.550038 -1.231644</span></span></code></pre>
<p>As can be seen the REML method is easier than manipulating the MS and
we arrive to the same results.</p>
</div>
<div class="section level4">
<h4 id="north-carolina-design-ii-factorial-design">North Carolina Design II (Factorial design)<a class="anchor" aria-label="anchor" href="#north-carolina-design-ii-factorial-design"></a>
</h4>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_expdesigns</span><span class="op">$</span><span class="va">car2</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span><span class="va">set</span><span class="op">+</span><span class="va">male</span><span class="op">+</span><span class="va">female</span><span class="op">+</span><span class="va">rep</span>, data<span class="op">=</span><span class="va">DT</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">setf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">set</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">repf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">rep</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">malef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">male</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">$</span><span class="va">femalef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">female</span><span class="op">)</span></span>
<span><span class="co">#levelplot(yield~male*female|set, data=DT, main="NC desing II")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   set male female rep   yield setf repf malef femalef</span></span>
<span><span class="co">## 1   1    1      1   1  831.03    1    1     1       1</span></span>
<span><span class="co">## 2   1    2      1   1 1046.55    1    1     2       1</span></span>
<span><span class="co">## 3   1    3      1   1  853.33    1    1     3       1</span></span>
<span><span class="co">## 4   1    4      1   1  940.00    1    1     4       1</span></span>
<span><span class="co">## 5   1    5      1   1  802.00    1    1     5       1</span></span>
<span><span class="co">## 6   1    1      2   1  625.93    1    1     1       2</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">DT</span>,<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">female</span>, <span class="va">male</span>, <span class="va">set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nmale</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">N</span><span class="op">[</span><span class="fl">1</span>,,<span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nfemale</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">N</span><span class="op">[</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nrep</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">N</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">nrep</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">nrep</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">nrep</span><span class="op">)</span> <span class="op">!=</span><span class="fl">0</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################</span></span>
<span><span class="co">## Expected Mean Square method</span></span>
<span><span class="co">##############################</span></span>
<span></span>
<span><span class="va">mix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span> <span class="va">setf</span> <span class="op">+</span> <span class="va">setf</span><span class="op">:</span><span class="va">repf</span> <span class="op">+</span> </span>
<span>             <span class="va">femalef</span><span class="op">:</span><span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">femalef</span><span class="op">:</span><span class="va">setf</span>, data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="va">MS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">mix1</span><span class="op">)</span>; <span class="va">MS</span></span></code></pre></div>
<pre><code><span><span class="co">## Analysis of Variance Table</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Response: yield</span></span>
<span><span class="co">##                    Df  Sum Sq Mean Sq F value    Pr(&gt;F)    </span></span>
<span><span class="co">## setf                1  847836  847836 45.6296 1.097e-09 ***</span></span>
<span><span class="co">## setf:repf           4  144345   36086  1.9421  0.109652    </span></span>
<span><span class="co">## setf:malef          8  861053  107632  5.7926 5.032e-06 ***</span></span>
<span><span class="co">## setf:femalef        8  527023   65878  3.5455  0.001227 ** </span></span>
<span><span class="co">## setf:femalef:malef 32  807267   25227  1.3577  0.129527    </span></span>
<span><span class="co">## Residuals          96 1783762   18581                      </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms1</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"setf:malef"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">ms2</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"setf:femalef"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">ms3</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"setf:femalef:malef"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="va">MS</span><span class="op">[</span><span class="st">"Residuals"</span>,<span class="st">"Mean Sq"</span><span class="op">]</span></span>
<span><span class="va">nrep</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">rep</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nfem</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">female</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nmal</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">male</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Vfm</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">ms3</span><span class="op">-</span><span class="va">mse</span><span class="op">)</span><span class="op">/</span><span class="va">nrep</span>; </span>
<span><span class="va">Vf</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">ms2</span><span class="op">-</span><span class="va">ms3</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">nrep</span><span class="op">*</span><span class="va">nmale</span><span class="op">)</span>; </span>
<span><span class="va">Vm</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">ms1</span><span class="op">-</span><span class="va">ms3</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">nrep</span><span class="op">*</span><span class="va">nfemale</span><span class="op">)</span>; </span>
<span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vm</span>; <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vf</span>; <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Vd</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="op">(</span><span class="va">Vfm</span><span class="op">)</span>; <span class="co"># assuming no inbreeding(4/(1+F)^2)</span></span>
<span><span class="va">Vg</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Va</span>,<span class="va">Vd</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Vg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Va"</span>,<span class="st">"Vd"</span><span class="op">)</span>; <span class="va">Vg</span></span></code></pre></div>
<pre><code><span><span class="co">##        Va        Vd </span></span>
<span><span class="co">## 10840.192  8861.659</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##############################</span></span>
<span><span class="co">## REML method</span></span>
<span><span class="co">##############################</span></span>
<span></span>
<span><span class="va">mix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">yield</span><span class="op">~</span> <span class="va">setf</span> <span class="op">+</span> <span class="va">setf</span><span class="op">:</span><span class="va">repf</span> ,</span>
<span>            random<span class="op">=</span><span class="op">~</span><span class="va">femalef</span><span class="op">:</span><span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">malef</span><span class="op">:</span><span class="va">setf</span> <span class="op">+</span> <span class="va">femalef</span><span class="op">:</span><span class="va">setf</span>, </span>
<span>            nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>            data<span class="op">=</span><span class="va">DT</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mix2</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span>; <span class="va">vc</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  VarComp VarCompSE    Zratio Constraint</span></span>
<span><span class="co">## femalef:malef:setf.yield-yield  2235.754  2320.298 0.9635634   Positive</span></span>
<span><span class="co">## malef:setf.yield-yield          5464.494  3484.990 1.5680085   Positive</span></span>
<span><span class="co">## femalef:setf.yield-yield        2722.922  2316.795 1.1752969   Positive</span></span>
<span><span class="co">## units.yield-yield              18569.305  2665.240 6.9672171   Positive</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Vfm</span> <span class="op">&lt;-</span> <span class="va">vc</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"VarComp"</span><span class="op">]</span></span>
<span><span class="va">Vm</span> <span class="op">&lt;-</span> <span class="va">vc</span><span class="op">[</span><span class="fl">2</span>,<span class="st">"VarComp"</span><span class="op">]</span></span>
<span><span class="va">Vf</span> <span class="op">&lt;-</span> <span class="va">vc</span><span class="op">[</span><span class="fl">3</span>,<span class="st">"VarComp"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vm</span>; <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Va</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="va">Vf</span>; <span class="co"># assuming no inbreeding (4/(1+F))</span></span>
<span><span class="va">Vd</span><span class="op">=</span><span class="fl">4</span><span class="op">*</span><span class="op">(</span><span class="va">Vfm</span><span class="op">)</span>; <span class="co"># assuming no inbreeding(4/(1+F)^2)</span></span>
<span><span class="va">Vg</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Va</span>,<span class="va">Vd</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Vg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Va"</span>,<span class="st">"Vd"</span><span class="op">)</span>; <span class="va">Vg</span></span></code></pre></div>
<pre><code><span><span class="co">##        Va        Vd </span></span>
<span><span class="co">## 10891.689  8943.017</span></span></code></pre>
<p>As can be seen, the REML method is easier than manipulating the MS
and we arrive to the same results.</p>
</div>
</div>
<div class="section level3">
<h3 id="gwas-by-gblup">4) GWAS by GBLUP<a class="anchor" aria-label="anchor" href="#gwas-by-gblup"></a>
</h3>
<p>Gualdron-Duarte et al.¬†(2014) and Bernal-Rubio et al.¬†(2016) proved
that in (SingleStep)GBLUP or RRBLUP/SNP-BLUP, dividing the estimate of
the marker effect by its standard error is mathematically equivalent to
fixed regression EMMAX GWAS, even if markers are estimated as random
effects in GBLUP and as fixed effects in EMMAX. That way fitting a GBLUP
model is enough to perform GWAS for additive and on-additive
effects.</p>
<p>Let us use the DT_cpdata dataset to explore the GWAS by GBLUP
method</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_cpdata</span><span class="op">)</span></span>
<span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_cpdata</span></span>
<span><span class="va">GT</span> <span class="op">&lt;-</span> <span class="va">GT_cpdata</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">200</span><span class="op">]</span></span>
<span><span class="va">MP</span> <span class="op">&lt;-</span> <span class="va">MP_cpdata</span></span>
<span><span class="co">#### create the variance-covariance matrix</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/A.mat.html">A.mat</a></span><span class="op">(</span><span class="va">GT</span><span class="op">)</span> <span class="co"># additive relationship matrix</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span> <span class="co"># to be used for degrees of freedom</span></span>
<span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># to be used for degrees of freedom (number of levels in fixed effects)</span></span></code></pre></div>
<p>First we fit a regular GWAS/EMMAX using the GWAS function available
in sommer that first calculates variance components and then fits a
regression marker by marker as a fixed effect.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###########################</span></span>
<span><span class="co">#### Regular GWAS/EMMAX approach</span></span>
<span><span class="co">###########################</span></span>
<span><span class="va">mix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span> <span class="op">+</span> <span class="va">Rowf</span> <span class="op">+</span> <span class="va">Colf</span>,</span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, M<span class="op">=</span><span class="va">GT</span>, gTerm <span class="op">=</span> <span class="st">"u:id"</span>,</span>
<span>             verbose <span class="op">=</span> <span class="cn">FALSE</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BB0000;">Performing GWAS evaluation</span></span></span>
<span><span class="co"><span style="color: #BB0000;">## </span></span></span></code></pre>
<p>To compare EMMAX to the approach proposed by Gualdron-Duarte et
al.¬†(2014) and Bernal-Rubio et al.¬†(2016) we will start fitting an
RRBLUP/SNP-BLUP model to show that the estimate of the marker effect by
its standard error is mathematically equivalent to fixed regression
EMMAX GWAS.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###########################</span></span>
<span><span class="co">#### GWAS by RRBLUP approach</span></span>
<span><span class="co">###########################</span></span>
<span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">GT</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">DT</span><span class="op">$</span><span class="va">id</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">mixRRBLUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>              random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span> <span class="op">+</span> <span class="va">Rowf</span> <span class="op">+</span> <span class="va">Colf</span>,</span>
<span>              rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>              verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span>
<span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="va">mixRRBLUP</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:Z`</span><span class="op">$</span><span class="va">color</span> <span class="co"># marker effects</span></span>
<span><span class="va">se.a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu">kronecker</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span>,<span class="va">mixRRBLUP</span><span class="op">$</span><span class="va">sigma</span><span class="op">$</span><span class="va">`u:Z`</span><span class="op">)</span> <span class="op">-</span> <span class="va">mixRRBLUP</span><span class="op">$</span><span class="va">PevU</span><span class="op">$</span><span class="va">`u:Z`</span><span class="op">$</span><span class="va">color</span><span class="op">)</span><span class="op">)</span> <span class="co"># SE of marker effects</span></span>
<span><span class="va">t.stat</span> <span class="op">&lt;-</span> <span class="va">a</span><span class="op">/</span><span class="va">se.a</span> <span class="co"># t-statistic</span></span>
<span><span class="va">pvalRRBLUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/TDist.html" class="external-link">dt</a></span><span class="op">(</span><span class="va">t.stat</span>,df<span class="op">=</span><span class="va">n</span><span class="op">-</span><span class="va">k</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="co"># -log10(pval)</span></span></code></pre></div>
<p>Instead of fitting the RRBLUP/SNP-BLUP model we can fit a GBLUP model
which is less computationally demanding and recover marker effects and
their standard errors from the genotype effects.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###########################</span></span>
<span><span class="co">#### GWAS by GBLUP approach</span></span>
<span><span class="co">###########################</span></span>
<span><span class="va">M</span><span class="op">&lt;-</span> <span class="va">GT</span></span>
<span><span class="va">MMT</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/matmult-methods.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="co">## MM' = additive relationship matrix</span></span>
<span><span class="va">MMTinv</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/solve-methods.html" class="external-link">solve</a></span><span class="op">(</span><span class="va">MMT</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">1e-6</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">MMT</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">MMT</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co">## inverse of MM'</span></span>
<span><span class="va">MTMMTinv</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">MMTinv</span> <span class="co"># M' %*% (M'M)-</span></span>
<span><span class="va">mixGBLUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mmer.html">mmer</a></span><span class="op">(</span><span class="va">color</span><span class="op">~</span><span class="fl">1</span>,</span>
<span>             random<span class="op">=</span><span class="op">~</span><span class="fu"><a href="../reference/vsr.html">vsr</a></span><span class="op">(</span><span class="va">id</span>, Gu<span class="op">=</span><span class="va">MMT</span><span class="op">)</span> <span class="op">+</span> <span class="va">Rowf</span> <span class="op">+</span> <span class="va">Colf</span>,</span>
<span>             rcov<span class="op">=</span><span class="op">~</span><span class="va">units</span>, nIters<span class="op">=</span><span class="fl">3</span>,</span>
<span>             verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>             data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span>
<span><span class="va">a.from.g</span> <span class="op">&lt;-</span><span class="va">MTMMTinv</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="va">mixGBLUP</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">color</span>,ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">var.g</span> <span class="op">&lt;-</span> <span class="fu">kronecker</span><span class="op">(</span><span class="va">MMT</span>,<span class="va">mixGBLUP</span><span class="op">$</span><span class="va">sigma</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">)</span> <span class="op">-</span> <span class="va">mixGBLUP</span><span class="op">$</span><span class="va">PevU</span><span class="op">$</span><span class="va">`u:id`</span><span class="op">$</span><span class="va">color</span></span>
<span><span class="va">var.a.from.g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">MMTinv</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="op">(</span><span class="va">var.g</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">MMTinv</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">M</span></span>
<span><span class="va">se.a.from.g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">var.a.from.g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">t.stat.from.g</span> <span class="op">&lt;-</span> <span class="va">a.from.g</span><span class="op">/</span><span class="va">se.a.from.g</span> <span class="co"># t-statistic</span></span>
<span><span class="va">pvalGBLUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/TDist.html" class="external-link">dt</a></span><span class="op">(</span><span class="va">t.stat.from.g</span>,df<span class="op">=</span><span class="va">n</span><span class="op">-</span><span class="va">k</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="co"># -log10(pval)</span></span></code></pre></div>
<p>Now we can look at the p-values coming from the 3 approaches to
indeed show that results are equivalent.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###########################</span></span>
<span><span class="co">#### Compare results</span></span>
<span><span class="co">###########################</span></span>
<span><span class="co"># plot(mix2$scores[,1], main="GWAS")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">pvalRRBLUP</span><span class="op">)</span>, main<span class="op">=</span><span class="st">"GWAS by RRBLUP/SNP-BLUP"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="v3.sommer.qg_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">pvalGBLUP</span><span class="op">)</span>, main<span class="op">=</span><span class="st">"GWAS by GBLUP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="v3.sommer.qg_files/figure-html/unnamed-chunk-26-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="final-remarks">Final remarks<a class="anchor" aria-label="anchor" href="#final-remarks"></a>
</h3>
<p>Keep in mind that mmer uses a direct inversion (DI) algorithm which
can be very slow for large datasets with many records. When datasets
have more records than coefficients to be estimated please shift to the
use of the mmec function.</p>
</div>
</div>
<div class="section level2">
<h2 id="literature">Literature<a class="anchor" aria-label="anchor" href="#literature"></a>
</h2>
<p>Covarrubias-Pazaran G. 2016. Genome assisted prediction of
quantitative traits using the R package sommer. PLoS ONE 11(6):1-15.</p>
<p>Covarrubias-Pazaran G. 2018. Software update: Moving the R package
sommer to multivariate mixed models for genome-assisted prediction. doi:
<a href="https://doi.org/10.1101/354639" class="external-link uri">https://doi.org/10.1101/354639</a></p>
<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants.
Second edition. Stemma Press. 390 pp.</p>
<p>Gilmour et al.¬†1995. Average Information REML: An efficient algorithm
for variance parameter estimation in linear mixed models. Biometrics
51(4):1440-1450.</p>
<p>Henderson C.R. 1975. Best Linear Unbiased Estimation and Prediction
under a Selection Model. Biometrics vol.¬†31(2):423-447.</p>
<p>Kang et al.¬†2008. Efficient control of population structure in model
organism association mapping. Genetics 178:1709-1723.</p>
<p>Lee, D.-J., Durban, M., and Eilers, P.H.C. (2013). Efficient
two-dimensional smoothing with P-spline ANOVA mixed models and nested
bases. Computational Statistics and Data Analysis, 61, 22 - 37.</p>
<p>Lee et al.¬†2015. MTG2: An efficient algorithm for multivariate linear
mixed model analysis based on genomic information. Cold Spring Harbor.
doi: <a href="http://dx.doi.org/10.1101/027201" class="external-link uri">http://dx.doi.org/10.1101/027201</a>.</p>
<p>Maier et al.¬†2015. Joint analysis of psychiatric disorders increases
accuracy of risk prediction for schizophrenia, bipolar disorder, and
major depressive disorder. Am J Hum Genet; 96(2):283-294.</p>
<p>Rodriguez-Alvarez, Maria Xose, et al.¬†Correcting for spatial
heterogeneity in plant breeding experiments with P-splines. Spatial
Statistics 23 (2018): 52-71.</p>
<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML
estimates of variance components. Paper invited for the 1993 American
Statistical Association Meeting, San Francisco.</p>
<p>Yu et al.¬†2006. A unified mixed-model method for association mapping
that accounts for multiple levels of relatedness. Genetics
38:203-208.</p>
<p>Tunnicliffe W. 1989. On the use of marginal likelihood in time series
model estimation. JRSS 51(1):15-27.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
