<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Build a hybrid marker matrix using parental genotypes from inbred individuals — build.HMM • sommer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Build a hybrid marker matrix using parental genotypes from inbred individuals — build.HMM"><meta name="description" content="Uses the 2 marker matrices from both sets of inbred or partially inbred parents and creates all possible combinations unless the user specifies which hybrid genotypes to build (custom.hyb argument). It returns the additive and dominance marker matrices (-1,0,1; homo,het,homo in additive and 0,1,0; homo,het,homo for dominance)."><meta property="og:description" content="Uses the 2 marker matrices from both sets of inbred or partially inbred parents and creates all possible combinations unless the user specifies which hybrid genotypes to build (custom.hyb argument). It returns the additive and dominance marker matrices (-1,0,1; homo,het,homo in additive and 0,1,0; homo,het,homo for dominance)."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Build a hybrid marker matrix using parental genotypes from inbred individuals</h1>

      <div class="d-none name"><code>build.HMM.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Uses the 2 marker matrices from both sets of inbred or partially inbred parents and creates all possible combinations unless the user specifies which hybrid genotypes to build (custom.hyb argument). It returns the additive and dominance marker matrices (-1,0,1; homo,het,homo in additive and 0,1,0; homo,het,homo for dominance).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">build.HMM</span><span class="op">(</span><span class="va">M1</span>,<span class="va">M2</span>, custom.hyb<span class="op">=</span><span class="cn">NULL</span>, return.combos.only<span class="op">=</span><span class="cn">FALSE</span>,separator<span class="op">=</span><span class="st">":"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-m-">M1<a class="anchor" aria-label="anchor" href="#arg-m-"></a></dt>
<dd><p>Matrix (\(n \times m\)) of unphased genotypes for \(n\) inbreds and \(m\) biallelic markers,
coded as {-1,0,1}. Fractional (imputed) and missing values (NA) are not allowed.</p></dd>


  <dt id="arg-m-">M2<a class="anchor" aria-label="anchor" href="#arg-m-"></a></dt>
<dd><p>Matrix (\(n \times m\)) of unphased genotypes for \(n\) inbreds and \(m\) biallelic markers,
coded as {-1,0,1}. Fractional (imputed) and missing values (NA) are not allowed.</p></dd>


<dt id="arg-custom-hyb">custom.hyb<a class="anchor" aria-label="anchor" href="#arg-custom-hyb"></a></dt>
<dd><p>A data frame with columns 'Var1' 'Var2', 'hybrid' which specifies which hybrids should be built using the M1 and M2 matrices provided.</p></dd>


<dt id="arg-return-combos-only">return.combos.only<a class="anchor" aria-label="anchor" href="#arg-return-combos-only"></a></dt>
<dd><p>A TRUE/FALSE statement inicating if the function should skip building the geotype matrix for hybrids and only return the data frame with all possible combinations to be build. In case the user wants to subset the hybrids before building the marker matrix.</p></dd>


<dt id="arg-separator">separator<a class="anchor" aria-label="anchor" href="#arg-separator"></a></dt>
<dd><p>Any desired character to be used when pasting the male and female columns to assign the name to the hybrids.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>


<p>It returns the marker matrix for hybrids coded as additive (-1,0,1; homo,het,homo) and dominance (0,1,0; homo,het,homo). This function is deviced for building marker matrices for hybrids coming from inbreds. If the parents are close to inbred &gt;F5 you can try deleting the heterozygote calls (0's) and imputing those cells with the most common genotype (1 or -1). The expectation is that for mostly inbred individuals this may not change drastically the result but will make the results more interpretable. For non-inbred parents (F1 to F3) the cross of an F1 x F1 has many possibilities and is not the intention of this function to build genotypes for heterzygote x heterozygote crosses.</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>It returns the marker matrix for hybrids coded as additive (-1,0,1; homo,het,homo) and dominance (0,1,0; homo,het,homo).</p>
<dl><dt>$HMM.add</dt>
<dd><p>marker matrix for hybrids coded as additive (-1,0,1; homo,het,homo)</p></dd>

<dt>$HMM.dom</dt>
<dd><p>marker matrix for hybrids coded as dominance (0,1,0; homo,het,homo)</p></dd>

<dt>$data.used</dt>
<dd><p>the data frame used to build the hybrid genotypes</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>


<p>Covarrubias-Pazaran G (2016) Genome assisted prediction of quantitative traits using the R package sommer. PLoS ONE 11(6): doi:10.1371/journal.pone.0156744</p>
<p>Nishio M and Satoh M. 2014. Including Dominance Effects in the Genomic BLUP Method for Genomic Evaluation. Plos One 9(1), doi:10.1371/journal.pone.0085792</p>
<p>Su G, Christensen OF, Ostersen T, Henryon M, Lund MS. 2012. Estimating Additive and Non-Additive Genetic Variances and Predicting Genetic Merits Using Genome-Wide Dense Single Nucleotide Polymorphism Markers. PLoS ONE 7(9): e45293. doi:10.1371/journal.pone.0045293</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="mmer.html">mmer</a></code>– the core function of the package</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### use Technow data as example</span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_technow</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_technow</span></span></span>
<span class="r-in"><span><span class="va">Md</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">Md_technow</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="va">Mf</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">Mf_technow</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## first get all possible hybrids</span></span></span>
<span class="r-in"><span><span class="va">res1</span> <span class="op">&lt;-</span> <span class="fu">build.HMM</span><span class="op">(</span><span class="va">Md</span>, <span class="va">Mf</span>, </span></span>
<span class="r-in"><span>                  return.combos.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res1</span><span class="op">$</span><span class="va">data.used</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Var1 Var2  hybrid</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  513  316 513:316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  518  316 518:316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  536  316 536:316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  545  316 545:316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  546  316 546:316</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  548  316 548:316</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## build the marker matrix for the first 50 hybrids</span></span></span>
<span class="r-in"><span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">build.HMM</span><span class="op">(</span><span class="va">Md</span>, <span class="va">Mf</span>,</span></span>
<span class="r-in"><span>                 custom.hyb <span class="op">=</span> <span class="va">res1</span><span class="op">$</span><span class="va">data.used</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span>,<span class="op">]</span></span></span>
<span class="r-in"><span>                 <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Building hybrid marker matrix for 50 hybrids</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Extracting M1 contribution</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Extracting M2 contribution</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done!!</span>
<span class="r-in"><span><span class="va">res2</span><span class="op">$</span><span class="va">HMM.add</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         SYN79 PZE.101000088 PZE.101000108 PZE.101000169 PZE.101000209</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 513:316     1             1             1             0             1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 518:316     1             1             1             0             1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 536:316     1             1             1             0             1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 545:316     1             1             1             0             1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 546:316     1             1             1             0             1</span>
<span class="r-in"><span><span class="va">res2</span><span class="op">$</span><span class="va">HMM.dom</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         SYN79 PZE.101000088 PZE.101000108 PZE.101000169 PZE.101000209</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 513:316     0             0             0             1             0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 518:316     0             0             0             1             0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 536:316     0             0             0             1             0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 545:316     0             0             0             1             0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 546:316     0             0             0             1             0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## now you can use the A.mat(), D.mat() and E.mat() functions</span></span></span>
<span class="r-in"><span><span class="co"># M &lt;- res2$HMM.add</span></span></span>
<span class="r-in"><span><span class="co"># A &lt;- A.mat(M)</span></span></span>
<span class="r-in"><span><span class="co"># D &lt;- D.mat(M)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

