<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>wheat lines dataset — DT_wheat • sommer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="wheat lines dataset — DT_wheat"><meta name="description" content="

Information from a collection of 599 historical CIMMYT wheat lines.  The wheat data set is from
CIMMYT's Global Wheat Program. Historically, this program has conducted numerous international
trials across a wide variety of wheat-producing environments. The environments represented in
these trials were grouped into four basic target sets of environments comprising four
main agroclimatic regions previously defined and widely used by CIMMYT's Global Wheat Breeding Program.
The phenotypic trait considered here was the average grain yield (GY) of the 599 wheat lines evaluated
in each of these four mega-environments.
A pedigree tracing back many generations was available, and the Browse application of
the International Crop Information System (ICIS), as described in  (McLaren et al. 2000, 2005) was used
for deriving the relationship matrix A among the 599 lines; it accounts for selection and inbreeding.
Wheat lines were recently genotyped using 1447 Diversity Array Technology (DArT) generated by
Triticarte Pty. Ltd. (Canberra, Australia; http://www.triticarte.com.au). The DArT markers
may take on two values, denoted by their presence or absence. Markers with a minor allele frequency
lower than 0.05 were removed, and missing genotypes were imputed with samples from the marginal
distribution of marker genotypes, that is, \(x_{ij}=Bernoulli(\hat p_j)\), where  \(\hat p_j\)
is the estimated allele frequency computed from the non-missing genotypes. The number of DArT
MMs after edition was 1279."><meta property="og:description" content="

Information from a collection of 599 historical CIMMYT wheat lines.  The wheat data set is from
CIMMYT's Global Wheat Program. Historically, this program has conducted numerous international
trials across a wide variety of wheat-producing environments. The environments represented in
these trials were grouped into four basic target sets of environments comprising four
main agroclimatic regions previously defined and widely used by CIMMYT's Global Wheat Breeding Program.
The phenotypic trait considered here was the average grain yield (GY) of the 599 wheat lines evaluated
in each of these four mega-environments.
A pedigree tracing back many generations was available, and the Browse application of
the International Crop Information System (ICIS), as described in  (McLaren et al. 2000, 2005) was used
for deriving the relationship matrix A among the 599 lines; it accounts for selection and inbreeding.
Wheat lines were recently genotyped using 1447 Diversity Array Technology (DArT) generated by
Triticarte Pty. Ltd. (Canberra, Australia; http://www.triticarte.com.au). The DArT markers
may take on two values, denoted by their presence or absence. Markers with a minor allele frequency
lower than 0.05 were removed, and missing genotypes were imputed with samples from the marginal
distribution of marker genotypes, that is, \(x_{ij}=Bernoulli(\hat p_j)\), where  \(\hat p_j\)
is the estimated allele frequency computed from the non-missing genotypes. The number of DArT
MMs after edition was 1279."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>wheat lines dataset</h1>

      <div class="d-none name"><code>DT_wheat.Rd</code></div>
    </div>

    <div class="ref-description section level2">


<p>Information from a collection of 599 historical CIMMYT wheat lines.  The wheat data set is from
CIMMYT's Global Wheat Program. Historically, this program has conducted numerous international
trials across a wide variety of wheat-producing environments. The environments represented in
these trials were grouped into four basic target sets of environments comprising four
main agroclimatic regions previously defined and widely used by CIMMYT's Global Wheat Breeding Program.
The phenotypic trait considered here was the average grain yield (GY) of the 599 wheat lines evaluated
in each of these four mega-environments.</p>
<p>A pedigree tracing back many generations was available, and the Browse application of
the International Crop Information System (ICIS), as described in  (McLaren <em>et al.</em> 2000, 2005) was used
for deriving the relationship matrix A among the 599 lines; it accounts for selection and inbreeding.</p>
<p>Wheat lines were recently genotyped using 1447 Diversity Array Technology (DArT) generated by
Triticarte Pty. Ltd. (Canberra, Australia; http://www.triticarte.com.au). The DArT markers
may take on two values, denoted by their presence or absence. Markers with a minor allele frequency
lower than 0.05 were removed, and missing genotypes were imputed with samples from the marginal
distribution of marker genotypes, that is, \(x_{ij}=Bernoulli(\hat p_j)\), where  \(\hat p_j\)
is the estimated allele frequency computed from the non-missing genotypes. The number of DArT
MMs after edition was 1279.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_wheat</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>
    <p>Matrix Y contains the average grain yield, column 1: Grain yield for environment 1 and so on.</p>
    </div>
    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>International Maize and Wheat Improvement Center (CIMMYT), Mexico.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>


<p>Covarrubias-Pazaran G (2016) Genome assisted prediction of quantitative traits using the R package sommer. PLoS ONE 11(6): doi:10.1371/journal.pone.0156744</p>
<p>McLaren, C. G., L. Ramos, C. Lopez, and W. Eusebio. 2000. “Applications of the geneaology manegment system.”
In <em>International Crop Information System. Technical  Development Manual, version VI</em>, edited by McLaren, C. G., J.W. White
and P.N. Fox. pp. 5.8-5.13. CIMMyT, Mexico: CIMMyT and IRRI.</p>
<p>McLaren, C. G., R. Bruskiewich, A.M. Portugal, and A.B. Cosico. 2005. The International Rice Information System.
A platform for meta-analysis of rice crop data. <em>Plant Physiology</em> <b>139</b>: 637-642.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>The core functions of the package <code><a href="mmer.html">mmer</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### For CRAN time limitations most lines in the </span></span></span>
<span class="r-in"><span><span class="co">#### examples are silenced with one '#' mark, </span></span></span>
<span class="r-in"><span><span class="co">#### remove them and run the examples using</span></span></span>
<span class="r-in"><span><span class="co">#### command + shift + C |OR| control + shift + C</span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># data(DT_wheat)</span></span></span>
<span class="r-in"><span><span class="co"># DT &lt;- DT_wheat</span></span></span>
<span class="r-in"><span><span class="co"># GT &lt;- GT_wheat</span></span></span>
<span class="r-in"><span><span class="co"># DTlong &lt;- data.frame(pheno=as.vector(DT), </span></span></span>
<span class="r-in"><span><span class="co">#                      env=sort(rep(1:4,nrow(DT))), </span></span></span>
<span class="r-in"><span><span class="co">#                      id=rep(rownames(DT),4))</span></span></span>
<span class="r-in"><span><span class="co"># DT &lt;- as.data.frame(DT);colnames(DT) &lt;- paste0("x",1:4);DT$line &lt;- rownames(DT);</span></span></span>
<span class="r-in"><span><span class="co"># rownames(GT) &lt;- DT$line</span></span></span>
<span class="r-in"><span><span class="co"># K &lt;- A.mat(GT) # additive relationship matrix</span></span></span>
<span class="r-in"><span><span class="co"># K[1:4,1:4]</span></span></span>
<span class="r-in"><span><span class="co"># ###=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># ###=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># ### using formula based 'mmer'</span></span></span>
<span class="r-in"><span><span class="co"># ###=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># ###=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># head(DT)</span></span></span>
<span class="r-in"><span><span class="co"># #### univariate</span></span></span>
<span class="r-in"><span><span class="co"># mix0 &lt;- mmer(x1~1,</span></span></span>
<span class="r-in"><span><span class="co">#              random = ~vsr(line,Gu=K),</span></span></span>
<span class="r-in"><span><span class="co">#              rcov=~units,</span></span></span>
<span class="r-in"><span><span class="co">#              data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(mix0)$varcomp</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># Ki &lt;- as(solve(K + diag(1e-4,ncol(K),ncol(K))), Class="dgCMatrix")</span></span></span>
<span class="r-in"><span><span class="co"># mix0b &lt;- mmec(x1~1,</span></span></span>
<span class="r-in"><span><span class="co">#              random = ~vsc(isc(line),Gu=Ki),</span></span></span>
<span class="r-in"><span><span class="co">#              rcov=~units,</span></span></span>
<span class="r-in"><span><span class="co">#              data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(mix0b)$varcomp</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

