<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ sommer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="A.mat.html">A.mat()</a></code>

        </dt>
        <dd>Additive relationship matrix</dd>
      </dl><dl><dt>

          <code><a href="AI.html">AI()</a></code>

        </dt>
        <dd>Average Information Algorithm</dd>
      </dl><dl><dt>

          <code><a href="AR1mat.html">AR1()</a></code>

        </dt>
        <dd>Autocorrelation matrix of order 1.</dd>
      </dl><dl><dt>

          <code><a href="ARMAmat.html">ARMA()</a></code>

        </dt>
        <dd>Autocorrelation Moving average.</dd>
      </dl><dl><dt>

          <code><a href="CSmat.html">CS()</a></code>

        </dt>
        <dd>Compound symmetry matrix</dd>
      </dl><dl><dt>

          <code><a href="D.mat.html">D.mat()</a></code>

        </dt>
        <dd>Dominance relationship matrix</dd>
      </dl><dl><dt>

          <code><a href="DT_augment.html">DT_augment</a></code>

        </dt>
        <dd>DT_augment design example.</dd>
      </dl><dl><dt>

          <code><a href="DT_btdata.html">DT_btdata</a></code>

        </dt>
        <dd>Blue Tit Data for a Quantitative Genetic Experiment</dd>
      </dl><dl><dt>

          <code><a href="DT_cornhybrids.html">DT_cornhybrids</a></code>

        </dt>
        <dd>Corn crosses and markers</dd>
      </dl><dl><dt>

          <code><a href="DT_cpdata.html">DT_cpdata</a></code>

        </dt>
        <dd>Genotypic and Phenotypic data for a CP population</dd>
      </dl><dl><dt>

          <code><a href="DT_example.html">DT_example</a></code>

        </dt>
        <dd>Broad sense heritability calculation.</dd>
      </dl><dl><dt>

          <code><a href="DT_expdesigns.html">DT_expdesigns</a></code>

        </dt>
        <dd>Data for different experimental designs</dd>
      </dl><dl><dt>

          <code><a href="DT_fulldiallel.html">DT_fulldiallel</a></code>

        </dt>
        <dd>Full diallel data for corn hybrids</dd>
      </dl><dl><dt>

          <code><a href="DT_gryphon.html">DT_gryphon</a></code>

        </dt>
        <dd>Gryphon data from the Journal of Animal Ecology</dd>
      </dl><dl><dt>

          <code><a href="DT_h2.html">DT_h2</a></code>

        </dt>
        <dd>Broad sense heritability calculation.</dd>
      </dl><dl><dt>

          <code><a href="DT_halfdiallel.html">DT_halfdiallel</a></code>

        </dt>
        <dd>half diallel data for corn hybrids</dd>
      </dl><dl><dt>

          <code><a href="DT_ige.html">DT_ige</a></code>

        </dt>
        <dd>Data to fit indirect genetic effects.</dd>
      </dl><dl><dt>

          <code><a href="DT_legendre.html">DT_legendre</a></code>

        </dt>
        <dd>Simulated data for random regression</dd>
      </dl><dl><dt>

          <code><a href="DT_mohring.html">DT_mohring</a></code>

        </dt>
        <dd>Full diallel data for corn hybrids</dd>
      </dl><dl><dt>

          <code><a href="DT_polyploid.html">DT_polyploid</a></code>

        </dt>
        <dd>Genotypic and Phenotypic data for a potato polyploid population</dd>
      </dl><dl><dt>

          <code><a href="DT_rice.html">DT_rice</a></code>

        </dt>
        <dd>Rice lines dataset</dd>
      </dl><dl><dt>

          <code><a href="DT_sleepstudy.html">DT_sleepstudy</a></code>

        </dt>
        <dd>Reaction times in a sleep deprivation study</dd>
      </dl><dl><dt>

          <code><a href="DT_technow.html">DT_technow</a></code>

        </dt>
        <dd>Genotypic and Phenotypic data from single cross hybrids (Technow et al.,2014)</dd>
      </dl><dl><dt>

          <code><a href="DT_wheat.html">DT_wheat</a></code>

        </dt>
        <dd>wheat lines dataset</dd>
      </dl><dl><dt>

          <code><a href="DT_yatesoats.html">DT_yatesoats</a></code>

        </dt>
        <dd>Yield of oats in a split-block experiment</dd>
      </dl><dl><dt>

          <code><a href="E.mat.html">E.mat()</a></code>

        </dt>
        <dd>Epistatic relationship matrix</dd>
      </dl><dl><dt>

          <code><a href="EM.html">EM()</a></code>

        </dt>
        <dd>Expectation Maximization Algorithm</dd>
      </dl><dl><dt>

          <code><a href="EMMA.html">MEMMA()</a></code>

        </dt>
        <dd>Multivariate Efficient Mixed Model Association Algorithm</dd>
      </dl><dl><dt>

          <code><a href="GWAS.html">GWAS()</a></code>

        </dt>
        <dd>Genome wide association study analysis</dd>
      </dl><dl><dt>

          <code><a href="H.html">H()</a></code>

        </dt>
        <dd>Two-way id by features table</dd>
      </dl><dl><dt>

          <code><a href="H.mat.html">H.mat()</a></code>

        </dt>
        <dd>Combined relationship matrix H</dd>
      </dl><dl><dt>

          <code><a href="LD.decay.html">LD.decay()</a></code>

        </dt>
        <dd>Calculation of linkage disequilibrium decay</dd>
      </dl><dl><dt>

          <code><a href="add.diallel.vars.html">add.diallel.vars()</a></code>

        </dt>
        <dd>add.diallel.vars</dd>
      </dl><dl><dt>

          <code><a href="adiag1.html">adiag1()</a></code>

        </dt>
        <dd>Binds arrays corner-to-corner</dd>
      </dl><dl><dt>

          <code><a href="anova_mmec.html">anova(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>anova form a GLMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="anova_mmer.html">anova(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>anova form a GLMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="atc.html">atc()</a></code>

        </dt>
        <dd>atc covariance structure</dd>
      </dl><dl><dt>

          <code><a href="atcg1234.html">atcg1234()</a></code>

        </dt>
        <dd>Letter to number converter</dd>
      </dl><dl><dt>

          <code><a href="atcg1234BackTransform.html">atcg1234BackTransform()</a></code>

        </dt>
        <dd>Letter to number converter</dd>
      </dl><dl><dt>

          <code><a href="atr.html">atr()</a></code>

        </dt>
        <dd>atr covariance structure</dd>
      </dl><dl><dt>

          <code><a href="bathy.colors.html">bathy.colors()</a></code>

        </dt>
        <dd>Generate a sequence of colors for plotting bathymetric data.</dd>
      </dl><dl><dt>

          <code><a href="bivariateRun.html">bivariateRun()</a></code>

        </dt>
        <dd>bivariateRun functionality</dd>
      </dl><dl><dt>

          <code><a href="build.HMM.html">build.HMM()</a></code>

        </dt>
        <dd>Build a hybrid marker matrix using parental genotypes from inbred individuals</dd>
      </dl><dl><dt>

          <code><a href="coef_mmec.html">coef(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>coef form a GLMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="coef_mmer.html">coef(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>coef form a GLMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="corImputation.html">corImputation()</a></code>

        </dt>
        <dd>Imputing a matrix using correlations</dd>
      </dl><dl><dt>

          <code><a href="covc.html">covc()</a></code>

        </dt>
        <dd>covariance between random effects</dd>
      </dl><dl><dt>

          <code><a href="csc.html">csc()</a></code>

        </dt>
        <dd>customized covariance structure</dd>
      </dl><dl><dt>

          <code><a href="csr.html">csr()</a></code>

        </dt>
        <dd>customized covariance structure</dd>
      </dl><dl><dt>

          <code><a href="dfToMatrix.html">dfToMatrix()</a></code>

        </dt>
        <dd>data frame to matrix</dd>
      </dl><dl><dt>

          <code><a href="dsc.html">dsc()</a></code>

        </dt>
        <dd>diagonal covariance structure</dd>
      </dl><dl><dt>

          <code><a href="dsr.html">dsr()</a></code>

        </dt>
        <dd>diagonal covariance structure</dd>
      </dl><dl><dt>

          <code><a href="fcm.html">fcm()</a></code>

        </dt>
        <dd>fixed effect constraint indication matrix</dd>
      </dl><dl><dt>

          <code><a href="fitted_mmec.html">fitted(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>fitted form a LMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="fitted_mmer.html">fitted(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>fitted form a LMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="fixm.html">fixm()</a></code>

        </dt>
        <dd>fixed indication matrix</dd>
      </dl><dl><dt>

          <code><a href="gvsr.html">gvsr()</a></code>

        </dt>
        <dd>general variance structure specification</dd>
      </dl><dl><dt>

          <code><a href="impute.html">imputev()</a></code>

        </dt>
        <dd>Imputing a numeric or character vector</dd>
      </dl><dl><dt>

          <code><a href="isc.html">isc()</a></code>

        </dt>
        <dd>identity covariance structure</dd>
      </dl><dl><dt>

          <code><a href="jet.colors.html">jet.colors()</a></code>

        </dt>
        <dd>Generate a sequence of colors alog the jet colormap.</dd>
      </dl><dl><dt>

          <code><a href="leg.html">leg()</a></code>

        </dt>
        <dd>Legendre polynomial matrix</dd>
      </dl><dl><dt>

          <code><a href="list2usmat.html">list2usmat()</a></code>

        </dt>
        <dd>list or vector to unstructured matrix</dd>
      </dl><dl><dt>

          <code><a href="logspace.html">logspace()</a></code>

        </dt>
        <dd>Decreasing logarithmic trend</dd>
      </dl><dl><dt>

          <code><a href="manhattan.html">manhattan()</a></code>

        </dt>
        <dd>Creating a manhattan plot</dd>
      </dl><dl><dt>

          <code><a href="map.plot.html">map.plot()</a></code>

        </dt>
        <dd>Creating a genetic map plot</dd>
      </dl><dl><dt>

          <code><a href="mmec.html">mmec()</a></code>

        </dt>
        <dd><strong>m</strong>ixed <strong>m</strong>odel <strong>e</strong>quations for <strong>c</strong> coefficients</dd>
      </dl><dl><dt>

          <code><a href="mmer.html">mmer()</a></code>

        </dt>
        <dd><strong>m</strong>ixed <strong>m</strong>odel <strong>e</strong>quations for <strong>r</strong> records</dd>
      </dl><dl><dt>

          <code><a href="neMarker.html">neMarker()</a></code>

        </dt>
        <dd>Effective population size based on marker matrix</dd>
      </dl><dl><dt>

          <code><a href="overlay.html">overlay()</a></code>

        </dt>
        <dd>Overlay Matrix</dd>
      </dl><dl><dt>

          <code><a href="plot.monitor.html">pmonitor()</a></code>

        </dt>
        <dd>plot the change of VC across iterations</dd>
      </dl><dl><dt>

          <code><a href="plot_mmec.html">plot(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>plot form a LMM plot with mmec</dd>
      </dl><dl><dt>

          <code><a href="plot_mmer.html">plot(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>plot form a LMM plot with mmer</dd>
      </dl><dl><dt>

          <code><a href="predict_mmec.html">predict(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>Predict form of a LMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="predict_mmer.html">predict(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>Predict form of a LMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="propMissing.html">propMissing()</a></code>

        </dt>
        <dd>Proportion of missing data</dd>
      </dl><dl><dt>

          <code><a href="r2.html">r2()</a></code>

        </dt>
        <dd>Reliability</dd>
      </dl><dl><dt>

          <code><a href="randef.html">randef()</a></code>

        </dt>
        <dd>extracting random effects</dd>
      </dl><dl><dt>

          <code><a href="redmm.html">redmm()</a></code>

        </dt>
        <dd>Reduced Model Matrix</dd>
      </dl><dl><dt>

          <code><a href="residuals_mmec.html">residuals(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>Residuals form a GLMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="residuals_mmer.html">residuals(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>Residuals form a GLMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="rrc.html">rrc()</a></code>

        </dt>
        <dd>reduced rank covariance structure</dd>
      </dl><dl><dt>

          <code><a href="simGECorMat.html">simGECorMat()</a></code>

        </dt>
        <dd>Create a GE correlation matrix for simulation purposes.</dd>
      </dl><dl><dt>

          <code><a href="sommer-package.html">sommer</a></code>

        </dt>
        <dd><strong>So</strong>lving <strong>M</strong>ixed <strong>M</strong>odel <strong>E</strong>quations in <strong>R</strong> <br><img src="figures/mai.png" height="10%" alt="Figure: mai.png"></dd>
      </dl><dl><dt>

          <code><a href="spl2Da.html">spl2Da()</a></code>

        </dt>
        <dd>Two-dimensional penalised tensor-product of marginal B-Spline basis.</dd>
      </dl><dl><dt>

          <code><a href="spl2Db.html">spl2Db()</a></code>

        </dt>
        <dd>Two-dimensional penalised tensor-product of marginal B-Spline basis.</dd>
      </dl><dl><dt>

          <code><a href="spl2Dc.html">spl2Dc()</a></code>

        </dt>
        <dd>Two-dimensional penalised tensor-product of marginal B-Spline basis.</dd>
      </dl><dl><dt>

          <code><a href="spl2Dmats.html">spl2Dmats()</a></code>

        </dt>
        <dd>Get Tensor Product Spline Mixed Model Incidence Matrices</dd>
      </dl><dl><dt>

          <code><a href="stackTrait.html">stackTrait()</a></code>

        </dt>
        <dd>Stacking traits in a dataset</dd>
      </dl><dl><dt>

          <code><a href="summary_mmec.html">summary(<i>&lt;mmec&gt;</i>)</a></code>

        </dt>
        <dd>summary form a GLMM fitted with mmec</dd>
      </dl><dl><dt>

          <code><a href="summary_mmer.html">summary(<i>&lt;mmer&gt;</i>)</a></code>

        </dt>
        <dd>summary form a GLMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="tps.html">tps()</a></code>

        </dt>
        <dd>Get Tensor Product Spline Mixed Model Incidence Matrices</dd>
      </dl><dl><dt>

          <code><a href="tpsmmbwrapper.html">tpsmmbwrapper()</a></code>

        </dt>
        <dd>Get Tensor Product Spline Mixed Model Incidence Matrices</dd>
      </dl><dl><dt>

          <code><a href="transformConstraints.html">transformConstraints()</a></code>

        </dt>
        <dd>transformConstraints</dd>
      </dl><dl><dt>

          <code><a href="transp.html">transp()</a></code>

        </dt>
        <dd>Creating color with transparency</dd>
      </dl><dl><dt>

          <code><a href="unsm.html">unsm()</a></code>

        </dt>
        <dd>unstructured indication matrix</dd>
      </dl><dl><dt>

          <code><a href="usc.html">usc()</a></code>

        </dt>
        <dd>unstructured covariance structure</dd>
      </dl><dl><dt>

          <code><a href="usr.html">usr()</a></code>

        </dt>
        <dd>unstructured covariance structure</dd>
      </dl><dl><dt>

          <code><a href="vpredict.html">vpredict()</a></code>

        </dt>
        <dd>vpredict form of a LMM fitted with mmer</dd>
      </dl><dl><dt>

          <code><a href="vs.html">vs()</a></code>

        </dt>
        <dd>variance structure specification</dd>
      </dl><dl><dt>

          <code><a href="vsc.html">vsc()</a></code>

        </dt>
        <dd>variance structure specification</dd>
      </dl><dl><dt>

          <code><a href="vsr.html">vsr()</a></code>

        </dt>
        <dd>variance structure specification</dd>
      </dl><dl><dt>

          <code><a href="wald.html">wald.test()</a></code> <code><a href="wald.html">print(<i>&lt;wald.test&gt;</i>)</a></code>

        </dt>
        <dd>Wald Test for Model Coefficients</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

