<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>mixed model equations for r records — mmer • sommer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="mixed model equations for r records — mmer"><meta name="description" content="

The mmer function uses the Direct-Inversion Newton-Raphson or Average Information coded in C++ using the Armadillo library to optimize dense matrix operations common in genomic selection models. These algorithms are intended to be used for problems of the type c &amp;gt; r (more coefficients to estimate than records in the dataset) and/or dense matrices. For problems with sparse data, or problems of the type r &amp;gt; c (more records in the dataset than coefficients to estimate), the MME-based algorithm in the mmec function is faster and we recommend to shift to use that function."><meta property="og:description" content="

The mmer function uses the Direct-Inversion Newton-Raphson or Average Information coded in C++ using the Armadillo library to optimize dense matrix operations common in genomic selection models. These algorithms are intended to be used for problems of the type c &amp;gt; r (more coefficients to estimate than records in the dataset) and/or dense matrices. For problems with sparse data, or problems of the type r &amp;gt; c (more records in the dataset than coefficients to estimate), the MME-based algorithm in the mmec function is faster and we recommend to shift to use that function."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-116716530-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-116716530-2');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sommer</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">4.3.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-o"></span> Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/v1.sommer.quick.start.html">Quick Start</a></li>
    <li><a class="dropdown-item" href="../articles/v2.sommer.changes.and.faqs.html">FAQ</a></li>
    <li><a class="dropdown-item" href="../articles/v3.sommer.qg.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/v4.sommer.gxe.html">GxE models</a></li>
    <li><a class="dropdown-item" href="../articles/v5.sommer.vs.lme4.html">sommer vs lme4</a></li>
    <li><a class="dropdown-item" href="../articles/v6.sommer.spatial.html">Spatial</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="https://github.com/covaruber/sommer" aria-label="GitHub repository"><span class="fa fa-github"></span> GitHub</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1><strong>m</strong>ixed <strong>m</strong>odel <strong>e</strong>quations for <strong>r</strong> records</h1>

      <div class="d-none name"><code>mmer.Rd</code></div>
    </div>

    <div class="ref-description section level2">


<p>The mmer function uses the Direct-Inversion Newton-Raphson or Average Information coded in C++ using the Armadillo library to optimize dense matrix operations common in genomic selection models. These algorithms are <strong>intended to be used for problems of the type c &gt; r (more coefficients to estimate than records in the dataset) and/or dense matrices</strong>. For problems with sparse data, or problems of the type r &gt; c (more records in the dataset than coefficients to estimate), the MME-based algorithm in the <code><a href="mmec.html">mmec</a></code> function is faster and we recommend to shift to use that function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mmer</span><span class="op">(</span><span class="va">fixed</span>, <span class="va">random</span>, <span class="va">rcov</span>, <span class="va">data</span>, <span class="va">weights</span>, <span class="va">W</span>, nIters<span class="op">=</span><span class="fl">20</span>, tolParConvLL <span class="op">=</span> <span class="fl">1e-03</span>, </span>
<span>     tolParInv <span class="op">=</span> <span class="fl">1e-06</span>, init<span class="op">=</span><span class="cn">NULL</span>, constraints<span class="op">=</span><span class="cn">NULL</span>,method<span class="op">=</span><span class="st">"NR"</span>, getPEV<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>     naMethodX<span class="op">=</span><span class="st">"exclude"</span>, naMethodY<span class="op">=</span><span class="st">"exclude"</span>,returnParam<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>     dateWarning<span class="op">=</span><span class="cn">TRUE</span>,date.warning<span class="op">=</span><span class="cn">TRUE</span>,verbose<span class="op">=</span><span class="cn">TRUE</span>, reshapeOutput<span class="op">=</span><span class="cn">TRUE</span>, stepWeight<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>     emWeight<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-fixed">fixed<a class="anchor" aria-label="anchor" href="#arg-fixed"></a></dt>
<dd><p>A formula specifying the <strong>response variable(s)</strong> <strong>and fixed effects</strong>, e.g.:</p>
<p><em>response ~ covariate</em> for univariate models</p>
<p><em>cbind(response.i,response.j) ~ covariate</em> for multivariate models</p>
<p>The <code><a href="fcm.html">fcm</a>()</code> function can be used to constrain fixed effects in multi-response models.</p>
<p></p></dd>


  <dt id="arg-random">random<a class="anchor" aria-label="anchor" href="#arg-random"></a></dt>
<dd><p>A formula specifying the name of the <strong>random effects</strong>, e.g. <em>random= ~ genotype + year</em>.</p>
<p>Useful functions can be used to fit heterogeneous variances and other special models (<em>see 'Special Functions' in the Details section for more information</em>):</p>
<p><code><a href="vsr.html">vsr</a>(...,Gu,Gti,Gtc)</code> is the main function to specify variance models and special structures for random effects. On the ... argument you provide the unknown variance-covariance structures (e.g., usr,dsr,atr,csr) and the random effect where such covariance structure will be used (the random effect of interest). Gu is used to provide known covariance matrices among the levels of the random effect, Gti initial values and Gtc for constraints. Auxiliar functions for building the variance models are:</p>
<p>** <code><a href="dsr.html">dsr</a>(x)</code>, <code><a href="usr.html">usr</a>(x)</code>, <code><a href="csr.html">csr</a>(x)</code> and <code><a href="atr.html">atr</a>(x,levs)</code> can be used to specify unknown diagonal, unstructured and customized unstructured and diagonal covariance structures to be estimated by REML.</p>
<p>** <code><a href="unsm.html">unsm</a>(x)</code>, <code><a href="fixm.html">fixm</a>(x)</code> and <code><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a>(x)</code> can be used to build easily matrices to specify constraints in the Gtc argument of the <code><a href="vsr.html">vsr</a>()</code> function.</p>
<p>** <code><a href="overlay.html">overlay</a>()</code>, <code><a href="spl2Da.html">spl2Da</a>()</code>, <code><a href="spl2Db.html">spl2Db</a>()</code>, and <code><a href="leg.html">leg</a>()</code> functions can be used to specify overlayed of design matrices of random effects, two dimensional spline and random regression models within the <code><a href="vsr.html">vsr</a>()</code> function.</p>
<p><code><a href="gvsr.html">gvsr</a>(...,Gu,Guc,Gti,Gtc)</code> is an alternative function to specify general variance structures between different random effects. An special case in the indirect genetic effect models. Is similar to the vsr function but in the ... argument the different random effects are provided.</p>
<p></p></dd>

  <dt id="arg-rcov">rcov<a class="anchor" aria-label="anchor" href="#arg-rcov"></a></dt>
<dd><p>A formula specifying the name of the <strong>error term</strong>, e.g., <em>rcov= ~ units</em>.</p>
<p>Special heterogeneous and special variance models and constraints for the residual part are the same used on the random term but the name of the random effect is always "units" which can be thought as a column with as many levels as rows in the data, e.g., <em>rcov=~vsr(dsr(covariate),units)</em></p>
<p></p></dd>

  <dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame containing the variables specified in the formulas for response, fixed, and random effects.</p></dd>

  <dt id="arg-weights">weights<a class="anchor" aria-label="anchor" href="#arg-weights"></a></dt>
<dd><p>Name of the covariate for weights. To be used for the product R = Wsi*R*Wsi, where * is the matrix product, Wsi is the square root of the inverse of W and R is the residual matrix.</p></dd>

  <dt id="arg-w">W<a class="anchor" aria-label="anchor" href="#arg-w"></a></dt>
<dd><p>Alternatively, instead of providing a vector of weights the user can specify an entire W matrix (e.g., when covariances exist). To be used first to produce Wis = solve(chol(W)), and then calculate R = Wsi*R*Wsi.t(), where * is the matrix product, and R is the residual matrix. Only one of the arguments weights or W should be used. If both are indicated W will be given the preference.</p></dd>

   <dt id="arg-niters">nIters<a class="anchor" aria-label="anchor" href="#arg-niters"></a></dt>
<dd><p>Maximum number of iterations allowed.</p></dd>

  <dt id="arg-tolparconvll">tolParConvLL<a class="anchor" aria-label="anchor" href="#arg-tolparconvll"></a></dt>
<dd><p>Convergence criteria for the change in log-likelihood.</p></dd>

  <dt id="arg-tolparinv">tolParInv<a class="anchor" aria-label="anchor" href="#arg-tolparinv"></a></dt>
<dd><p>Tolerance parameter for matrix inverse used when singularities are encountered in the estimation procedure.</p></dd>

  <dt id="arg-init">init<a class="anchor" aria-label="anchor" href="#arg-init"></a></dt>
<dd><p>Initial values for the variance components. By default this is NULL and initial values for the variance components are provided by the algorithm, but in case the user want to provide initial values for ALL var-cov components this argument is functional. It has to be provided as a list, where each list element corresponds to one random effect (1x1 matrix) and if multitrait model is pursued each element of the list is a matrix of variance covariance components among traits for such random effect. Initial values can also be provided in the Gti argument of the <a href="vsr.html">vsr</a> function. Is highly encouraged to use the Gti and Gtc arguments of the <a href="vsr.html">vsr</a> function instead of this argument, but these argument can be used to provide all initial values at once</p></dd>

  <dt id="arg-constraints">constraints<a class="anchor" aria-label="anchor" href="#arg-constraints"></a></dt>
<dd><p>When initial values are provided these have to be accompanied by their constraints. See the <a href="vsr.html">vsr</a> function for more details on the constraints. Is highly encouraged to use the Gti and Gtc arguments of the <a href="vsr.html">vsr</a> function instead of this argument but these argument can be used to provide all constraints at once.</p></dd>

  <dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>This refers to the method or algorithm to be used for estimating variance components. Direct-inversion Newton-Raphson <strong>NR</strong> and Average Information <strong>AI</strong> (Tunnicliffe 1989; Gilmour et al. 1995; Lee et al. 2015).</p></dd>

  <dt id="arg-getpev">getPEV<a class="anchor" aria-label="anchor" href="#arg-getpev"></a></dt>
<dd><p>A TRUE/FALSE value indicating if the program should return the predicted error variance and variance for random effects. This option is provided since this can take a long time for certain models where p is &gt; n by a big extent.</p></dd>

   <dt id="arg-namethodx">naMethodX<a class="anchor" aria-label="anchor" href="#arg-namethodx"></a></dt>
<dd><p>One of the two possible values; "include" or "exclude". If "include" is selected then the function will impute the X matrices for fixed effects with the median value. If "exclude" is selected it will get rid of all rows with missing values for the X (fixed) covariates. The default is "exclude". The "include" option should be used carefully.</p></dd>

   <dt id="arg-namethody">naMethodY<a class="anchor" aria-label="anchor" href="#arg-namethody"></a></dt>
<dd><p>One of the three possible values; "include", "include2" or "exclude" (default) to treat the observations in response variable to be used in the estimation of variance components. The first option "include" will impute the response variables for all rows with the median value, whereas "include2" imputes the responses only for rows where there is observation(s) for at least one of the responses (only available in the multi-response models). If "exclude" is selected (default) it will get rid of rows in response(s) where missing values are present for at least one of the responses.</p></dd>

   <dt id="arg-returnparam">returnParam<a class="anchor" aria-label="anchor" href="#arg-returnparam"></a></dt>
<dd><p>A TRUE/FALSE value to indicate if the program should return the parameters to be used for fitting the model instead of fitting the model.</p></dd>

   <dt id="arg-datewarning">dateWarning<a class="anchor" aria-label="anchor" href="#arg-datewarning"></a></dt>
<dd><p>A TRUE/FALSE value to indicate if the program should warn you when is time to update the sommer package.</p></dd>

   <dt id="arg-date-warning">date.warning<a class="anchor" aria-label="anchor" href="#arg-date-warning"></a></dt>
<dd><p>A TRUE/FALSE value to indicate if the program should warn you when is time to update the sommer package. This argument will be removed soon, just left for backcompatibility.</p></dd>

  <dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>A TRUE/FALSE value to indicate if the program should return the progress of the iterative algorithm.</p></dd>

  <dt id="arg-reshapeoutput">reshapeOutput<a class="anchor" aria-label="anchor" href="#arg-reshapeoutput"></a></dt>
<dd><p>A TRUE/FALSE value to indicate if the output should be reshaped to be easier to interpret for the user, some information is missing from the multivariate models for an easy interpretation.</p></dd>

  <dt id="arg-stepweight">stepWeight<a class="anchor" aria-label="anchor" href="#arg-stepweight"></a></dt>
<dd><p>A vector of values (of length equal to the number of iterations) indicating the weight used to multiply the update (delta) for variance components at each iteration. If NULL the 1st iteration will be multiplied by 0.5, the 2nd by 0.7, and the rest by 0.9. This argument can help to avoid that variance components go outside the parameter space in the initial iterations which doesn't happen very often with the NR method but it can be detected by looking at the behavior of the likelihood. In that case you may want to give a smaller weight to the initial 8-10 iterations.</p></dd>

  <dt id="arg-emweight">emWeight<a class="anchor" aria-label="anchor" href="#arg-emweight"></a></dt>
<dd><p>A vector of values (of length equal to the number of iterations) indicating with values between 0 and 1 the weight assigned to the EM information matrix. And the values 1 - emWeight will be applied to the NR/AI information matrix to produce a joint information matrix.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>


<p>The use of this function requires a good understanding of mixed models. Please review the 'sommer.quick.start' vignette and pay attention to details like format of your random and fixed variables (e.g. character and factor variables have different properties when returning BLUEs or BLUPs, please see the 'sommer.changes.and.faqs' vignette).</p>
<p><strong>For tutorials</strong> on how to perform different analysis with sommer please look at the vignettes by typing in the terminal:</p>
<p>vignette("v1.sommer.quick.start")</p>
<p>vignette("v2.sommer.changes.and.faqs")</p>
<p>vignette("v3.sommer.qg")</p>
<p>vignette("v4.sommer.gxe")</p>

<p><strong>Citation</strong></p>
<p>Type <em>citation("sommer")</em> to know how to cite the sommer package in your publications.</p>
<p><strong>Special variance structures</strong></p>
<p><code><a href="vsr.html">vsr</a>(<a href="atr.html">atr</a>(x,levels),y)</code></p>
<p>can be used to specify heterogeneous variance for the "y" covariate at specific levels of the covariate "x", e.g., <em>random=~vsr(at(Location,c("A","B")),ID)</em> fits a variance component for ID at levels A and B of the covariate Location.</p>
<p><code><a href="vsr.html">vsr</a>(<a href="dsr.html">dsr</a>(x),y)</code></p>
<p>can be used to specify a diagonal covariance structure for the "y" covariate for all levels of the covariate "x", e.g., <em>random=~vsr(dsr(Location),ID)</em>  fits a variance component for ID at all levels of the covariate Location.</p>
<p><code><a href="vsr.html">vsr</a>(<a href="usr.html">usr</a>(x),y)</code></p>
<p>can be used to specify an unstructured covariance structure for the "y" covariate for all levels of the  covariate "x", e.g., <em>random=~vsr(usr(Location),ID)</em>  fits variance and covariance components for ID at all levels of the covariate Location.</p>
<p><code><a href="vsr.html">vsr</a>(<a href="overlay.html">overlay</a>(...,rlist=NULL,prefix=NULL))</code></p>
<p>can be used to specify overlay of design matrices between consecutive random effects specified, e.g., <em>random=~vsr(overlay(male,female))</em> overlays (overlaps) the incidence matrices for the male and female random effects to obtain a single variance component for both effects. The `rlist` argument is a list with each element being a numeric value that multiplies the incidence matrix to be overlayed. See <code><a href="overlay.html">overlay</a></code> for details.Can be combined with vsr().</p>
<p><code><a href="vsr.html">vsr</a>(<a href="leg.html">leg</a>(x,n),y)</code></p>
<p>can be used to fit a random regression model using a numerical variable <code>x</code> that marks the trayectory for the random effect <code>y</code>. The leg function can be combined with the special functions <code>dsr</code>, <code>usr</code> <code>at</code> and <code>csr</code>. For example <em>random=~vsr(leg(x,1),y)</em> or <em>random=~vsr(usr(leg(x,1)),y)</em>.</p>
<p><code><a href="vsr.html">vsr</a>(x,Gtc=<a href="fcm.html">fcm</a>(v))</code></p>
<p>can be used to constrain fixed effects in the multi-response mixed models. This is a vector that specifies if the fixed effect is to be estimated for such trait. For example <em>fixed=cbind(response.i, response.j)~vsr(Rowf, Gtc=fcm(c(1,0)))</em> means that the fixed effect Rowf should only be estimated for the first response and the second should only have the intercept.</p>
<p><code><a href="gvsr.html">gvsr</a>(x,y)</code></p>
<p>can be used to fit variance and covariance parameters between two or more random effects. For example, indirect genetic effect models.</p>
<p><code><a href="spl2Da.html">spl2Da</a>(x.coord, y.coord, at.var, at.levels))</code></p>
<p>can be used to fit a 2-dimensional spline (e.g., spatial modeling) using coordinates <code>x.coord</code> and <code>y.coord</code> (in numeric class) assuming a single variance component. The 2D spline can be fitted at specific levels using the <code>at.var</code> and <code>at.levels</code> arguments. For example <em>random=~spl2Da(x.coord=Row,y.coord=Range,at.var=FIELD)</em>.</p>
<p><code><a href="spl2Db.html">spl2Db</a>(x.coord, y.coord, at.var, at.levels))</code></p>
<p>can be used to fit a 2-dimensional spline (e.g., spatial modeling) using coordinates <code>x.coord</code> and <code>y.coord</code> (in numeric class) assuming multiple variance components. The 2D spline can be fitted at specific levels using the <code>at.var</code> and <code>at.levels</code> arguments. For example <em>random=~spl2Db(x.coord=Row,y.coord=Range,at.var=FIELD)</em>.</p>
<p><strong>S3 methods</strong></p>
<p>S3 methods are available for some parameter extraction such as <code><a href="fitted_mmer.html">fitted.mmer</a></code>, <code><a href="residuals_mmer.html">residuals.mmer</a></code>, <code><a href="summary_mmer.html">summary.mmer</a></code>, <code><a href="randef.html">randef</a></code>, <code><a href="coef_mmer.html">coef.mmer</a></code>, <code><a href="anova_mmer.html">anova.mmer</a></code>, <code><a href="plot_mmer.html">plot.mmer</a></code>, and <code><a href="predict_mmer.html">predict.mmer</a></code> to obtain adjusted means. In addition, the <code><a href="vpredict.html">vpredict</a></code> function (replacement of the pin function) can be used to estimate standard errors for linear combinations of variance components (e.g., ratios like h2).</p>
<p><strong>Additional Functions</strong></p>
<p>Additional functions for genetic analysis have been included such as relationship matrix building (<code><a href="A.mat.html">A.mat</a></code>, <code><a href="D.mat.html">D.mat</a></code>, <code><a href="E.mat.html">E.mat</a></code>, <code><a href="H.mat.html">H.mat</a></code>), build a genotypic hybrid marker matrix (<code><a href="build.HMM.html">build.HMM</a></code>), plot of genetic maps (<code><a href="map.plot.html">map.plot</a></code>), and manhattan plots (<code><a href="manhattan.html">manhattan</a></code>). If you need to build a pedigree-based relationship matrix use the <code>getA</code> function from the pedigreemm package.</p>
<p><strong>Bug report and contact</strong></p>
<p>If you have any technical questions or suggestions please post it in https://stackoverflow.com or https://stats.stackexchange.com</p>
<p>If you have any bug report please go to https://github.com/covaruber/sommer or send me an email to address it asap, just make sure you have read the vignettes carefully before sending your question.</p>
<p><strong>Example Datasets</strong></p>
<p>The package has been equiped with several datasets to learn how to use the sommer package:</p>
<p>* <code><a href="DT_halfdiallel.html">DT_halfdiallel</a></code>, <code><a href="DT_fulldiallel.html">DT_fulldiallel</a></code> and <code><a href="DT_mohring.html">DT_mohring</a></code> datasets have examples to fit half and full diallel designs.</p>
<p>* <code><a href="DT_h2.html">DT_h2</a></code> to calculate heritability</p>
<p>* <code><a href="DT_cornhybrids.html">DT_cornhybrids</a></code> and <code><a href="DT_technow.html">DT_technow</a></code> datasets to perform genomic prediction in hybrid single crosses</p>
<p>* <code><a href="DT_wheat.html">DT_wheat</a></code> dataset to do genomic prediction in single crosses in species displaying only additive effects.</p>
<p>* <code><a href="DT_cpdata.html">DT_cpdata</a></code> dataset to fit genomic prediction models within a biparental population coming from 2 highly heterozygous parents including additive, dominance and epistatic effects.</p>
<p>* <code><a href="DT_polyploid.html">DT_polyploid</a></code> to fit genomic prediction and GWAS analysis in polyploids.</p>
<p>* <code><a href="DT_gryphon.html">DT_gryphon</a></code> data contains an example of an animal model including pedigree information.</p>
<p>* <code><a href="DT_btdata.html">DT_btdata</a></code> dataset contains an animal (birds) model.</p>
<p>* <code><a href="DT_legendre.html">DT_legendre</a></code> simulated dataset for random regression model.</p>
<p>* <code><a href="DT_sleepstudy.html">DT_sleepstudy</a></code> dataset to know how to translate lme4 models to sommer models.</p>
<p>* <code><a href="DT_ige.html">DT_ige</a></code> dataset to show how to fit indirect genetic effect models.</p>
<p><strong>Models Enabled</strong></p>
<p>For details about the models enabled and more information about the covariance structures please check the help page of the package (<code><a href="sommer-package.html">sommer</a></code>).</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If all parameters are correctly indicated the program will return a list with the following information:</p>
<dl><dt>Vi</dt>
<dd><p>the inverse of the phenotypic variance matrix V^- = (ZGZ+R)^-1</p></dd>

<dt>P</dt>
<dd><p>the projection matrix  Vi - [Vi*(X*Vi*X)^-*Vi]</p></dd>

<dt>sigma</dt>
<dd><p>a list with the values of the variance-covariance components with one list element for each random effect.</p></dd>

<dt>sigma_scaled</dt>
<dd><p>a list with the values of the scaled variance-covariance components with one list element for each random effect.</p></dd>

<dt>sigmaSE</dt>
<dd><p>Hessian matrix containing the variance-covariance for the variance components. SE's can be obtained taking the square root of the diagonal values of the Hessian.</p></dd>

<dt>Beta</dt>
<dd><p>a data frame for trait BLUEs (fixed effects).</p></dd>

<dt>VarBeta</dt>
<dd><p>a variance-covariance matrix for trait BLUEs</p></dd>

<dt>U</dt>
<dd><p>a list (one element for each random effect) with a data frame for trait BLUPs.</p></dd>

<dt>VarU</dt>
<dd><p>a list (one element for each random effect) with the variance-covariance matrix for trait BLUPs.</p></dd>

<dt>PevU</dt>
<dd><p>a list (one element for each random effect) with the predicted error variance matrix for trait BLUPs.</p></dd>

<dt>fitted</dt>
<dd><p>Fitted values y.hat=XB</p></dd>

<dt>residuals</dt>
<dd><p>Residual values e = Y - XB</p></dd>

<dt>AIC</dt>
<dd><p>Akaike information criterion</p></dd>

<dt>BIC</dt>
<dd><p>Bayesian information criterion</p></dd>

<dt>convergence</dt>
<dd><p>a TRUE/FALSE statement indicating if the model converged.</p></dd>

<dt>monitor</dt>
<dd><p>The values of log-likelihood and variance-covariance components across iterations during the REML estimation.</p></dd>

<dt>percChange</dt>
<dd><p>The percent change of variance components across iterations. There should be one column less than the number of iterations. Calculated as percChange = ((x_i/x_i-1) - 1) * 100 where i is the ith iteration.</p></dd>

<dt>dL</dt>
<dd><p>The vector of first derivatives of the likelihood with respect to the ith variance-covariance component.</p></dd>

<dt>dL2</dt>
<dd><p>The matrix of second derivatives of the likelihood with respect to the i.j th variance-covariance component.</p></dd>

<dt>method</dt>
<dd><p>The method for extimation of variance components specified by the user.</p></dd>

<dt>call</dt>
<dd><p>Formula for fixed, random and rcov used.</p></dd>

<dt>constraints</dt>
<dd><p>contraints used in the mixed models for the random effects.</p></dd>

<dt>constraintsF</dt>
<dd><p>contraints used in the mixed models for the fixed effects.</p></dd>

<dt>data</dt>
<dd><p>The dataset used in the model after removing missing records for the response variable.</p></dd>

<dt>dataOriginal</dt>
<dd><p>The original dataset used in the model.</p></dd>

<dt>terms</dt>
<dd><p>The name of terms for responses, fixed, random and residual effects in the model.</p></dd>

<dt>termsN</dt>
<dd><p>The number of effects associated to fixed, random and residual effects in the model.</p></dd>

<dt>sigmaVector</dt>
<dd><p>a vectorized version of the sigma element (variance-covariance components) to match easily the standard errors of the var-cov components stored in the element sigmaSE.</p></dd>

<dt>reshapeOutput</dt>
<dd><p>The value provided to the mmer function for the argument with the same name.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>


<p>Covarrubias-Pazaran G. Genome assisted prediction of quantitative traits using the R package sommer. PLoS ONE 2016, 11(6): doi:10.1371/journal.pone.0156744</p>
<p>Covarrubias-Pazaran G. 2018. Software update: Moving the R package sommer to multivariate mixed models for genome-assisted prediction. doi: https://doi.org/10.1101/354639</p>
<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants. Second edition. Stemma Press. 390 pp.</p>
<p>Gilmour et al. 1995. Average Information REML: An efficient algorithm for variance parameter estimation in linear mixed models. Biometrics 51(4):1440-1450.</p>
<p>Kang et al. 2008. Efficient control of population structure in model organism association mapping. Genetics 178:1709-1723.</p>
<p>Lee, D.-J., Durban, M., and Eilers, P.H.C. (2013). Efficient two-dimensional smoothing with P-spline ANOVA mixed models and nested bases. Computational Statistics and Data Analysis, 61, 22 - 37.</p>
<p>Lee et al. 2015. MTG2: An efficient algorithm for multivariate linear mixed model analysis based on genomic information. Cold Spring Harbor. doi: http://dx.doi.org/10.1101/027201.</p>
<p>Maier et al. 2015. Joint analysis of psychiatric disorders increases accuracy of risk prediction for schizophrenia, bipolar disorder, and major depressive disorder. Am J Hum Genet; 96(2):283-294.</p>
<p>Rodriguez-Alvarez, Maria Xose, et al. Correcting for spatial heterogeneity in plant breeding experiments with P-splines. Spatial Statistics 23 (2018): 52-71.</p>
<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML estimates of variance components. Paper invited for the 1993 American Statistical Association Meeting, San Francisco.</p>
<p>Yu et al. 2006. A unified mixed-model method for association mapping that accounts for multiple levels of relatedness. Genetics 38:203-208.</p>
<p>Tunnicliffe W. 1989. On the use of marginal likelihood in time series model estimation. JRSS 51(1):15-27.</p>
<p>Zhang et al. 2010. Mixed linear model approach adapted for genome-wide association studies. Nat. Genet. 42:355-360.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Giovanny Covarrubias-Pazaran</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### For CRAN time limitations most lines in the </span></span></span>
<span class="r-in"><span><span class="co">#### examples are silenced with one '#' mark, </span></span></span>
<span class="r-in"><span><span class="co">#### remove them and run the examples</span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### EXAMPLES</span></span></span>
<span class="r-in"><span><span class="co">#### Different models with sommer</span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DT_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="va">DT_example</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   Name     Env Loc Year     Block Yield    Weight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33  Manistee(MSL292-A) CA.2013  CA 2013 CA.2013.1     4 -1.904711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 65          CO02024-9W CA.2013  CA 2013 CA.2013.1     5 -1.446958</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 66  Manistee(MSL292-A) CA.2013  CA 2013 CA.2013.2     5 -1.516271</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 67            MSL007-B CA.2011  CA 2011 CA.2011.2     5 -1.435510</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 68           MSR169-8Y CA.2013  CA 2013 CA.2013.1     5 -1.469051</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 103         AC05153-1W CA.2013  CA 2013 CA.2013.1     6 -1.307167</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### Univariate homogeneous variance models  ####</span></span></span>
<span class="r-in"><span><span class="co">####=========================================####</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Compound simmetry (CS) model</span></span></span>
<span class="r-in"><span><span class="va">ans1</span> <span class="op">&lt;-</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span></span>
<span class="r-in"><span>             random<span class="op">=</span> <span class="op">~</span> <span class="va">Name</span> <span class="op">+</span> <span class="va">Env</span><span class="op">:</span><span class="va">Name</span>,</span></span>
<span class="r-in"><span>             rcov<span class="op">=</span> <span class="op">~</span> <span class="va">units</span>,</span></span>
<span class="r-in"><span>             data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1      -31.2668   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2      -23.2804   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3      -20.4746   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4      -20.1501   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     5      -20.1454   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     6      -20.1454   1:16:17      0           0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Multivariate Linear Mixed Model fit by REML         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> **********************  sommer 4.3  ********************** </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          logLik      AIC      BIC Method Converge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Value -20.14538 46.29075 55.95182     NR     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variance-Covariance components:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      VarComp VarCompSE Zratio Constraint</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name.Yield-Yield       3.682     1.691  2.177   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Env:Name.Yield-Yield   5.173     1.495  3.460   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> units.Yield-Yield      4.366     0.647  6.748   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fixed effects:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Trait      Effect Estimate Std.Error t.value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Yield (Intercept)   16.496    0.6855  24.065</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Yield  EnvCA.2012   -5.777    0.7558  -7.643</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Yield  EnvCA.2013   -6.380    0.7960  -8.015</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Groups and observations:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Yield</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name        41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Env:Name   123</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the '$' sign to access results and parameters</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####===========================================####</span></span></span>
<span class="r-in"><span><span class="co">#### Univariate heterogeneous variance models  ####</span></span></span>
<span class="r-in"><span><span class="co">####===========================================####</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Compound simmetry (CS) + Diagonal (DIAG) model</span></span></span>
<span class="r-in"><span><span class="va">ans2</span> <span class="op">&lt;-</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span></span>
<span class="r-in"><span>             random<span class="op">=</span> <span class="op">~</span><span class="va">Name</span> <span class="op">+</span> <span class="fu"><a href="vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">Name</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>             rcov<span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>             data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1      -31.2668   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2      -19.8549   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3      -15.9797   1:16:17      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4      -15.4374   1:16:18      1           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     5      -15.43   1:16:18      1           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     6      -15.4298   1:16:18      1           0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Multivariate Linear Mixed Model fit by REML         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> **********************  sommer 4.3  ********************** </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          logLik      AIC      BIC Method Converge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Value -15.42983 36.85965 46.52072     NR     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variance-Covariance components:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           VarComp VarCompSE Zratio Constraint</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name.Yield-Yield            2.963     1.496  1.980   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:Name.Yield-Yield   10.146     4.507  2.251   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:Name.Yield-Yield    1.878     1.870  1.004   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:Name.Yield-Yield    6.629     2.503  2.649   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:units.Yield-Yield   4.942     1.525  3.242   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:units.Yield-Yield   5.725     1.312  4.363   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:units.Yield-Yield   2.560     0.640  4.000   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fixed effects:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Trait      Effect Estimate Std.Error t.value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Yield (Intercept)   16.508    0.8268  19.965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Yield  EnvCA.2012   -5.817    0.8575  -6.783</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Yield  EnvCA.2013   -6.412    0.9356  -6.854</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Groups and observations:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Yield</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name            41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:Name    41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:Name    41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:Name    41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ============================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the '$' sign to access results and parameters</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####===========================================####</span></span></span>
<span class="r-in"><span><span class="co">####  Univariate unstructured variance models  ####</span></span></span>
<span class="r-in"><span><span class="co">####===========================================####</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ans3</span> <span class="op">&lt;-</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Yield</span><span class="op">~</span><span class="va">Env</span>,</span></span>
<span class="r-in"><span>             random<span class="op">=</span><span class="op">~</span> <span class="fu"><a href="vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="usr.html">usr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">Name</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>             rcov<span class="op">=</span><span class="op">~</span><span class="fu"><a href="vsr.html">vsr</a></span><span class="op">(</span><span class="fu"><a href="dsr.html">dsr</a></span><span class="op">(</span><span class="va">Env</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>             data<span class="op">=</span><span class="va">DT</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1      -37.9059   1:16:18      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2      -17.9745   1:16:18      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3      -12.2427   1:16:18      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4      -11.5121   1:16:18      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     5      -11.5001   1:16:18      0           0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     6      -11.4997   1:16:18      0           0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ans3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             Multivariate Linear Mixed Model fit by REML            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *************************  sommer 4.3  ************************* </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          logLik      AIC      BIC Method Converge</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Value -11.49971 28.99943 38.66049     NR     TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variance-Covariance components:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  VarComp VarCompSE Zratio Constraint</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:Name.Yield-Yield          15.665    5.4207  2.890   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:CA.2011:Name.Yield-Yield   6.110    2.4851  2.459   Unconstr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:Name.Yield-Yield           4.530    1.8208  2.488   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:CA.2011:Name.Yield-Yield   6.384    3.0659  2.082   Unconstr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:CA.2012:Name.Yield-Yield   0.393    1.5234  0.258   Unconstr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:Name.Yield-Yield           8.597    2.4838  3.461   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:units.Yield-Yield          4.970    1.5323  3.243   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:units.Yield-Yield          5.673    1.3008  4.361   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:units.Yield-Yield          2.557    0.6393  4.000   Positive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fixed effects:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Trait      Effect Estimate Std.Error t.value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Yield (Intercept)   16.331    0.8137  20.070</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Yield  EnvCA.2012   -5.696    0.7404  -7.693</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Yield  EnvCA.2013   -6.271    0.8191  -7.656</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Groups and observations:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      Yield</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2011:Name            41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:CA.2011:Name    82</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2012:Name            41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:CA.2011:Name    82</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:CA.2012:Name    82</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CA.2013:Name            41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ==================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Use the '$' sign to access results and parameters</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># ####==========================================####</span></span></span>
<span class="r-in"><span><span class="co"># #### Multivariate homogeneous variance models ####</span></span></span>
<span class="r-in"><span><span class="co"># ####==========================================####</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ## Multivariate Compound simmetry (CS) model</span></span></span>
<span class="r-in"><span><span class="co"># DT$EnvName &lt;- paste(DT$Env,DT$Name)</span></span></span>
<span class="r-in"><span><span class="co"># ans4 &lt;- mmer(cbind(Yield, Weight) ~ Env,</span></span></span>
<span class="r-in"><span><span class="co">#               random= ~ vsr(Name, Gtc = unsm(2)) + vsr(EnvName,Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               rcov= ~ vsr(units, Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(ans4)</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ####=============================================####</span></span></span>
<span class="r-in"><span><span class="co"># #### Multivariate heterogeneous variance models  ####</span></span></span>
<span class="r-in"><span><span class="co"># ####=============================================####</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ## Multivariate Compound simmetry (CS) + Diagonal (DIAG) model</span></span></span>
<span class="r-in"><span><span class="co"># ans5 &lt;- mmer(cbind(Yield, Weight) ~ Env,</span></span></span>
<span class="r-in"><span><span class="co">#               random= ~ vsr(Name, Gtc = unsm(2)) + vsr(dsr(Env),Name, Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               rcov= ~ vsr(dsr(Env),units, Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(ans5)</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ####===========================================####</span></span></span>
<span class="r-in"><span><span class="co"># #### Multivariate unstructured variance models ####</span></span></span>
<span class="r-in"><span><span class="co"># ####===========================================####</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ans6 &lt;- mmer(cbind(Yield, Weight) ~ Env,</span></span></span>
<span class="r-in"><span><span class="co">#               random= ~ vsr(usr(Env),Name, Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               rcov= ~ vsr(dsr(Env),units, Gtc = unsm(2)),</span></span></span>
<span class="r-in"><span><span class="co">#               data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(ans6)</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># #### EXAMPLE SET 2</span></span></span>
<span class="r-in"><span><span class="co"># #### 2 variance components</span></span></span>
<span class="r-in"><span><span class="co"># #### one random effect with variance covariance structure</span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># data("DT_cpdata")</span></span></span>
<span class="r-in"><span><span class="co"># DT &lt;- DT_cpdata</span></span></span>
<span class="r-in"><span><span class="co"># GT &lt;- GT_cpdata</span></span></span>
<span class="r-in"><span><span class="co"># MP &lt;- MP_cpdata</span></span></span>
<span class="r-in"><span><span class="co"># head(DT)</span></span></span>
<span class="r-in"><span><span class="co"># GT[1:4,1:4]</span></span></span>
<span class="r-in"><span><span class="co"># #### create the variance-covariance matrix</span></span></span>
<span class="r-in"><span><span class="co"># A &lt;- A.mat(GT)</span></span></span>
<span class="r-in"><span><span class="co"># #### look at the data and fit the model</span></span></span>
<span class="r-in"><span><span class="co"># mix1 &lt;- mmer(Yield~1,</span></span></span>
<span class="r-in"><span><span class="co">#              random=~vsr(id, Gu=A) + Rowf,</span></span></span>
<span class="r-in"><span><span class="co">#              rcov=~units,</span></span></span>
<span class="r-in"><span><span class="co">#              data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(mix1)$varcomp</span></span></span>
<span class="r-in"><span><span class="co"># #### calculate heritability</span></span></span>
<span class="r-in"><span><span class="co"># vpredict(mix1, h1 ~ V1/(V1+V3) )</span></span></span>
<span class="r-in"><span><span class="co"># #### multi trait example</span></span></span>
<span class="r-in"><span><span class="co"># mix2 &lt;- mmer(cbind(Yield,color)~1,</span></span></span>
<span class="r-in"><span><span class="co">#               random = ~ vsr(id, Gu=A, Gtc = unsm(2)) + # unstructured at trait level</span></span></span>
<span class="r-in"><span><span class="co">#                             vsr(Rowf, Gtc=diag(2)) + # diagonal structure at trait level</span></span></span>
<span class="r-in"><span><span class="co">#                                 vsr(Colf, Gtc=diag(2)), # diagonal structure at trait level</span></span></span>
<span class="r-in"><span><span class="co">#               rcov = ~ vsr(units, Gtc = unsm(2)), # unstructured at trait level</span></span></span>
<span class="r-in"><span><span class="co">#               data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(mix2)</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># #### EXAMPLE SET 3</span></span></span>
<span class="r-in"><span><span class="co"># #### comparison with lmer, install 'lme4' </span></span></span>
<span class="r-in"><span><span class="co"># #### and run the code below</span></span></span>
<span class="r-in"><span><span class="co"># ####=========================================####</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># #### lmer cannot use var-cov matrices so we will not </span></span></span>
<span class="r-in"><span><span class="co"># #### use them in this comparison example</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># library(lme4)</span></span></span>
<span class="r-in"><span><span class="co"># library(sommer)</span></span></span>
<span class="r-in"><span><span class="co"># data("DT_cornhybrids")</span></span></span>
<span class="r-in"><span><span class="co"># DT &lt;- DT_cornhybrids</span></span></span>
<span class="r-in"><span><span class="co"># DTi &lt;- DTi_cornhybrids</span></span></span>
<span class="r-in"><span><span class="co"># GT &lt;- GT_cornhybrids</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># fm1 &lt;- lmer(Yield ~ Location + (1|GCA1) + (1|GCA2) + (1|SCA),</span></span></span>
<span class="r-in"><span><span class="co">#             data=DT )</span></span></span>
<span class="r-in"><span><span class="co"># out &lt;- mmer(Yield ~ Location,</span></span></span>
<span class="r-in"><span><span class="co">#              random = ~ GCA1 + GCA2 + SCA,</span></span></span>
<span class="r-in"><span><span class="co">#              rcov = ~ units,</span></span></span>
<span class="r-in"><span><span class="co">#              data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(fm1)</span></span></span>
<span class="r-in"><span><span class="co"># summary(out)</span></span></span>
<span class="r-in"><span><span class="co"># ### same BLUPs for GCA1, GCA2, SCA than lme4</span></span></span>
<span class="r-in"><span><span class="co"># plot(out$U$GCA1$Yield, ranef(fm1)$GCA1[,1])</span></span></span>
<span class="r-in"><span><span class="co"># plot(out$U$GCA2$Yield, ranef(fm1)$GCA2[,1])</span></span></span>
<span class="r-in"><span><span class="co"># vv=which(abs(out$U$SCA$Yield) &gt; 0)</span></span></span>
<span class="r-in"><span><span class="co"># plot(out$U$SCA$Yield[vv], ranef(fm1)$SCA[,1])</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># ### a more complex model specifying which locations</span></span></span>
<span class="r-in"><span><span class="co"># head(DT)</span></span></span>
<span class="r-in"><span><span class="co"># out2 &lt;- mmer(Yield ~ Location,</span></span></span>
<span class="r-in"><span><span class="co">#               random = ~ vsr(atr(Location,c("3","4")),GCA2) +</span></span></span>
<span class="r-in"><span><span class="co">#                          vsr(atr(Location,c("3","4")),SCA),</span></span></span>
<span class="r-in"><span><span class="co">#               rcov = ~ vsr(dsr(Location),units),</span></span></span>
<span class="r-in"><span><span class="co">#               data=DT)</span></span></span>
<span class="r-in"><span><span class="co"># summary(out2)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/covaruber" class="external-link">Giovanny Covarrubias-Pazaran</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

